# TRIAD
Scripts for analysing the composition of libraries generated by TRIAD cloning

## Setup and installation
### Alignment software
- PEAR: https://cme.h-its.org/exelixis/web/software/pear/
- samtools version 1.9
- htslib version 1.9
- bowtie2 version 2.3
- emboss version 6.6

All are currently (August 2019) available as packages in standard Ubuntu repositories, apart from PEAR, which now requires a separate download due to licensing.

The versions given here were used to process the data; compatibility with newer or older versions is unclear.

### Python and Conda dependencies
These are a set of scripts that was developed and tested within the Anaconda tools. The environment is given in TRIAD.yml. To install, go to the TRIAD directory and use:
```
conda env create -f TRIAD.yml
```

Then activate the environment with `source activate TRIAD` or `conda activate TRIAD`. The latter is preferred for modern conda v.4.4 or higher.

While BioPython is included in the conda environment file, you may run into an issue where BioPython cannot be loaded. The workaround is to first install `pip3` with your preferred package manager, then install BioPython with `sudo pip3 install -c conda-forge biopython`. 

### Short version of scripts

1a. If working on a cluster and it is difficult to install PEAR, assemble the reads separately:
```
pear -f $forwardReads -r $reverseReads -o $baseName.$activity --keep-original --min-overlap 5 --min-assembly-length 0 --quality-threshold 15 --max-uncalled-base 0.01
```
1b. Run `count.sh reads_fw reads_rv reference.fa base_name activity` or `count_PC.sh [arguments]` as appropriate to environment.

Arguments:
- reads_fw : (only for PC version) path to forward fastq.gz reads
- reads_rv: (only for PC version) path to reverse reads
- reference.fa : filepath to reference fasta file
- base_name : Usually a shorthand for what gene we are looking at, eg. PTE
- activity : A label for what fraction / activity gate / input library these reads came from

Summary of the scripts:
#### Preprocessing
- If using paired end reads: merge reads with PEAR. Take the opportunity to filter out very broken data.
- Align all reads against reference.
- While we have SAM files, take the opportunity to calculate depth / position.
- Extract reads that are correctly mapped, keep the name.
- Throw away reads the fully match reference. This is faster than NW alignment for all reads later.
- Feed interesting reads to EMBOSS Needleman-Wunsch aligner.

#### Counting substitutions, deletions and combinations for each sequencing file
- Load a dictionary of all interesting mutations we're considering
- Read a read+reference into a SeqRecord
- Various checks that the read is not broken
    Barcodes: As long as the read does not contain insertions, the barcode is ignored and does not contribute to detected mutations
- If the mutation is defined as interesting, figure out what kind it is and add to valid_counts dictionary
- Add the mutation to a dictionary counting everything
- Save both dictionaries for later viewing.
 
 #### Inspect results and generate figures
 Start a jupyter notebook with `jupyter lab` and have a look at results in `TRIAD_composition_figures.ipynb`.
