{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {},
   "outputs": [],
   "source": [
    "#!/usr/bin/env python3\n",
    "\n",
    "import sys\n",
    "import pickle\n",
    "import re\n",
    "import os\n",
    "import csv\n",
    "import pprint\n",
    "\n",
    "import pandas as pd\n",
    "import numpy as np\n",
    "import matplotlib.pyplot as plt\n",
    "from collections import defaultdict\n",
    "\n",
    "# Demand Python 3.\n",
    "if sys.version_info[0] < 3:\n",
    "    print(\"Python 3 is required, but you are using Python %i.%i.%i\") % (\n",
    "        sys.version_info[0], sys.version_info[1], sys.version_info[2])\n",
    "    sys.exit(1)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "We need to change directory to wherever the input files are located."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {},
   "outputs": [],
   "source": [
    "os.chdir(\"/mnt/c//Users/Maya/Dropbox/PTE_results\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Later, load in the results for analysis\n",
    "with open('S6_full_all.p', 'rb') as f:\n",
    "    all_ref = pickle.load(f)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "dict_keys(['d9', 'd6', 'i6', 'd3', 'i3', 'i9'])\n"
     ]
    }
   ],
   "source": [
    "print(all_ref['S6'].keys())"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "The `all_ref` dictionary contains all information collected from sequenced, aggregated by protein background (here wtPTE, dubbed S6 for all fractions) and then by fraction / type of library. There is no filterin on this point, so `all_ref` includes all sequencing error, frameshifts, variants with extra point mutations etc.\n",
    "\n",
    "Each key in the dictionary takes the format:\n",
    "\n",
    "```\n",
    "protein_mutation : {'total': integer_count, \n",
    "                    'dna_hgvs': {'dna_in_HGVS_format': count},\n",
    "                    'dna': {'dna_in_tuple_format': count} # this is for each DNA mutation that generates that protein mutation\n",
    "                   }\n",
    "```\n",
    "So, if we want to know all mutations in a certain position in a library, we need `all_ref['S6']['d3](protein_tuple)['dna'].items()` or some such."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 66,
   "metadata": {},
   "outputs": [],
   "source": [
    "def classify_dna(dna_error):\n",
    "    \"\"\"\n",
    "    Given a dna_error tuple, it describes the type of mutation: \n",
    "    d(eletion), i(nsertion), s(ubstitution), f(rameshift), b(roken), wt \n",
    "    or a combination of these.\n",
    "    \"\"\"\n",
    "    if dna_error is None:  # empty or broken reads\n",
    "        return 'b'\n",
    "    elif dna_error == ():\n",
    "        return 'wt'\n",
    "    elif len(dna_error) > 1:\n",
    "        # expect substitutions\n",
    "        if dna_error[-1][1] == 'f':  # frameshifts are always the last mutation\n",
    "            return 'f'\n",
    "        else:\n",
    "            for k in range(len(dna_error)):\n",
    "                if dna_error[k][1] == 'i':\n",
    "                    return 'si'\n",
    "                elif dna_error[k][1] == 'd':\n",
    "                    return 'sd'\n",
    "            return 's'\n",
    "    elif len(dna_error) == 1:\n",
    "        if dna_error[0][1] == 'f':\n",
    "            return 'f'\n",
    "        elif dna_error[0][1] == 'd':\n",
    "            return 'd' + dna_error[0][2]  # length of deletion\n",
    "        elif dna_error[0][1] == 'i':\n",
    "            return 'i' + str(len(dna_error[0][2]))\n",
    "        elif dna_error[0][1] == 's':\n",
    "            return 's'\n",
    "    else:\n",
    "        return 'b'\n",
    "\n",
    "def classify_protein(mutation):\n",
    "    \"\"\"\n",
    "    Detect whether a protein mutation is an insertion/deletion/substituion and if more than 1, (non)consecutive\n",
    "    :param mutation:\n",
    "    :return:\n",
    "    \"\"\"\n",
    "    if mutation is None:  # came from empty or broken reads\n",
    "        return 'b'\n",
    "    elif mutation == ():\n",
    "        return 'b'\n",
    "    else:\n",
    "        m = []\n",
    "        for pos in range(len(mutation)):\n",
    "            t = mutation[pos][1]\n",
    "            if t != 'i':\n",
    "                m.append(t)\n",
    "            else:\n",
    "                m.append(t + str(len(mutation[pos][2])))\n",
    "        # need to distinguish between consecutive mutations and likely sequencing errors\n",
    "        if 'f' in m:\n",
    "            return 'f'\n",
    "        elif len(m) == 1:\n",
    "            if mutation[0][-1] == '*':\n",
    "                return '*'\n",
    "            else:\n",
    "                return ''.join(m)\n",
    "        elif len(m) >= 1:\n",
    "            if mutation[-1][-1] == '*':\n",
    "                return '*'\n",
    "            else:\n",
    "                c = is_mutation_consecutive(mutation) + '-' + ''.join(m)\n",
    "                return c\n",
    "\n",
    "\n",
    "def is_mutation_consecutive(mutation):\n",
    "    \"\"\"\n",
    "    If only consecutve amino acids are affected, return 'c', else return 'nc'\n",
    "    :param mutation:\n",
    "    :return:\n",
    "    \"\"\"\n",
    "    for pos in range(1, len(mutation)):\n",
    "        if mutation[pos][0] != (mutation[pos - 1][0] + 1) :\n",
    "            return 'nc'\n",
    "    return 'c'\n",
    "\n",
    "\n",
    "def dna_protein_correspondence(dna_type):\n",
    "    \"\"\"\n",
    "    Given a type of dna mutation (d3, d6 etc.), it translates this\n",
    "    into what effect this should have on protein for target mutations.\n",
    "    \"\"\"\n",
    "    dna_to_prot_type = {'d3':['d', 'c-sd', '*'], 'd6':['c-dd', 'c-sdd', '*'], 'd9':['c-ddd', 'c-sddd', '*'],\n",
    "                        'i3':['i1','c-si1', '*'], 'i6': ['i2','c-si2', '*'], 'i9':['i3','c-si3', '*']\n",
    "                       }\n",
    "    \n",
    "    try:\n",
    "        return dna_to_prot_type[dna_type]\n",
    "    except KeyError:\n",
    "        return []       "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "metadata": {},
   "outputs": [],
   "source": [
    "def get_dna_composition(all_references, cutoff=10):\n",
    "    \"\"\"\n",
    "    Classify all detected mutations by what kind they are: substitution, deletion, insertion (including length), frameshift, or other.\n",
    "    Collect the results in a dataframe, both for mutation counts and associated sequencing reads.\n",
    "    Relatively simple logic because two different DNA mutations can give rise to one protein mutation, but not vice versa.\n",
    "    \"\"\"\n",
    "\n",
    "    dna_count = {}\n",
    "    dna_reads = {}\n",
    "    \n",
    "    dna_to_stop = {'d3': 0, 'd6': 0, 'd9': 0}\n",
    "    \n",
    "    for background in all_references.keys():\n",
    "        for fraction in all_references[background].keys():\n",
    "            # print('Analysing background {0} and fraction {1}. Unusal mutations: '.format(background, fraction))\n",
    "            distinct_mutations = 0\n",
    "            total_count = 0\n",
    "            dna_count[background + '.' + fraction] = {'s': 0, 'd3': 0, 'd6': 0, 'd9': 0, 'i3': 0, 'i6': 0, 'i9': 0, 'f': 0,\n",
    "                                                'b': 0, 'other': 0, 'sd': 0, 'si': 0}\n",
    "            dna_reads[background + '.' + fraction] = {'s': 0, 'd3': 0, 'd6': 0, 'd9': 0, 'i3': 0, 'i6': 0, 'i9': 0, 'f': 0,\n",
    "                                                'b': 0, 'other': 0, 'sd': 0, 'si': 0}\n",
    "            \n",
    "            # now look through all protein mutations in the fraction, expected and not\n",
    "            for prot_mutation, prot_entry in all_references[background][fraction].items():\n",
    "                if prot_entry['total'] < cutoff:\n",
    "                    continue\n",
    "                    # find all DNA entries with high enough counts\n",
    "                    # same protein mutation can arise from multiple DNA mutations\n",
    "                    \n",
    "                for dna_mutation, dna_c in prot_entry['dna'].items():\n",
    "\n",
    "                    if dna_c < cutoff:\n",
    "                        continue\n",
    "\n",
    "                    # sometimes a mutations will have a sequencing error / silent substitution elsewhere, which biases the count\n",
    "                    dna_mut_type = classify_dna(dna_mutation)\n",
    "                    prot_mut_type = classify_protein(prot_mutation)\n",
    "\n",
    "                    if dna_mut_type == 'wt':\n",
    "                        continue\n",
    "                    elif dna_mut_type == 'b':\n",
    "                        continue\n",
    "                    \n",
    "                    # it's possible that the DNA mutation looks valid but there is something off on the protein level\n",
    "                    # skip those\n",
    "\n",
    "                    if prot_mut_type is None:\n",
    "                        print(\"None\", prot_mut_type, prot_mutation, dna_mut_type, dna_mutation)\n",
    "                        continue\n",
    "                    elif (dna_mut_type in ['d3', 'd6', 'd9', 'i3', 'i6', 'i9']) and (prot_mut_type not in dna_protein_correspondence(dna_mut_type)):\n",
    "                        continue                       \n",
    "\n",
    "                    try:\n",
    "                        dna_count[background + '.' + fraction][dna_mut_type] += 1\n",
    "                        dna_reads[background + '.' + fraction][dna_mut_type] += dna_c\n",
    "                        distinct_mutations += 1\n",
    "                        total_count += dna_c                     \n",
    "                    except KeyError:\n",
    "                        dna_count[background + '.' + fraction]['other'] += 1\n",
    "                        dna_reads[background + '.' + fraction]['other'] += dna_c\n",
    "\n",
    "\n",
    "    return dna_count, dna_reads"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "metadata": {},
   "outputs": [],
   "source": [
    "def get_full_protein_composition(all_references, cutoff=10):\n",
    "    \"\"\"\n",
    "    This version does no filtering for correspondence between DNA and protein mutations, just checking what we see in the library.\n",
    "    Probably want to use cutoff >1 for deletion libraries.\n",
    "    \"\"\"\n",
    "    protein_count = {}\n",
    "    protein_reads = {}\n",
    "    \n",
    "    for background in all_references.keys():\n",
    "        for fraction in all_references[background].keys():\n",
    "            distinct_mutations = 0\n",
    "            total_count = 0\n",
    "            \n",
    "            protein_count[background + '.' + fraction] = {'s': 0, 'd': 0, 'c-dd': 0, 'c-ddd': 0, 'i1': 0, 'i2': 0,\n",
    "                            'i3': 0, 'c-ss': 0, 'c-sd': 0, 'c-sdd': 0, 'c-sddd': 0, 'c-si1': 0, 'c-si2': 0, 'c-si3': 0,\n",
    "                            '*': 0, 'f': 0, 'other': 0, 'b': 0}\n",
    "            protein_reads[background + '.' + fraction] = {'s': 0, 'd': 0, 'c-dd': 0, 'c-ddd': 0, 'i1': 0, 'i2': 0,\n",
    "                            'i3': 0, 'c-ss': 0, 'c-sd': 0, 'c-sdd': 0, 'c-sddd': 0, 'c-si1': 0, 'c-si2': 0, 'c-si3': 0,\n",
    "                            '*': 0, 'f': 0, 'other': 0, 'b': 0}\n",
    "            \n",
    "            for prot_mutation, entry in all_references[background][fraction].items():\n",
    "                mut_total = entry['total']\n",
    "                if entry['total'] < cutoff:\n",
    "                    continue\n",
    "                    \n",
    "                # find all DNA entries with high enough counts\n",
    "                # make sure the contributing DNA mutatios don't involve spurious extra substitutions\n",
    "                prot_type = classify_protein(prot_mutation)\n",
    "\n",
    "                try:\n",
    "                    protein_count[background + '.' + fraction][prot_type] += 1\n",
    "                    protein_reads[background + '.' + fraction][prot_type] += mut_total\n",
    "                    distinct_mutations += 1\n",
    "                    total_count += mut_total\n",
    "                except KeyError:\n",
    "                    # print(prot_type, mutation, mut_total)\n",
    "                    protein_count[background + '.' + fraction]['other'] += 1\n",
    "                    protein_reads[background + '.' + fraction]['other'] += mut_total\n",
    "\n",
    "    return protein_count, protein_reads"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 54,
   "metadata": {},
   "outputs": [],
   "source": [
    "def get_filetered_protein_composition(all_references, cutoff=5):\n",
    "\n",
    "    \"\"\"\n",
    "    This version does no filtering for correspondence between DNA and protein mutations, just checking what we see in the library.\n",
    "    Probably want to use cutoff >1 for deletion libraries.\n",
    "    \"\"\"\n",
    "    protein_count = {}\n",
    "    protein_reads = {}\n",
    "    allowed_dna = ['d3', 'd6', 'd9', 'i3', 'i6', 'i9']\n",
    "    allowed_protein = [ 'd', 'c-sd', 'c-dd', 'c-sdd', 'c-ddd','c-sddd', 'i1','c-si1', 'i2','c-si2', 'i3','c-si3','*','s', 'c-ss']\n",
    "    \n",
    "    for background in all_references.keys():\n",
    "        for fraction in all_references[background].keys():\n",
    "            distinct_mutations = 0\n",
    "            total_count = 0\n",
    "            \n",
    "            protein_count[background + '.' + fraction] = {'s': 0, 'd': 0, 'c-dd': 0, 'c-ddd': 0, 'i1': 0, 'i2': 0,\n",
    "                            'i3': 0, 'c-ss': 0, 'c-sd': 0, 'c-sdd': 0, 'c-sddd': 0, 'c-si1': 0, 'c-si2': 0, 'c-si3': 0,\n",
    "                            '*': 0}\n",
    "            protein_reads[background + '.' + fraction] = {'s': 0, 'd': 0, 'c-dd': 0, 'c-ddd': 0, 'i1': 0, 'i2': 0,\n",
    "                            'i3': 0, 'c-ss': 0, 'c-sd': 0, 'c-sdd': 0, 'c-sddd': 0, 'c-si1': 0, 'c-si2': 0, 'c-si3': 0,\n",
    "                            '*': 0}\n",
    "            \n",
    "            for prot_mutation, entry in all_references[background][fraction].items():\n",
    "                mut_total = entry['total']\n",
    "                if entry['total'] < cutoff:\n",
    "                    continue\n",
    "                # any valid protein muttion should be within the ones expected in the fraction\n",
    "                # find all DNA entries with high enough counts\n",
    "                # make sure the contributing DNA mutatios don't involve spurious extra substitutions\n",
    "                prot_type = classify_protein(prot_mutation)\n",
    "                if prot_type not in allowed_protein:\n",
    "                    continue\n",
    "\n",
    "                # make sure that the protein mutation arises from what we consider a valid DNA mutation\n",
    "                # need special handling for STOPs\n",
    "                valid_dna = False\n",
    "                for dna_error, dna_c in entry['dna'].items():\n",
    "                    dna_type = classify_dna(dna_error)\n",
    "                    if prot_type == '*':\n",
    "                        if dna_type == fraction:\n",
    "                            valid_dna = True\n",
    "                            break\n",
    "                    elif (dna_type in allowed_dna) and prot_type in dna_protein_correspondence(dna_type):\n",
    "                        valid_dna = True\n",
    "                        break\n",
    "                \n",
    "                if valid_dna:\n",
    "                    protein_count[background + '.' + fraction][prot_type] += 1\n",
    "                    protein_reads[background + '.' + fraction][prot_type] += mut_total\n",
    "                    distinct_mutations += 1\n",
    "                    total_count += mut_total\n",
    "\n",
    "    return protein_count, protein_reads"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Baseline composition of libraries"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Before we get started with composition of the actual libraries, let's have a look at the baseline: what we could detect if all mutations were represented in the library, once."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 47,
   "metadata": {},
   "outputs": [],
   "source": [
    "with open('baseline/PTE_baseline_d369i3.p', 'rb') as f:\n",
    "    base_long = pickle.load(f)\n",
    "\n",
    "base = {'S6': {} }\n",
    "for k, fraction in base_long['S6'].items():\n",
    "    new_key = k[-2:]\n",
    "    base['S6'][new_key] = fraction"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 73,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>S6.d3</th>\n",
       "      <th>S6.d6</th>\n",
       "      <th>S6.d9</th>\n",
       "      <th>S6.i3</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>d3</th>\n",
       "      <td>748</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>d6</th>\n",
       "      <td>0</td>\n",
       "      <td>747</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>d9</th>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>730</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>i3</th>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>46170</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "    S6.d3  S6.d6  S6.d9  S6.i3\n",
       "d3    748      0      0      0\n",
       "d6      0    747      0      0\n",
       "d9      0      0    730      0\n",
       "i3      0      0      0  46170"
      ]
     },
     "execution_count": 73,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "dna_count, dna_reads = get_dna_composition(base, 1)\n",
    "protein_count, protein_reads = get_full_protein_composition(base, 1)\n",
    "filt_prot_count, filt_prot_reads = get_filetered_protein_composition(base, 1)\n",
    "\n",
    "\n",
    "row_index_dna = ['d3', 'd6', 'd9', 'i3']\n",
    "fractions_base = ['S6.d3', 'S6.d6', 'S6.d9', 'S6.i3']\n",
    "\n",
    "df_dna_count = pd.DataFrame(dna_count, index=row_index_dna, columns=fractions_base)\n",
    "df_dna_reads = pd.DataFrame(dna_reads, index=row_index_dna, columns=fractions_base)\n",
    "\n",
    "df_dna_count"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 74,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>S6.d3</th>\n",
       "      <th>S6.d6</th>\n",
       "      <th>S6.d9</th>\n",
       "      <th>S6.i3</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>d</th>\n",
       "      <td>325</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>c-sd</th>\n",
       "      <td>259</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>c-dd</th>\n",
       "      <td>0</td>\n",
       "      <td>328</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>c-sdd</th>\n",
       "      <td>0</td>\n",
       "      <td>252</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>c-ddd</th>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>328</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>c-sddd</th>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>217</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>i1</th>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>6661</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>c-si1</th>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>5887</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>*</th>\n",
       "      <td>6</td>\n",
       "      <td>10</td>\n",
       "      <td>6</td>\n",
       "      <td>456</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "        S6.d3  S6.d6  S6.d9  S6.i3\n",
       "d         325      0      0      0\n",
       "c-sd      259      0      0      0\n",
       "c-dd        0    328      0      0\n",
       "c-sdd       0    252      0      0\n",
       "c-ddd       0      0    328      0\n",
       "c-sddd      0      0    217      0\n",
       "i1          0      0      0   6661\n",
       "c-si1       0      0      0   5887\n",
       "*           6     10      6    456"
      ]
     },
     "execution_count": 74,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "row_index_protein = [ 'd', 'c-sd', 'c-dd', 'c-sdd', 'c-ddd','c-sddd', 'i1','c-si1','*']\n",
    "df_protein_reads = pd.DataFrame(protein_reads, index=row_index_protein, columns=fractions_base)\n",
    "df_protein_count = pd.DataFrame(protein_count, index=row_index_protein, columns=fractions_base)\n",
    "\n",
    "df_protein_count"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 75,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>S6.d3</th>\n",
       "      <th>S6.d6</th>\n",
       "      <th>S6.d9</th>\n",
       "      <th>S6.i3</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>d</th>\n",
       "      <td>325</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>c-sd</th>\n",
       "      <td>259</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>c-dd</th>\n",
       "      <td>0</td>\n",
       "      <td>327</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>c-sdd</th>\n",
       "      <td>0</td>\n",
       "      <td>252</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>c-ddd</th>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>328</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>c-sddd</th>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>217</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>i1</th>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>6661</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>c-si1</th>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>5887</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>*</th>\n",
       "      <td>6</td>\n",
       "      <td>10</td>\n",
       "      <td>6</td>\n",
       "      <td>456</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "        S6.d3  S6.d6  S6.d9  S6.i3\n",
       "d         325      0      0      0\n",
       "c-sd      259      0      0      0\n",
       "c-dd        0    327      0      0\n",
       "c-sdd       0    252      0      0\n",
       "c-ddd       0      0    328      0\n",
       "c-sddd      0      0    217      0\n",
       "i1          0      0      0   6661\n",
       "c-si1       0      0      0   5887\n",
       "*           6     10      6    456"
      ]
     },
     "execution_count": 75,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df_protein_reads_filt = pd.DataFrame(filt_prot_count, index=row_index_protein, columns=fractions_base)\n",
    "\n",
    "df_protein_reads_filt"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Composition of real libraries\n",
    "\n",
    "## Table 1\n",
    "\n",
    "1. Generate CSV files that give overall composition of libraries on protein and DNA level"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 23,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>S6.d3</th>\n",
       "      <th>S6.d6</th>\n",
       "      <th>S6.d9</th>\n",
       "      <th>S6.i3</th>\n",
       "      <th>S6.i6</th>\n",
       "      <th>S6.i9</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>d3</th>\n",
       "      <td>639</td>\n",
       "      <td>224</td>\n",
       "      <td>278</td>\n",
       "      <td>208</td>\n",
       "      <td>210</td>\n",
       "      <td>266</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>d6</th>\n",
       "      <td>28</td>\n",
       "      <td>690</td>\n",
       "      <td>53</td>\n",
       "      <td>124</td>\n",
       "      <td>176</td>\n",
       "      <td>169</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>d9</th>\n",
       "      <td>5</td>\n",
       "      <td>241</td>\n",
       "      <td>613</td>\n",
       "      <td>80</td>\n",
       "      <td>66</td>\n",
       "      <td>75</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>i3</th>\n",
       "      <td>2</td>\n",
       "      <td>10</td>\n",
       "      <td>0</td>\n",
       "      <td>20872</td>\n",
       "      <td>310</td>\n",
       "      <td>154</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>i6</th>\n",
       "      <td>2</td>\n",
       "      <td>5</td>\n",
       "      <td>5</td>\n",
       "      <td>4</td>\n",
       "      <td>107165</td>\n",
       "      <td>932</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>i9</th>\n",
       "      <td>0</td>\n",
       "      <td>8</td>\n",
       "      <td>1</td>\n",
       "      <td>18</td>\n",
       "      <td>18</td>\n",
       "      <td>103720</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>s</th>\n",
       "      <td>6140</td>\n",
       "      <td>7303</td>\n",
       "      <td>6909</td>\n",
       "      <td>18296</td>\n",
       "      <td>18923</td>\n",
       "      <td>14528</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>sd</th>\n",
       "      <td>2333</td>\n",
       "      <td>1901</td>\n",
       "      <td>1785</td>\n",
       "      <td>1038</td>\n",
       "      <td>1878</td>\n",
       "      <td>2164</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>si</th>\n",
       "      <td>4</td>\n",
       "      <td>474</td>\n",
       "      <td>7</td>\n",
       "      <td>4712</td>\n",
       "      <td>6600</td>\n",
       "      <td>5586</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>other</th>\n",
       "      <td>35</td>\n",
       "      <td>167</td>\n",
       "      <td>252</td>\n",
       "      <td>332</td>\n",
       "      <td>378</td>\n",
       "      <td>422</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "       S6.d3  S6.d6  S6.d9  S6.i3   S6.i6   S6.i9\n",
       "d3       639    224    278    208     210     266\n",
       "d6        28    690     53    124     176     169\n",
       "d9         5    241    613     80      66      75\n",
       "i3         2     10      0  20872     310     154\n",
       "i6         2      5      5      4  107165     932\n",
       "i9         0      8      1     18      18  103720\n",
       "s       6140   7303   6909  18296   18923   14528\n",
       "sd      2333   1901   1785   1038    1878    2164\n",
       "si         4    474      7   4712    6600    5586\n",
       "other     35    167    252    332     378     422"
      ]
     },
     "execution_count": 23,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "dna_count, dna_reads = get_dna_composition(all_ref, 1)\n",
    "\n",
    "row_index_dna = ['d3', 'd6', 'd9', 'i3', 'i6', 'i9', 's', 'sd', 'si', 'other']\n",
    "fractions = ['S6.d3', 'S6.d6', 'S6.d9', 'S6.i3', 'S6.i6', 'S6.i9']\n",
    "\n",
    "df_dna_count = pd.DataFrame(dna_count, index=row_index_dna, columns=fractions)\n",
    "df_dna_reads = pd.DataFrame(dna_reads, index=row_index_dna, columns=fractions)\n",
    "\n",
    "df_dna_count"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 70,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>S6.d3</th>\n",
       "      <th>S6.d6</th>\n",
       "      <th>S6.d9</th>\n",
       "      <th>S6.i3</th>\n",
       "      <th>S6.i6</th>\n",
       "      <th>S6.i9</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>d</th>\n",
       "      <td>302</td>\n",
       "      <td>146</td>\n",
       "      <td>188</td>\n",
       "      <td>143</td>\n",
       "      <td>151</td>\n",
       "      <td>183</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>c-sd</th>\n",
       "      <td>267</td>\n",
       "      <td>73</td>\n",
       "      <td>77</td>\n",
       "      <td>148</td>\n",
       "      <td>157</td>\n",
       "      <td>173</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>c-dd</th>\n",
       "      <td>19</td>\n",
       "      <td>322</td>\n",
       "      <td>42</td>\n",
       "      <td>82</td>\n",
       "      <td>135</td>\n",
       "      <td>128</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>c-sdd</th>\n",
       "      <td>11</td>\n",
       "      <td>292</td>\n",
       "      <td>19</td>\n",
       "      <td>109</td>\n",
       "      <td>131</td>\n",
       "      <td>136</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>c-ddd</th>\n",
       "      <td>5</td>\n",
       "      <td>145</td>\n",
       "      <td>307</td>\n",
       "      <td>73</td>\n",
       "      <td>53</td>\n",
       "      <td>65</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>c-sddd</th>\n",
       "      <td>1</td>\n",
       "      <td>78</td>\n",
       "      <td>220</td>\n",
       "      <td>56</td>\n",
       "      <td>93</td>\n",
       "      <td>72</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>i1</th>\n",
       "      <td>2</td>\n",
       "      <td>19</td>\n",
       "      <td>0</td>\n",
       "      <td>4683</td>\n",
       "      <td>284</td>\n",
       "      <td>111</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>c-si1</th>\n",
       "      <td>0</td>\n",
       "      <td>72</td>\n",
       "      <td>0</td>\n",
       "      <td>3468</td>\n",
       "      <td>282</td>\n",
       "      <td>107</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>i2</th>\n",
       "      <td>3</td>\n",
       "      <td>4</td>\n",
       "      <td>4</td>\n",
       "      <td>4</td>\n",
       "      <td>37279</td>\n",
       "      <td>681</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>c-si2</th>\n",
       "      <td>0</td>\n",
       "      <td>2</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>21686</td>\n",
       "      <td>450</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>i3</th>\n",
       "      <td>0</td>\n",
       "      <td>5</td>\n",
       "      <td>1</td>\n",
       "      <td>14</td>\n",
       "      <td>15</td>\n",
       "      <td>64187</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>c-si3</th>\n",
       "      <td>0</td>\n",
       "      <td>4</td>\n",
       "      <td>0</td>\n",
       "      <td>4</td>\n",
       "      <td>3</td>\n",
       "      <td>31000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>*</th>\n",
       "      <td>241</td>\n",
       "      <td>362</td>\n",
       "      <td>289</td>\n",
       "      <td>1342</td>\n",
       "      <td>1148</td>\n",
       "      <td>981</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>s</th>\n",
       "      <td>1968</td>\n",
       "      <td>2002</td>\n",
       "      <td>1994</td>\n",
       "      <td>2740</td>\n",
       "      <td>2322</td>\n",
       "      <td>2236</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>c-ss</th>\n",
       "      <td>233</td>\n",
       "      <td>274</td>\n",
       "      <td>266</td>\n",
       "      <td>1095</td>\n",
       "      <td>1129</td>\n",
       "      <td>873</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>f</th>\n",
       "      <td>969</td>\n",
       "      <td>2647</td>\n",
       "      <td>2056</td>\n",
       "      <td>5374</td>\n",
       "      <td>6587</td>\n",
       "      <td>7687</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>other</th>\n",
       "      <td>3430</td>\n",
       "      <td>4133</td>\n",
       "      <td>3603</td>\n",
       "      <td>11234</td>\n",
       "      <td>14538</td>\n",
       "      <td>11957</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "        S6.d3  S6.d6  S6.d9  S6.i3  S6.i6  S6.i9\n",
       "d         302    146    188    143    151    183\n",
       "c-sd      267     73     77    148    157    173\n",
       "c-dd       19    322     42     82    135    128\n",
       "c-sdd      11    292     19    109    131    136\n",
       "c-ddd       5    145    307     73     53     65\n",
       "c-sddd      1     78    220     56     93     72\n",
       "i1          2     19      0   4683    284    111\n",
       "c-si1       0     72      0   3468    282    107\n",
       "i2          3      4      4      4  37279    681\n",
       "c-si2       0      2      1      0  21686    450\n",
       "i3          0      5      1     14     15  64187\n",
       "c-si3       0      4      0      4      3  31000\n",
       "*         241    362    289   1342   1148    981\n",
       "s        1968   2002   1994   2740   2322   2236\n",
       "c-ss      233    274    266   1095   1129    873\n",
       "f         969   2647   2056   5374   6587   7687\n",
       "other    3430   4133   3603  11234  14538  11957"
      ]
     },
     "execution_count": 70,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "all_protein_count, all_protein_reads = get_full_protein_composition(all_ref, 1)\n",
    "\n",
    "row_index_protein = [ 'd', 'c-sd', 'c-dd', 'c-sdd', 'c-ddd','c-sddd', 'i1','c-si1', 'i2','c-si2', 'i3','c-si3','*','s', 'c-ss','f', 'other']\n",
    "df_protein_reads = pd.DataFrame(all_protein_reads, index=row_index_protein, columns=fractions)\n",
    "df_protein_count = pd.DataFrame(all_protein_count, index=row_index_protein, columns=fractions)\n",
    "\n",
    "df_protein_count"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 76,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>S6.d3</th>\n",
       "      <th>S6.d6</th>\n",
       "      <th>S6.d9</th>\n",
       "      <th>S6.i3</th>\n",
       "      <th>S6.i6</th>\n",
       "      <th>S6.i9</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>d</th>\n",
       "      <td>302</td>\n",
       "      <td>142</td>\n",
       "      <td>185</td>\n",
       "      <td>136</td>\n",
       "      <td>136</td>\n",
       "      <td>174</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>c-sd</th>\n",
       "      <td>223</td>\n",
       "      <td>65</td>\n",
       "      <td>69</td>\n",
       "      <td>59</td>\n",
       "      <td>58</td>\n",
       "      <td>67</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>c-dd</th>\n",
       "      <td>19</td>\n",
       "      <td>320</td>\n",
       "      <td>39</td>\n",
       "      <td>73</td>\n",
       "      <td>121</td>\n",
       "      <td>110</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>c-sdd</th>\n",
       "      <td>9</td>\n",
       "      <td>234</td>\n",
       "      <td>13</td>\n",
       "      <td>44</td>\n",
       "      <td>44</td>\n",
       "      <td>50</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>c-ddd</th>\n",
       "      <td>4</td>\n",
       "      <td>144</td>\n",
       "      <td>307</td>\n",
       "      <td>61</td>\n",
       "      <td>42</td>\n",
       "      <td>51</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>c-sddd</th>\n",
       "      <td>1</td>\n",
       "      <td>74</td>\n",
       "      <td>180</td>\n",
       "      <td>16</td>\n",
       "      <td>21</td>\n",
       "      <td>21</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>i1</th>\n",
       "      <td>2</td>\n",
       "      <td>5</td>\n",
       "      <td>0</td>\n",
       "      <td>4671</td>\n",
       "      <td>200</td>\n",
       "      <td>89</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>c-si1</th>\n",
       "      <td>0</td>\n",
       "      <td>3</td>\n",
       "      <td>0</td>\n",
       "      <td>3359</td>\n",
       "      <td>79</td>\n",
       "      <td>39</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>i2</th>\n",
       "      <td>2</td>\n",
       "      <td>4</td>\n",
       "      <td>4</td>\n",
       "      <td>4</td>\n",
       "      <td>37116</td>\n",
       "      <td>609</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>c-si2</th>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>21402</td>\n",
       "      <td>296</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>i3</th>\n",
       "      <td>0</td>\n",
       "      <td>5</td>\n",
       "      <td>1</td>\n",
       "      <td>14</td>\n",
       "      <td>15</td>\n",
       "      <td>63663</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>c-si3</th>\n",
       "      <td>0</td>\n",
       "      <td>3</td>\n",
       "      <td>0</td>\n",
       "      <td>4</td>\n",
       "      <td>3</td>\n",
       "      <td>30601</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>*</th>\n",
       "      <td>5</td>\n",
       "      <td>8</td>\n",
       "      <td>5</td>\n",
       "      <td>370</td>\n",
       "      <td>40</td>\n",
       "      <td>39</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "        S6.d3  S6.d6  S6.d9  S6.i3  S6.i6  S6.i9\n",
       "d         302    142    185    136    136    174\n",
       "c-sd      223     65     69     59     58     67\n",
       "c-dd       19    320     39     73    121    110\n",
       "c-sdd       9    234     13     44     44     50\n",
       "c-ddd       4    144    307     61     42     51\n",
       "c-sddd      1     74    180     16     21     21\n",
       "i1          2      5      0   4671    200     89\n",
       "c-si1       0      3      0   3359     79     39\n",
       "i2          2      4      4      4  37116    609\n",
       "c-si2       0      1      1      0  21402    296\n",
       "i3          0      5      1     14     15  63663\n",
       "c-si3       0      3      0      4      3  30601\n",
       "*           5      8      5    370     40     39"
      ]
     },
     "execution_count": 76,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "row_index_protein_filt = [ 'd', 'c-sd', 'c-dd', 'c-sdd', 'c-ddd','c-sddd', 'i1','c-si1', 'i2','c-si2', 'i3','c-si3','*']\n",
    "all_protein_count_f, all_protein_reads_f = get_filetered_protein_composition(all_ref, 1)\n",
    "df_protein_reads_filt = pd.DataFrame(all_protein_count_f, index=row_index_protein_filt, columns=fractions)\n",
    "\n",
    "df_protein_reads_filt"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 122,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "d3\n",
      "Detected mutations not in baseline:\n",
      "((334, 'd'),) (('999', 'd', '3'),) 59\n",
      "((334, 's', 'C'), (335, 'd')) (('1001', 'd', '3'),) 28\n",
      "\n",
      "\n",
      "d6\n",
      "Detected mutations not in baseline:\n",
      "((333, 's', 'C'), (334, 'd'), (335, 'd')) (('998', 'd', '6'),) 3\n",
      "((334, 'd'), (335, 'd')) (('1000', 'd', '6'),) 2\n",
      "((333, 'd'), (334, 'd')) (('997', 'd', '6'),) 200\n",
      "((333, 'd'), (334, 'd')) (('996', 'd', '6'),) 5\n",
      "((332, 's', 'G'), (333, 'd'), (334, 'd')) (('995', 'd', '6'),) 4\n",
      "\n",
      "\n",
      "d9\n",
      "Detected mutations not in baseline:\n",
      "((332, 'd'), (333, 'd'), (334, 'd')) (('994', 'd', '9'),) 214\n",
      "((332, 'd'), (333, 'd'), (334, 'd')) (('992', 'd', '9'),) 8\n",
      "((333, 'd'), (334, 'd'), (335, 'd')) (('996', 'd', '9'),) 1\n",
      "((333, 'd'), (334, 'd'), (335, 'd')) (('997', 'd', '9'),) 2\n",
      "\n",
      "\n"
     ]
    }
   ],
   "source": [
    "fracs = ['d3', 'd6', 'd9']\n",
    "\n",
    "for fraction in fracs:\n",
    "    print(fraction)\n",
    "    print('Detected mutations not in baseline:')\n",
    "    for prot_mutation, entry in all_ref['S6'][fraction].items():\n",
    "        if classify_protein(prot_mutation) in dna_protein_correspondence(fraction):\n",
    "            for dna_mutation, count in entry['dna'].items():\n",
    "                if classify_dna(dna_mutation) == fraction:\n",
    "                    try:\n",
    "                        base['S6']['d_baseline'][prot_mutation]['dna'][dna_mutation]\n",
    "                    except KeyError:\n",
    "                        print(prot_mutation, dna_mutation, count)\n",
    "    print()\n",
    "    print()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Figure 3A (after normalization for baseline from codon ambiguity).\n",
    "\n",
    "Generate data for a histogram of tranposon insertion sites: best used for -3 bp library, it shows how many times a mutation is detected at each DNA position. Spikes corresponds to sites close to tranposon preferred insertion sequence (GC rich). This data still needs to be normalized for whether a position can be detected due to codon ambiguity.\n",
    "\n",
    "InDel placement is inheretnly ambiguous. For example, if the original sequence is GCT TCA and we observe GC- --A, it could just as well come from G-- -CA. Any algorithm will arbitrarily favour one over another. Here, if the true deletions were one of each, the use of Needleman-Wunsch algorith to place deletions will favour placing InDels earlier in the sequence. Hence, G-- -CA will be reported twice and GC- --A zero times. When we are considering the transposon preference, the best we can do is to split the observed count across all positions where it could have come from, equally.\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 123,
   "metadata": {},
   "outputs": [],
   "source": [
    "def find_transposon_histogram(all_references, background, baseline='baseline', transposon='d3'):\n",
    "    \"\"\"\n",
    "    Find all mutation of a certain type and count where they are in DNA\n",
    "    :return: dict\n",
    "    \"\"\"\n",
    "    hist = defaultdict(int)\n",
    "\n",
    "    for prot_mutation in all_references[background][baseline]:\n",
    "        if prot_mutation is None:\n",
    "            continue\n",
    "        elif len(prot_mutation) <= 2:\n",
    "            for dna_mutation, count in all_references[background][baseline][prot_mutation]['dna'].items():\n",
    "                if classify_dna(dna_mutation) == transposon:\n",
    "                    hist[int(dna_mutation[0][0])] += count\n",
    "\n",
    "    return hist\n",
    "\n",
    "histogram = find_transposon_histogram(all_ref, 'S6', baseline='d3', transposon='d3')\n",
    "\n",
    "df_hist = pd.DataFrame(histogram, index=['raw_insertions'], columns=list(range(1000)))\n",
    "df_hist = df_hist.fillna(0).transpose()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Figure 3B: Number of observed deletions per position"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 64,
   "metadata": {},
   "outputs": [],
   "source": [
    "def insertion_frequencies(all_references):\n",
    "    \"\"\"\n",
    "    Generate data for number of different insertions per position.\n",
    "    :param all_references:\n",
    "    :return:\n",
    "    \"\"\"\n",
    "    ins_freq = {}  # for each library ('d3', 'i6', etc.) give a dictionary {1: 5 times, 2: 3 times...}\n",
    "    for background in all_references.keys():\n",
    "        for fraction in all_references[background].keys():\n",
    "            ins_freq['.'.join([background,fraction])] = defaultdict(int)\n",
    "            for prot_mutation in all_references[background][fraction].keys():  # this will go through all protein mutations\n",
    "                for dna_mutation, c in all_references[background][fraction][prot_mutation]['dna'].items():\n",
    "                    # sometimes a mutations will have a sequencing error / silent substitution elsewhere, which biases the count\n",
    "                    dna_mut_type = classify_dna(dna_mutation)\n",
    "                    prot_mut_type = classify_protein(prot_mutation)\n",
    "                    if (dna_mut_type == fraction) and (prot_mut_type in dna_protein_correspondence(dna_mut_type)):\n",
    "                        ins_freq['.'.join([background,fraction])][int(dna_mutation[0][0])] += 1\n",
    "    return ins_freq"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "ins_freq = insertion_frequencies(all_ref)\n",
    "\n",
    "df_ins_freq = pd.DataFrame(ins_freq, columns=fractions).fillna(0).astype('int32')\n",
    "df_ins_freq['S6.deletions'] = df_ins_freq['S6.d3'] + df_ins_freq['S6.d6'] + df_ins_freq['S6.d9']\n",
    "df_ins_freq.to_csv('Fig3BC_all_ref.csv')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 65,
   "metadata": {},
   "outputs": [],
   "source": [
    "base_ins_freq = insertion_frequencies(base)\n",
    "\n",
    "df_base_ins_freq = pd.DataFrame(base_ins_freq, columns=fractions).fillna(0).astype('int32')\n",
    "df_base_ins_freq['S6.deletions'] = df_base_ins_freq['S6.d3'] + df_base_ins_freq['S6.d6'] + df_base_ins_freq['S6.d9']\n",
    "df_base_ins_freq.to_csv('Fig3BC_base.csv')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 148,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "array([[<matplotlib.axes._subplots.AxesSubplot object at 0x7f83335ed198>,\n",
       "        <matplotlib.axes._subplots.AxesSubplot object at 0x7f83341212e8>,\n",
       "        <matplotlib.axes._subplots.AxesSubplot object at 0x7f833418ba20>],\n",
       "       [<matplotlib.axes._subplots.AxesSubplot object at 0x7f83342502e8>,\n",
       "        <matplotlib.axes._subplots.AxesSubplot object at 0x7f83342cc978>,\n",
       "        <matplotlib.axes._subplots.AxesSubplot object at 0x7f83341e19b0>],\n",
       "       [<matplotlib.axes._subplots.AxesSubplot object at 0x7f83341e8518>,\n",
       "        <matplotlib.axes._subplots.AxesSubplot object at 0x7f833438dc50>,\n",
       "        <matplotlib.axes._subplots.AxesSubplot object at 0x7f833438d2e8>]],\n",
       "      dtype=object)"
      ]
     },
     "execution_count": 148,
     "metadata": {},
     "output_type": "execute_result"
    },
    {
     "data": {
      "image/png": "iVBORw0KGgoAAAANSUhEUgAAAXoAAAEICAYAAABRSj9aAAAABHNCSVQICAgIfAhkiAAAAAlwSFlzAAALEgAACxIB0t1+/AAAADh0RVh0U29mdHdhcmUAbWF0cGxvdGxpYiB2ZXJzaW9uMy4xLjAsIGh0dHA6Ly9tYXRwbG90bGliLm9yZy+17YcXAAAgAElEQVR4nO3de5xdZX3v8c/XIOoJlyQGQoTUEYm1UaqGKDknWEetEVCJtWqxHAmKpVo9yiE9NWB7pNoLWEHBG2JRgyeAVMiLFPASkam3giSIBAiBiBFCIhECgYSqXH7nj+eZYWWyZ2bPzF57r1n7+3699mvvddlr/dY8a377Wc+6PIoIzMysvp7W6QDMzKxcTvRmZjXnRG9mVnNO9GZmNedEb2ZWc070ZmY150RfEkl9kt7T6TisPC7j+qtLGXd9opd0hKQfS9ouaZukH0l6eZ62n6SLJD0k6UFJy8e4jldLWpuX84CkFZIObO2W2FDaUcatXpaNTpv+jyXpI5LulvSwpEsk7dPaLSlHVyf6XEhXAp8BpgEHAn8P/DbPcjnwK+C5wP7AJ8e4qtuA10fEFOA5wJ3AF8YeuTWrjWXc6mVZk9pYxscD7wQWkP6Pn5XXWXl7dDqADnsBQERcnIf/C/gOgKSFwCygNyKeyNN/OtSCJL2OVOgzga8B6p8WEfcNmv0J4JAWxG8ja0sZj3ZZ1lJtKWPgTcAFEXFPnvdM4HuS3hcRj7Zuc1qvq2v0wB3AE5KWSTpK0tTCtPnAemBZbm65QdKrGi1E0nTgMuBvgenAz0m/+sV5fk/SQ6Sd8K+BT7R+c6yBdpVx08uylmtXGYtdE7+AZwCzW7cp5ejqRB8RDwNHAAF8Cfi1pJWSZgAHAQuBa4EDgLOAK/LOMNjRwG0R8Y2IeAz4NOlQsbiuu3PTzXTSjnR7SZtlBW0s49Esy1qojWX8TeA9knok7Qt8OI//b2VsVyt1daIHiIh1EXFCRBwEvJjU9vZpUs17Y0RcEBGPRcQlwD0Mqqlnz8nT+pcZxeFB69sGLCPtbN3edNYWbSrj0SzLWqxNZfxl4GKgD7iV9OMBsKnFm9NyXZ/oiyLiduCrpB3lZlINoRlbSO2AQDo7XxxuYA/SSaEJcca+Tkos49Esy0pUVhlHxJMR8dGI6Mk/KLcC9+ZXpXV1opf0QklLJB2Uh2cB7wCuA1YAUyUtljRJ0ltJZ/N/1GBRVwEvkvSWXEv/IOkwsX89b5H0+5KeJmk/4Gzgp7l2byVqVxmPclnWQm38P54m6fn5Mss5pP/jj0XEk+VuYQtERNe+SAV+KekXeWd+/yKwT57+SmAtsANYDbyy8N3zgPMKw0eSTgptBz4L/AfwnjztfwG/yOv4FXAJ8NxOb383vNpVxiMty6+JX8akq3vWA48CvwRO6fS2N/tS3gAzM6uprm66MTPrBk70ZmY150RvZlZzTvRmZjVXiRt2pk+fHj09PQPDO3fuZPLkyZ0LaBQmSqyD41yzZs39EbFfu9bvMi6fy3jsJkqsYy7jTl/2ExEcdthhUXTttdfGRDFRYh0cJ7A6XMZNmSixuozHbqLEOtYydtONmVnNVaLpxsavZ+lVw07/6pHVPyy1oY1UvuAynujKLGPX6M3Mas6J3sys5pzozcxqzonezKzmnOjNzGrOid7MrOac6M26hKSNktZKuknS6jxumqRVku7M71PzeEk6V9IGSTdLmtvZ6G08nOjNusurI+KlETEvDy8FromI2cA1eRjgKGB2fp0EfKHtkVrLNJXoXRMwq61FpM7qye9vLoy/MN9pfx0wRdLMTgRo4zeaO2NfHRH3F4b7awJnSFqahz/MrjWBw0k1gcNbFK+ZjV0A35EUwBcj4nxgRkRsAYiILZL2z/MeCNxT+O6mPG5LcYGSTiLV+JkxYwZ9fX0D03bs2LHLcJVVIdYlhz4+4jxjjXM8j0BYBPTmz8uAPlKiH6gJANdJmiJpZv/OZGYdsyAiNudkvkrS7cPMqwbjdut3NP9YnA8wb9686O3tHZjW19dHcbjKqhDrCU0+AmEscTab6F0TGEJVYh2pNjBSnJI2Ao8ATwCPR8Q8SdOArwM9wEbg7RHxoCQB5wBHkzpKPiEibhz3RlipImJzft8qaQXwCuC+/opYbprZmmffBMwqfP0gYHNbA7aWaTbRuyYwhKrEOlJtoMmagJvnakrSZOBpEfFI/rwQ+BiwElgMnJHfr8hfWQl8QNIlpLLd7qPyiaupRO+aQNdy81x9zABWpIMx9gAuiohvSboBuFTSicDdwNvy/FeTjtg2kI7a3tX+kK1VRkz0rgl0jZY3z1l1RMRdwEsajH8AeG2D8QG8vw2hWRs0U6N3TaA7tLx5zudhWqfMKzKs/kZM9K4JdIcymud8HqZ1yrwiw+rPd8YakiZL2rv/M6l57haeap6D3Zvnjs83x83HzXNmleauBA3cPGdWa0705uY5K8Xae7eP2OS08Yw3tCma7uamGzOzmnOiNzOrOSd6M7Oaq2Qbvdv2zMxaxzV6M7Oac6I3M6s5J3ozs5pzojczqzknejOzmnOiNzOruUpeXmn150to20vSLOBC4ADgSeD8iDhH0unAXwC/zrOeFhFX5++cCpxI6l7ygxHx7bYHbi0xYo1e0ixJ10paJ+lWSR/K40+XdK+km/Lr6MJ3TpW0QdJ6Sa8vcwPMrCmPA0si4g+A+cD7Jc3J0z4VES/Nr/4kPwc4FngRcCTweUmTOhG4jV8zNfr+HeTG/CjbNZJW5WmfiohPFmcetIM8B/iupBdExBOtDNzMmpcfI93fW9gjktaRegUbyiLgkoj4LfALSRtIfRT8Z+nBWss10/GId5Ca82F9d5HUA7wMuB5YQOr683hgNalS9yDpf/y6wtf6u4scvKwhexGb8ayRe8aqSo9YVeidq8xexEbVRu8dZHdV2EFg5L/XCHH6qK1LSNoLuAw4OSIelvQF4OOkriA/DpwFvJsmu4scrhexzyy/grPWDp9iNh7XO+z0dql7L2JNJ3rvII1VYQeBkXeS4XYQH7V1B0lPJ/0PL4+IywEi4r7C9C8BV+bBprqLtImhqUTvHaR7+Khtd1U4ahvvYb1S92EXAOsi4uzC+JmFbiD/hNSFJKTuIi+SdDbpqG028JOxxm+dNWKi9w7SPXzU1lgVjtpacFi/AHgnsFbSTXncacA7JL2UVH4bgb8EiIhbJV0K3EZq2nu/m+YmrmZq9N5BuoCP2uotIn5I4x/oq4f5zj8C/1haUNY2zVx14x2k5nzUZlZvvjPWwEdtZrXmRG8+ajOrOT/UzMys5pzozcxqzonezKzmnOjNzGrOid7MrOac6M3Mas6J3sys5pzozcxqzonezKzmnOjNzGrOid7MrOac6M3Maq60RC/pSEnrJW2QtLSs9VhnuHzrz2VcH6UkekmTgM8BRwFzSI+7nVPGuqz9XL715zKul7Jq9K8ANkTEXRHxO+ASUofSVg8u3/pzGddIWc+jPxC4pzC8CTi8OEOx42hgh6T1hcnTgfuHW4HObEGUrTFirFXw6jN3i/O541jciOULLuN2cxmPS63LuKxEP2Ln0cWOo3f7srQ6IuaVEVirTZRYWxznqDsHLzmeUk2UWF3GYzdRYh1rnGU13bjz6Hpz+dafy7hGykr0NwCzJT1P0p7AsaQOpStLUp+k9zQ5b0g6ZIzrOU3Sv47luxUy4cq3lSTtkHRwp+Momcu4RmVcSqKPiMeBDwDfBtYBl0bEraNYRMNDQQBJR0j6saTtkrZJ+pGkl+dp+0m6SNJDkh6UtHxcGzL+WHslbSqOi4h/ioimflBabMg4R6sF5dvSeJo1jn1nl1gjYq+IuCt/72RJd0l6WNJmSZ+S1Km+mF3GJZRx/u5cSd/PPwD3SfpQmzZpsLH9TSNiwryAfYCHgHcAk4BnAQuBP8zTfwCcDewLPB142SiW3Qe8p8l5Azikifl6gU2d/rv5Vd6+AzwfmJI/TwO+B5zS6e3txleJZTwd2AocBzwD2Bv4g05v76j+Np0OYJQFOQ94aIhpC4GNwKQml/U64HZgO/BZ4D+KiR54N6km8yCpVvPcwrSBRJ8L/pPA3cB9wHl5B5sM/BfwJLAjv54DnA78v8KyjgFuzTtoX3EHytvz18DNOc6vA88s7HxX5u9tyzvx0zpdRlV9tXjfafhDDzwb+C7w+U5vbze+yipj4J+Ar3V6+8bzmmiPQLgDeELSMklHSZpamDYfWA8sk/SApBskvarRQiRNBy4D/paUMH8OLChMfzNwGvAWYD9SEr14iJjOBF4AvBQ4hHRZ2v+NiJ2km002RzoM3CsidjmZJekFebkn5/VcDfx7bhPt93bgSOB5wB8CJ+TxS0gnzPYDZuR4d7sqwga0ZN9pRNKfS3qYdNnbS4AvtjRya1ZZZTwf2JabhLZK+ndJv9fq4MvU0UQ/0i3Wkp4h6et5+vWkQ+MjSAntS8CvJa2UNIN0VcBC4FrgAOAs4Iqc1Ac7GrgtIr4REY8BnwZ+VZj+l8A/R8S6SG2Va4D5kjYW45Qk4C9IPwTr8/vvA822wf8ZcFVErMpxfJJ0NPA/CvOcGxGbI2Ib8O+kHxSAx4CZpCONxyLiB8AFeUe8pdHKlJyb/543S5rbZJxjNtoyltRTRhwR8TAj7zt/TDqCu5NB+46kEyT9WtJNedTbCsu+KCL2If3gn0c6siuFpC+7jBsrsYwPAhYDHwJ+D/gFQ1f8xq2UMu7gYdYkUk36YGBP4GfAnEHz/BVwXv58LPD1QdNfCKwm/dHPAX4xaPpaYFGDdS8F/m3QuP8kN90At5GaWh7KryeA3wB/lOMMUu19//z5UeC3ed7twI68nF4GtdFTaLoBvgD8y6Dp1wHH5c8bgT8e4rt7k37M7sqvpTm+ucAtQ/zNjwa+SbpGej5wfdXLuMTYivvOuaQfzmKcdxT3HdKR1Gfz5yHP0eRtuLzEuF3GbS7jPO9XCvM9O0/fd6KUcSdr9M3cYr0IWJY/fwN4ba5FAxARtwNfBV5MasdutuliC4VrhPMyi9cM3wP8ZURMITW/fDcinhkR389x9ruf1A5/GvCliJgSEftGxF79IY4Qx2YKd7YV4rh3pA2IiEciYklEHAy8CTiFdIJp2zBfWwRcGMl1wBRJM0da1ziMu4zLMmjfeQR4bFCc+4xx0XuQTtCWIu+DLuMmtLCMB+eW/s+lbEMZZdzJRN/oFusDh5onUhPKo8DfSToIQNIs0hn264AVwFRJiyVNkvTW/P0fNVj3VcCLJL1F6VK4D5Kae/qdB5wq6UV5Gb+S1H8YN3C5ZEQ8STpE/HPgbfkw6kpJx+VZ7gOeLWnfIf4GlwJvkPRaSU8ntbv/FvjxEPMPkPRGSYfkf5iHSUcdT4zwtWb+5q00ljLeTqoxtZSkF0paMsS+cycwqX/fITWJTWH3fedPJd2cPx+Ql/MeSfvnz3OAU4FrWh3/KLiMW1zGwFeAP5H00vx/+nfADyPioVZvQ5NGXcadTPTN3GI9eJ4nSYc010vaSSrAW4AlkdqwjyFdpbKd1JSxKCLuB5B0nqTzAPK4twFnAA8AsykUeESsIJ1kvQS4EPhTUs2+kQ/n7+4EeoDDgI/m5dxOOmy8S+na3efssrER64H/CXyGdHTwJuBNucYxktmkKzx2kJqdPh8RfSN8p6nb2ltoLGXcaJ5WeIT0rJbd9h1S2X2Hp/adNwJXFvcd4FVAT0T8YV7ev+T3BcDavMyr8+u0EuJvlsu4xWUcEd8jlelVpMssDyFV7jpl1H/PTt3YAc3dYt0/z6Zc894L+JPIDVWDRToheegQ0947aPhbpJNnDUXE14CvSfrvwOkR8e5CnKdFxIY832+A/51f/Y933VZYzrvZ1emD1rOCdDTSKIaeQcOnFz5/CvjUUPEPod23tY+ljPdl+MPWMYmIe0lXMO1G6aa2PSPi0Dx86qDvvnfQV/bojzEi3tXqWMfJZdzAeMo4T/8C6ZxaFYy6jDtZo2/mFuuVpLPdAG8FvjdUki/RiHEOah87hnT9fRWtBI7PZ+3nA9sjYkuJ63MZt5/LuLHuLuMyzho3+yKdPb6DdNb+I3ncx4Bj8udnAv8GbAB+Ahxc0Tj/mXTT089Il3e+sENxXkw60fwY6Vf/ROC9wHvzdJE6k/g56YqkeRX427mMXcYu45LLWPmLZmZWUxPtzlgzMxulTp6MHTB9+vTo6ekZGN65cyeTJ0/uXEBt0OltXLNmzf0RsR8MXIZ2IelysieB8yPiHEnTSM/X6SHdvPX2iHgwX9J5DulQ+FHghIi4cbj1Vb2MqxRPq2IplnE7VK2Mu2H9TZdxJ9qgBr8OO+ywKLr22muj7jq9jcDqeKpNcCYwN3/em9SOOQf4BLA0j18KnBlPtXWO6s68qpdxleJpVSzFMm7Hq2pl3A3rb7aM3XRjRMSWyDXyiHiEdLXBgex6R+My4M35c7vvvjSzcahE002d9Sy9quH4JYc+zgl52sYz3tDOkIal9MCplwHXAzMiX7YVEVv67wBl6DvzdrnES4WOo2fMmEFfX9/AtK3btvOZ5VcMG8uhBw51Q3Hr7dixY5f4OqlKsYzH2nu3D+zjQ6nSvl9nTvQ2QNJepMc3nxwRDw/zOJJRdxw9b9686O3tHZj2meVXcNba4Xe/jcf1Dju9lfr6+ijG10lVisXqwU03BkB+hsdlwPKIuDyPvq+/SSa/b83j3XG02QTiRG/9T828AFgXEWcXJhXvaFwMXFEY3867L81sHJpO9PmJkD+VdGUefp5SJwJ3KnUqsGce35ZOBqylFgDvBF4j6ab8Opr00LfXSbqT1PXiGXn+q0nPwN9AenrnX3UgZjNr0mja6D9Euhqj/xnOZwKfiohL8pPfTiQ99OdE4MGIOETSsXm+P2thzNZiEfFDhn629msbzB/A+0sNysxapqkafX6+8xuAf83DAl5D6kQAdr/0ru2dDJiZWWPN1ug/DfwN6WYaSJ0GPBSpEwHY9cH3u3QyIKm/k4H7iwsc7tK7ulxeBukyykZmPOupaXXZVjOrphETvaQ3AlsjYo2k3v7RDWYdrnutUV16V6fLy4a6jnjJoY8PXF7YzssIzaz7NFOjXwAck0/OPZPURv9p0t2Qe+RaffHyurZ0MmBmZs0ZsY0+Ik6NiIMi9XZ0LKnTgONIz2t+a55t8KV3ne5kwMzMsvFcR/9h4BRJG0ht8Bfk8ReQOsTeAJxCehiWmZl1yKgegRCp8+m+/Pku4BUN5vkNqeNtMzOrAN8Za2ZWc070ZmY150Rv1iUkbZS0Nj/iYnUeN03Sqvwok1WSpubxknRufpTJzZLmdjZ6Gw8/ptgqa6hn+ffzs8zH5NURUbx5cSlwTUScIWlpHv4wcBQwO78OJz3e5PB2B2ut4Rq9WXdzL2JdwInerHsE8B1Ja/IjSGBQL2LASL2I2QTkphuz7rEgIjbnLiFXSbp9mHmbepTJcM+sKj7PaShlPuep08/M6vT6i5zozbpERGzO71slrSDdB3OfpJm5T+BR9yJW5e4iO/3MrE6vv8hNN2ZdQNJkSXv3fwYWArfgXsS6gmv0Zt1hBrAidw2xB3BRRHxL0g3ApZJOBO7mqbvarwaOJvUi9ijwrvaHbK3iRG/WBfIjS17SYPwDuBex2nPTjZlZzblGXxO+ucjMhuIavSHpy5K2SrqlMM63xpvVRFfX6F0LHvBV4LPAhYVxvjW+QkbaV6Gr9lcbpa5O9JZExPcl9QwavQjozZ+Xkfoh+DCFW+OB6yRN6b8Ouz3RPsXJz6w5zXQOPotU0zsAeBI4PyLOkTQN+DrQA2wE3h4RDypdv3UO6dKsR4ETIuLGcsK3Eu1ya3y+mxKGvjV+t0Q/3rsmW6HZOxOrdBdjo1ia+VtVJX6rnmZq9I8DSyLixnzDxRpJq4AT8KF9N2rq1ngY/12TrdDsnZdVuouxUSwnNHP0UuJdpjaxNdM5+Jb+GnlEPAKsI9Xg/NS7eruvv9zGcmu8mVXHqKpUuR33ZcD1jPPQfrjD+nYdRrfjgUtDraPYdFHmevqNYR39t8afwe63xn9A0iWkIzXfGm9WcU0nekl7AZcBJ0fEw/lW6oazNhi326H9cIf17TqMHulwuBWHwkOtY8mhjw80XZS5nn7DrUPSxaQTr9MlbQI+SkrwvjW+CT4pbFXXVKKX9HRSkl8eEZfn0eN66p1VR0S8Y4hJvjW+TYo/FksOfbypNnmzZjVz1Y2AC4B1EXF2YZIP7a3yfK+EWXM1+gXAO4G1km7K407Dh/ZmZhPCiIk+In5I43Z38KG9dYFm2uDbsQyzsfKzbszMas6PQLCu1l/T9glQqzPX6M3Mas6J3sys5pzozcxqzonezKzmnOjNzGrOid7MrOZ8eaWZdYwfUdEertGbmdWcE72ZWc050ZuZ1ZwTvZlZzVXyZOzae7eP3GOST9KYmTWlkonezAzcTWOruOnGzKzmSkv0ko6UtF7SBklLy1qPdYbLt/5cxvVRSqKXNAn4HHAUMAd4h6Q5ZazL2s/lW38u43opq43+FcCGiLgLIHcUvgi4raT1WXu5fOtvwpTxUO34o+lMpu7t/GUl+gOBewrDm4DDizNIOgk4KQ/ukLS+MHk6cP9wK9CZLYhyBGWu44OFbezQtjx3HIsbsXxh/GXcTh+sUDxjjcVlvKvR/B1L+h9sx/Y3VcZlJfpGnYnHLgMR5wPnN/yytDoi5pURWFVM8G0csXxhYpVxleKpSCwTvoy7ff1FZZ2M3QTMKgwfBGwuaV3Wfi7f+nMZ10hZif4GYLak50naEzgWWFnSuipH0g5JB3c6jhJ1dfl2CZdxjZSS6CPiceADwLeBdcClEXHrKBbR8FCwbJKOkPRjSdslbZP0I0kvz9P2k3SRpIckPShp+VDLiYi9CiexpkhaJmlrfp2eZ+vINrZCC8oXqrf9VYqn47HUpIy7ff0DFLFbs1tXkrQPcDfwPuBSYE/glcCvIuJmST8g1XL+HngUeHFE/LSJ5X4F2AtYDOwPXAP8Q0R8pZQNMTMbxIk+kzQP+G5ETGkwbSHp1/n5EfFEE8sKYHZEbJB0P3BURNyQp52Wh1/Z2i0wM2vMj0B4yh3AE7mZ5ShJUwvT5gPrgWWSHpB0g6RXjWLZGvT5xS2I18ysKZVL9J267ToiHgaOIF1C9iXg15JWSppBuuJgIXAtcABwFnCFpOlNLPpbwFJJe0s6AvhbYF9Jt0r6UBnbUnWdvLVe0ixJ10paVywDSdMkrZJ0Z36fOtKyWhjTJEk/lXRlHn6epOtzLF/PJ0MnlHaVsaSNktZKuknS6jyuYVkqOTfHdLOkuWNY35fzubZbCuNGvT5Ji/P8d0paPP6/xAgiojIvYBLwc+BgUhv5z4A5HYrlhcBq4GLgHOAXg6avBRYN8d0ADsmfpwHLgV+Rjgr+NW/j3qSjiI5sX7eWMTATmJs/D5QB8AlgaR6/FDizjTGdAlwEXJmHLwWOzZ/PA97X6XKrahkDG4Hpg8Y1LEvgaOCbpKPq+cD1Y1jfHwFzgVvGur6cE+7K71Pz56lllknVavQDt11HxO+A/tuu2y4ibge+SmpmuZkGN4s0uZxtEXFcRBwQEb8PbAV+EhGPkK5mOLBFIU8UHS3jiNgSETfmz8UyWAQsy7MtA97cjngkHQS8gVQBQJKA1wDfaHcsLdTp/+OhynIRcGEk1wFTJM0czYIj4vvAtnGu7/XAqpwbHgRWAUeOJo7Rqlqib3TbdVsSoaQXSlqS//GQNAt4B3AdsAKYmg+3Jkl6a47rR00s9/mSnp2/dxTpdvF/kNQDvAy4vpwtqqyOlfFgg8pgRkRsgfRjQLpCqh0+DfwN8GQefjbwUKTLG6GDf59xaGcZB/AdSWuUHscAQ5dlWXGNdn1t/x+oWscjTd12XZJHSM/yOEXSFOAh4Erg/0TEw5KOAT5PeqLf7aRmm/SsGuk8gIh4b4PlHkb6Z55CaiY4Dvgl8B/AyZHODXSTTpbxU0FIewGXkcsgVaTbHsMbga0RsUZSb//oBrNOtEvj2rkNCyJis6T9gVWSbq9IXMOtr+1lXLVE37HbriPiXuDtw0z/AXDoENPeO2hYhc+XktpcAZD0dNIPyPKIuHycYU9EHb+1PpfBZexaBvdJmhkRW/Lh9dY2hLIAOEbS0cAzgX3IlQJJe+Ra/UR89EDbyjgiNuf3rZJWkJqNhirLsuIa7fo2Ab2Dxve1II4hVa3ppta3Xef21wuAdRFxdqfj6ZCOlvEwZbCSdFMb+f2KsmOJiFMj4qCI6CH9Hb4XEceRru56aztjabG2lLGkyZL27v9MujLuFoYuy5XA8flqmPnA9v4ml3Ea7fq+DSyUNDVfobMwjytPmWd6x/Iinam+g3TW/iOdjqfF29Z/+ebNwE35dXSn4+qmMh6qDEht49cAd+b3aW2Oq5enrro5GPgJsAH4N+AZnS6zKpZx/jv9LL9u7V/PUGVJajL5XI5pLTBvDOu8GNgCPEaqmZ84lvUB787luwF4V9nl4TtjzcxqrmpNN2Zm1mKVOBk7ffr06OnpGRjeuXMnkydP7lxAozBRYh0c55o1a+6PiP06GJKZtUklEn1PTw+rV68eGO7r66O3t7dzAY3CRIl1cJySftm5aMysndx0Y2ZWc5Wo0Q+29t7tI/beXvde283MWsU1ejOzmnOiNzOrOSd6M7Oac6I3M6s5J3ozs5pzojczqzknejOzmnOiNzOrOSd6M7Oac6I3M6s5J3ozs5pzojczqzknejOzmnOiNzOrOSd6M7Oac6I3M6s5J3ozs5pzojczqzknejOzmnOiNzOruaYSvaSNktZKuknS6jxumqRVku7M71PzeEk6V9IGSTdLmlvmBpiZ2fBGU6N/dUS8NCLm5eGlwDURMRu4Jg8DHAXMzq+TgC+0KlgzMxu98TTdLAKW5c/LgDcXxl8YyXXAFEkzx7EeMzMbh2YTfQDfkbRG0kl53IyI2AKQ3/fP4w8E7il8d1MeZ2ZmHbBHk/MtiIjNkvYHVkm6fZh51WBc7DZT+sE4CWDGjBn09fUNTJvxLFhy6OPDBlScv5N27NhRmViGM1HiNLPWayrRR8Tm/L5V0grgFcB9kmZGxJbcNLM1z74JmHo7VFcAAAQpSURBVFX4+kHA5gbLPB84H2DevHnR29s7MO0zy6/grLXDh7bxuN5hp7dLX18fxdiraqLEaWatN2LTjaTJkvbu/wwsBG4BVgKL82yLgSvy55XA8fnqm/nA9v4mHjMza79mavQzgBWS+ue/KCK+JekG4FJJJwJ3A2/L818NHA1sAB4F3tXyqM3MrGkjJvqIuAt4SYPxDwCvbTA+gPe3JDozMxs33xlrZlZzTvRmZjXnRG9mVnNO9GZmNedEb2ZWc070ZmY150RvZlZzTvRmZjXnRG9mVnNO9GZmNedEb2ZWc070ZmY150RvZlZzTvRmZjXnRG9mVnNO9GZmNedEb2ZWc070ZmY150RvZlZzTvRmZjXnRG9mVnN7dDqAsepZetWw0zee8YY2RWJmVm2u0ZuZ1ZwTvZlZzTnRm5nVXGmJXtKRktZL2iBpaVnrMTOz4ZVyMlbSJOBzwOuATcANklZGxG1lrK+RkU7Wgk/Ymll3KOuqm1cAGyLiLgBJlwCLgLYl+mb4yh0z6wZlJfoDgXsKw5uAw4szSDoJOCkP7pC0vjB5OnB/SbE1TWc2NVslYm3C4Dif26lAzKy9ykr0ajAudhmIOB84v+GXpdURMa+MwFptosQ6UeI0s9Yr62TsJmBWYfggYHNJ6zIzs2GUlehvAGZLep6kPYFjgZUlrcvMzIZRStNNRDwu6QPAt4FJwJcj4tZRLKJhk05FTZRYJ0qcZtZiioiR5zIzswnLd8aamdWcE72ZWc1VLtF3+tEJkr4saaukWwrjpklaJenO/D41j5ekc3OsN0uaW/jO4jz/nZIWlxDnLEnXSlon6VZJH6pqrGbWYRFRmRfpxO3PgYOBPYGfAXPaHMMfAXOBWwrjPgEszZ+XAmfmz0cD3yTdNzAfuD6Pnwbcld+n5s9TWxznTGBu/rw3cAcwp4qx+uWXX519Va1GP/DohIj4HdD/6IS2iYjvA9sGjV4ELMuflwFvLoy/MJLrgCmSZgKvB1ZFxLaIeBBYBRzZ4ji3RMSN+fMjwDrSHcmVi9XMOqtqib7RoxMO7FAsRTMiYgukBAvsn8cPFW9bt0NSD/Ay4Pqqx2pm7Ve1RD/ioxMqZqh427YdkvYCLgNOjoiHh5t1iJgm2t/czEapaom+qo9OuC83c5Dft+bxQ8Xblu2Q9HRSkl8eEZdXOVYz65yqJfqqPjphJdB/Ncpi4IrC+OPzFS3zge25ueTbwEJJU/NVLwvzuJaRJOACYF1EnF3lWM2ss8p6euWYxPgfnTBuki4GeoHpkjYBHwXOAC6VdCJwN/C2PPvVpKtZNgCPAu/K27FN0sdJP1wAH4uIwSd4x2sB8E5graSb8rjTKhqrmXWQH4FgZlZzVWu6MTOzFnOiNzOrOSd6M7Oac6I3M6s5J3ozs5pzojczqzknejOzmvv/VetWbVMl1dQAAAAASUVORK5CYII=\n",
      "text/plain": [
       "<Figure size 432x288 with 9 Axes>"
      ]
     },
     "metadata": {
      "needs_background": "light"
     },
     "output_type": "display_data"
    }
   ],
   "source": [
    "df_ins_freq.hist()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Figure S7 - not currently in the paper\n",
    "Reverse the data: how many mutations occer once, twice, three times etc."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 189,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "DNA: d6 (('2', 'd', '6'),) 1 PROTEIN: ()\n",
      "\n",
      "DNA: d6 (('27', 'd', '6'),) 1 PROTEIN: ()\n",
      "\n",
      "DNA: d6 (('1002', 'd', '6'),) 1 PROTEIN: ((334, 's', 'W'), (335, 'd'))\n",
      "\n",
      "DNA: d6 (('928', 'd', '6'),) 1 PROTEIN: ((310, 'd'), (311, 'd'), (335, 's', 'I'))\n",
      "\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>S6.d3</th>\n",
       "      <th>S6.d6</th>\n",
       "      <th>S6.d9</th>\n",
       "      <th>S6.i3</th>\n",
       "      <th>S6.i6</th>\n",
       "      <th>S6.i9</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>30</td>\n",
       "      <td>30</td>\n",
       "      <td>30</td>\n",
       "      <td>5579</td>\n",
       "      <td>81421</td>\n",
       "      <td>93662</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>18</td>\n",
       "      <td>35</td>\n",
       "      <td>28</td>\n",
       "      <td>3424</td>\n",
       "      <td>17979</td>\n",
       "      <td>8469</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>23</td>\n",
       "      <td>23</td>\n",
       "      <td>16</td>\n",
       "      <td>2296</td>\n",
       "      <td>5049</td>\n",
       "      <td>1177</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>13</td>\n",
       "      <td>19</td>\n",
       "      <td>22</td>\n",
       "      <td>1787</td>\n",
       "      <td>1625</td>\n",
       "      <td>260</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>11</td>\n",
       "      <td>12</td>\n",
       "      <td>10</td>\n",
       "      <td>1294</td>\n",
       "      <td>574</td>\n",
       "      <td>90</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "   S6.d3  S6.d6  S6.d9  S6.i3  S6.i6  S6.i9\n",
       "1     30     30     30   5579  81421  93662\n",
       "2     18     35     28   3424  17979   8469\n",
       "3     23     23     16   2296   5049   1177\n",
       "4     13     19     22   1787   1625    260\n",
       "5     11     12     10   1294    574     90"
      ]
     },
     "execution_count": 189,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "def dna_mutation_frequencies(all_references):\n",
    "    \"\"\"\n",
    "    Generate data for a histogram of DNA mutation frequencies - how many occur once, twice, ...\n",
    "    :param all_references:\n",
    "    :return:\n",
    "    \"\"\"\n",
    "    freq = {}  # for each library ('d3', 'i6', etc.) give a dictionary {1: 5 times, 2: 3 times...}\n",
    "    for background in all_references.keys():\n",
    "        for fraction in all_references[background].keys():\n",
    "            freq['.'.join([background,fraction])] = defaultdict(int)\n",
    "            for prot_mutation, entry in all_references[background][fraction].items():\n",
    "                for dna_mutation, dna_c in entry['dna'].items():\n",
    "                    dna_mut_type = classify_dna(dna_mutation)\n",
    "                    prot_mut_type = classify_protein(prot_mutation)\n",
    "                    if dna_mut_type == fraction:\n",
    "                        if prot_mut_type in dna_protein_correspondence(dna_mut_type):\n",
    "                            freq['.'.join([background,fraction])][dna_c] += 1\n",
    "                        elif dna_mut_type in ['d6']:\n",
    "                            print('DNA:', dna_mut_type, dna_mutation, dna_c, 'PROTEIN:', prot_mutation)\n",
    "                            print()\n",
    "    return freq\n",
    "\n",
    "hist_Poisson = dna_mutation_frequencies(all_ref)\n",
    "df_hist_Poisson = pd.DataFrame(hist_Poisson, columns=fractions).fillna(0).astype('int32')\n",
    "df_hist_Poisson.head()\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 190,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "S6.d3       639\n",
       "S6.d6       690\n",
       "S6.d9       613\n",
       "S6.i3     20872\n",
       "S6.i6    107165\n",
       "S6.i9    103720\n",
       "dtype: int64"
      ]
     },
     "execution_count": 190,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df_hist_Poisson.sum()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "index = number of times a mutation is observed\n",
    "\n",
    "column = number of such mutations\n",
    "\n",
    "This still needs manual clipping."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 142,
   "metadata": {},
   "outputs": [],
   "source": [
    "def plot_S7_Poisson(library,title, cutoffs, top_axes):\n",
    "    \n",
    "    # set the cutoffs betwen which we aggregate values, such that we count:\n",
    "    # - 1, 2, 3, ... up to 49\n",
    "    # - 50-74\n",
    "    # - 75-99 etc.\n",
    "    binned_counts = []\n",
    "    hist_data = df_hist_Poisson[library].iloc[:cutoffs[0] + len(cutoffs)]\n",
    "    main_max = df_hist_Poisson[library].max() # the axis for bottom group will cut off here\n",
    "\n",
    "    for i in range(len(cutoffs) - 1 ):\n",
    "        val = df_hist_Poisson[library].loc[cutoffs[i]:cutoffs[i+1]].sum()\n",
    "        binned_counts.append(val)\n",
    "        hist_data.iloc[[cutoffs[0] + i ]] = val\n",
    "    # the last bin is open-ended and needs special handling\n",
    "    hist_data.iloc[-1] = df_hist_Poisson[library].loc[cutoffs[-1]:].sum()\n",
    "    binned_counts.append(df_hist_Poisson[library].loc[cutoffs[-1]:].sum())\n",
    "    # this now contains the correct values, but labels are wrong\n",
    "    # the labesls for aggregate bin need to be done separately anyway, so this is no issue\n",
    "    \n",
    "    f, (ax, ax2) = plt.subplots(2, 1, sharex=True, gridspec_kw={'height_ratios': [1, 4]}, figsize=(8,5))\n",
    "\n",
    "    f.suptitle(title)\n",
    "    \n",
    "    bins1 = list(range(1, cutoffs[0] + 1))\n",
    "    bins2 = list(range(cutoffs[0]+6, cutoffs[0]+6 + 6*(len(cutoffs)), 6))\n",
    "    bins =  bins1 + bins2\n",
    "    ax.bar(bins, hist_data) # first arguement is x-position, second is height\n",
    "    ax2.bar(bins, hist_data)\n",
    "\n",
    "    # zoom-in / limit the view to different portions of the data\n",
    "    ax.set_ylim(top_axes[0], top_axes[1])  # outliers only\n",
    "    ax2.set_ylim(0, main_max + 1)  # most of the data\n",
    "    \n",
    "    ax2.set_xlabel('Read count (X)')\n",
    "    ax2.set_ylabel('Number of mutations')\n",
    "\n",
    "    # hide the spines between ax and ax2\n",
    "    ax.spines['bottom'].set_visible(False)\n",
    "    ax2.spines['top'].set_visible(False)\n",
    "\n",
    "    # set up the tick label positions and locatio\n",
    "\n",
    "    xlabels_main_pos = list(range(0,cutoffs[0] + 1, 10))\n",
    "    xlabels_main_pos[0] = 1 # first bin is one observation, not zero\n",
    "    xlabels_main = [str(pos) for pos in xlabels_main_pos]\n",
    "    \n",
    "    xlabels_grouped = []\n",
    "    for i in range(len(cutoffs)):\n",
    "        xlabels_grouped.append(str(cutoffs[i] +1) + '+')\n",
    " \n",
    "    xlabels = xlabels_main + xlabels_grouped\n",
    "    xlabels_pos = xlabels_main_pos + bins2                    \n",
    "    ax2.xaxis.set_ticks(xlabels_pos)\n",
    "    ax2.set_xticklabels(xlabels)\n",
    "    # hide the axes labels on top\n",
    "    ax.axes.get_xaxis().set_visible(False) \n",
    "\n",
    "\n",
    "    # This looks pretty good, and was fairly painless, but you can get that\n",
    "    # cut-out diagonal lines look with just a bit more work. The important\n",
    "    # thing to know here is that in axes coordinates, which are always\n",
    "    # between 0-1, spine endpoints are at these locations (0,0), (0,1),\n",
    "    # (1,0), and (1,1).  Thus, we just need to put the diagonals in the\n",
    "    # appropriate corners of each of our axes, and so long as we use the\n",
    "    # right transform and disable clipping.\n",
    "\n",
    "    d = .012  # how big to make the diagonal lines in axes coordinates\n",
    "    # arguments to pass to plot, just so we don't keep repeating them\n",
    "    kwargs = dict(transform=ax.transAxes, color='k', clip_on=False)\n",
    "    ax.plot((-d, +d), (-4*d, +4*d), **kwargs)        # top-left diagonal\n",
    "    ax.plot((1 - d, 1 + d), (-4*d, +4*d), **kwargs)  # top-right diagonal\n",
    "\n",
    "    kwargs.update(transform=ax2.transAxes)  # switch to the bottom axes\n",
    "    ax2.plot((-d, +d), (1 - d, 1 + d), **kwargs)  # bottom-left diagonal\n",
    "    ax2.plot((1 - d, 1 + d), (1 - d, 1 + d), **kwargs)  # bottom-right diagonal\n",
    "\n",
    "#     f.savefig('FigS7_' + library +'.png', format='png', dpi=600)\n",
    "    plt.show()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 143,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "image/png": "iVBORw0KGgoAAAANSUhEUgAAAfMAAAFhCAYAAABkoUMDAAAABHNCSVQICAgIfAhkiAAAAAlwSFlzAAALEgAACxIB0t1+/AAAADh0RVh0U29mdHdhcmUAbWF0cGxvdGxpYiB2ZXJzaW9uMy4xLjAsIGh0dHA6Ly9tYXRwbG90bGliLm9yZy+17YcXAAAgAElEQVR4nO3deZQmdX3v8feHzQVQQAYyLOOMyo0aF9QR8WIMiwuKYYt4xQ0E75gbURBiRKNBjuYEYtxjvBkDQnIJYEAWBRVCAE9iRHYYBCPCqAgyKCjgghn43j+emkMzdE9XL9VPV/f7dc5z+qnfU8u3err7M1X1q1+lqpAkSf213rALkCRJU2OYS5LUc4a5JEk9Z5hLktRzhrkkST1nmEuS1HOGuSSSVJKnDbsOSZNjmEtzUJLXJ/lukl8kWZXk5CRPGHZdkrphmEtz038Au1TVE4GnABsAHxluSZK6YphLc1BV/aiqfjqi6UFgvNPor05yS5KfJvlokvUAkhyc5D+SfKY50r8pyR6dFS9pwjYYdgGSupHkJcB5wBOAXwH7jbPIfsBSYBPgX4HvAv/QfPYi4AxgS2B/4EtJllTV3R2ULmmCPDKX5qiq+vfmNPt2wEeBleMscnxV3V1VPwQ+CRw44rNVwCer6r+r6nQGQb9XB2VLmgTDXOq5JG9Mcn/z+uran1fVj4GvAaeNs6ofjXj/A2CbEdM/rkc+lWntzyUNkWEu9VxVnVJVmzSvV40x2wbAU8dZ1fYj3i8Cbh8xvW2SrONzSUNkmEtzUHO0vigDTwb+ErhonMXek2TzJNsDhwOnj/hsK+BdSTZMcgDwDOD8ToqXNGF2gJPmpmcCxwObA/cwCN73jbPMOcCVwBOBk4ATRnx2GbAD8FPgTuC1VfWz6S1Z0mTlkZfBJOmRkhwMvK2qXjLsWiSNztPskiT1nGEuSVLPeZpdkqSe88hckqSeM8wlSeo5w1ySpJ4zzCVJ6jnDXJKknjPMJUnqOcNckqSeM8wlSeo5w1ySpJ4zzCVJ6jnDXJKknjPMJUnqOcNckqSeM8wlSeo5w1ySpJ4zzCVJ6jnDXJKknjPMJUnquSmHeZITk6xKsmJE20eT3JTkuiRnJdlsxGfvS3Jzku8meeVUty9J0nw3HUfmJwF7rtV2IfCsqnoO8F/A+wCSPBN4PfB7zTJ/l2T9aahBkqR5a8phXlXfAO5eq+2CqlrdTH4L2K55vw9wWlU9UFW3AjcDO021BkmS5rOZuGZ+CPDV5v22wI9GfHZb0yZJkiZpgy5XnuTPgdXAKWuaRpmtxlh2GbAMYOONN37B05/+9E5qlCRptrnyyit/WlUL2s7fWZgnOQh4DbBHVa0J7NuA7UfMth1w+2jLV9VyYDnA0qVL64orruiqVEmSZpUkP5jI/J2cZk+yJ/BeYO+q+tWIj84FXp/kMUmWADsA3+6iBkmS5ospH5knORXYFdgyyW3AMQx6rz8GuDAJwLeq6o+r6oYkXwS+w+D0+zuq6sGp1iBJ0nyWh8+Az16eZpckzSdJrqyqpW3ndwQ4SZJ6zjCXJKnnOr01TZI0dYuPPu9RbSuP22sIlWi28shckqSeM8wlSeo5w1ySpJ4zzCVJ6jnDXJKknjPMJUnqOcNckqSeM8wlSeo5w1ySpJ4zzCVJ6jnDXJKknjPMJUnqOcNckqSem3KYJzkxyaokK0a0bZHkwiTfa75u3rQnyaeT3JzkuiTPn+r2JUma76bjyPwkYM+12o4GLqqqHYCLmmmAVwE7NK9lwOemYfuSJM1rUw7zqvoGcPdazfsAJzfvTwb2HdH+jzXwLWCzJAunWoMkSfNZV9fMt66qOwCar1s17dsCPxox321NmyRJmqSZ7gCXUdpq1BmTZUmuSHLFXXfd1XFZkiT1V1dhfuea0+fN11VN+23A9iPm2w64fbQVVNXyqlpaVUsXLFjQUZmSJPVfV2F+LnBQ8/4g4JwR7W9perXvDPxizel4SZI0ORtMdQVJTgV2BbZMchtwDHAc8MUkhwI/BA5oZj8feDVwM/Ar4K1T3b4kSfPdlMO8qg4c46M9Rpm3gHdMdZuSJOlhjgAnSVLPGeaSJPWcYS5JUs9N+Zq5JEkzZfHR5z2qbeVxew2hktnFI3NJknrOMJckqecMc0mSes4wlySp5wxzSZJ6zjCXJKnnDHNJknrO+8wlSa2sfY+393fPHh6ZS5LUc4a5JEk9Z5hLktRzhrkkST3XaZgneXeSG5KsSHJqkscmWZLksiTfS3J6ko0mss7777+fz372s7z85S9n9erVXZUuSdKskIGXJjkryZ6jzdNZmCfZFngXsLSqngWsD7weOB74RFXtANwDHNpmfbfddhvvfe972X777TnssMO47777+MlPftJV+ZIkDVWSjZK8CbgCuBT4fWCr0ebt+ta0DYDHJflv4PHAHcDuwBuaz08GPgR8bqwVJHnhFltswZIlS3jooYfYf//9OfLII3nxi1/ccemSJM28u+++G+B3gFuBbYCbgD8G/qmqfjXaMqmqzgpKcjjwl8CvgQuAw4FvVdXTms+3B77aHLmPXG594G+BNwGbAA8BnwQ+U1UrOytYkqQhSfK7DHLyYOBxwIXAJ4CvV9VD61q2syPzJJsD+wBLgJ8D/wK8apRZa8QyTwAOYXB6fgmwEvggcGJV3dtVrZIkDUOSALsBRwJ7Ab8F/h/wyaq6vu16ujzN/jLg1qq6CyDJl4D/CWyWZIOqWg1sB9zefP584GLgCcB/AO8Bzq6qBzusUZKkYboQ2AO4CzgW+FxV3TnRlXQZ5j8Edk7yeAan2fdgcBH/YuC1wGnAQcA5zfwrgFMZHIV/u8O6JEmaLc5kkH2nVNVvJruSrq+ZHwv8L2A1cDXwNmBbBkG+RdP2pqp6oLMiJEma4zoNc0mS1D1HgJMkqecMc0mSes4wlySp5wxzSZJ6zjCXJKnnDHNJknrOMJckqecMc0mSes4wlySp5wxzSZJ6zjCXJKnnDHNJknrOMJckqecMc0mSes4wlySp53ob5kmeNOwaJEmaKUk2SfLY0T7rVZhn4OVJzgduTbLZsGuSJKlLSbZLcjzwI+CQ0ebZYGZLmpzmfyJvAN4NPAu4E/goUMOsS5KkriR5IYPcO4DBwfeZwGWjzTurwzzJVsD/Af4E2Aq4DjgYOK2qHhhiaZIkTbsk6wP7MAjxlwD3Ap8GPlNVK8dablaGeZJnMdiRNwKPWbBgAaeddhq77bbbc5KcBJw0zPokaSYtPvq8R7WtPG6vodcxG2oYVh3T7b777uPEE09kyZIl3HrrrQC/Bo4AvlBV9463/Ky5Zp5kvSSvSnIBcD1wIHAi8PRFixax++67k2S4RUqSNI1+8IMfcNRRR7HddttxxBFHsM0223DmmWcCfKeqPtUmyGF2HZkvBc4H7gDeDyyvqp8BLF26dJh1SZLUiX322YcVK1bwute9jne/+9288IUvnNR6OgvzptPaN4DHNNs5o6qOSbIEOA3YArgKeHNV/Ra4HNgb+HozLUnSnLZ8+XIWLlzI9ttvP6X1dHma/QFg96p6LrAjsGeSnYHjgU9U1Q7APcChADXwZYNckjRf7LTTTlMOcugwzJtwvr+Z3LB5FbA7cEbTfjKwb1c1SJI0H3TaAS7J+kmuAVYBFwLfB35eVaubWW4Dtu2yBkmS5rpOO8BV1YPAjs1IbWcBzxhtttGWTbIMWAawaNGiaa1rrt7aIEman2bk1rSq+jlwCbAzsFmSNf+J2A64fYxlllfV0qpaumDBgpkoU5KkXuoszJMsWDN2epLHAS8DbgQuBl7bzHYQcE5XNUiSNB90eZp9IXByMzTdesAXq+orSb4DnJbkI8DVwAkd1iBJ0pzXWZhX1XXA80ZpvwXYqavtSpI038ya4VwlSdLkGOaSJPWcYS5JUs8Z5pIk9ZxhLklSzxnmkiT1nGEuSVLPGeaSJPWcYS5JUs8Z5pIk9ZxhLklSzxnmkiT1nGEuSVLPGeaSJPWcYS5JUs8Z5pIk9ZxhLklSzxnmkiT1nGEuSVLPGeaSJPWcYS5JUs8Z5pIk9ZxhLklSz3UW5km2T3JxkhuT3JDk8Kb9Q0l+nOSa5vXqrmqQJGk+2KDDda8Gjqqqq5JsClyZ5MLms09U1d90uG1JkuaNzsK8qu4A7mje35fkRmDbrrYnSdJ8NSPXzJMsBp4HXNY0HZbkuiQnJtl8jGWWJbkiyRV33XXXTJQpSVIvdR7mSTYBzgSOqKp7gc8BTwV2ZHDk/rHRlquq5VW1tKqWLliwoOsyJUnqrU7DPMmGDIL8lKr6EkBV3VlVD1bVQ8DngZ26rEGSpLmuy97sAU4Abqyqj49oXzhitv2AFV3VIEnSfNBlb/ZdgDcD1ye5pml7P3Bgkh2BAlYCb++wBkmS5rwue7P/O5BRPjq/q21KkjQfOQKcJEk9Z5hLktRzhrkkST03bpgnOaAZjpUkH0jypSTP7740SZLURpsj8w82w7G+BHglcDKDgV8kSdIs0CbMH2y+7gV8rqrOATbqriRJkjQRbcL8x0n+HngdcH6Sx7RcTpIkzYA295m/DtgT+Juq+nkzgtt7ui1r9lh89HmPalt53F5DqESSpNGNe4RdVb8CzgF+mWQRsCFwU9eFSZKkdsY9Mk/yTuAY4E7goaa5gOd0WJckSWqpzWn2w4HfraqfdV2MJEmauDYd2X4E/KLrQiRJ0uS0OTK/BbgkyXnAA2saRz7WVJIkDU+bMP9h89oI7y+XJGnWGTfMq+pYgGZI16qq+zuvSpIktdZmbPZnJbkaWAHckOTKJL/XfWmSJKmNNh3glgNHVtWTq+rJwFHA57stS5IktdUmzDeuqovXTFTVJcDGnVUkSZImpFVv9iQfBP6pmX4TcGt3JUmSpIloc2R+CLAA+BJwVvP+rV0WJUmS2mvTm/0e4F0zUIskSZqEMcM8ySer6ogkX2YwFvsjVNXenVYmSZJaWdeR+Zpr5H8zmRUn2R74R+B3GDygZXlVfSrJFsDpwGJgJfC65uhfkiRNwpjXzKvqyubtjlV16cgXsGOLda8GjqqqZwA7A+9I8kzgaOCiqtoBuKiZliRJk9SmA9xBo7QdPN5CVXVHVV3VvL8PuBHYFtgHOLmZ7WRg31aVSpKkUa3rmvmBwBuAJUnOHfHRpsCEHoeaZDHwPOAyYOuqugMGgZ9kqwnWLEmSRljXNfNvAncAWwIfG9F+H3Bd2w0k2QQ4Eziiqu5N0na5ZcAygEWLFrXdnCRJ886YYV5VPwB+ALx4sitPsiGDID+lqr7UNN+ZZGFzVL4QWDXG9pczGEqWpUuXPqo3vSRJGmjzoJWdk1ye5P4kv03yYJJ7WywX4ATgxrWefX4uD1+HPwg4ZzKFS5KkgTbDuf4t8HrgX4ClwFuAp7VYbhfgzcD1Sa5p2t4PHAd8McmhDJ6TfsBEi5YkSQ9rE+ZU1c1J1q+qB4EvJPlmi2X+HRjrAvkeE6hRkiStQ5sw/1WSjYBrkvw1g05xPjVNkqRZos195m9u5jsM+CWwPbB/l0VJkqT22oT5vlX1m6q6t6qOraojgdd0XZgkSWqnsxHgJEnSzJiREeAkSVJ3Oh8BTpIkdavTEeAkSVL3xr01Lcl9wJrhVDcCNgR+WVVP6LIwSZLUzrhhXlWbjpxOsi+wU2cVSZKkCWnTm/0RqupsYPcOapEkSZPQ5jT7yAFi1mMwPrtPMZMkaZZoM5zrH454vxpYCezTSTWSJGnC2lwzf+tMFCJJkianzWn2JcA7gcUj56+qvbsrS5IktdXmNPvZwAnAl4GHui1nuBYffd4jplcet9eQKpEkqb02Yf6bqvp055VIkqRJaRPmn0pyDHAB8MCaxqq6qrOqJElSa23C/NkMnmm+Ow+fZi+811ySpFmhTZjvBzylqn7bdTGSJGni2owAdy2wWdeFSJKkyWlzZL41cFOSy3nkNXNvTZMkaRZoE+bHdF6FJEmatDYjwF06E4VIkqTJmfBT09pKcmKSVUlWjGj7UJIfJ7mmeb26q+1LkjRfdBbmwEnAnqO0f6Kqdmxe53e4fUmS5oUxwzzJRc3X4yez4qr6BnD3JOuSJEktreua+cIkfwDsneQ0ICM/nMIIcIcleQtwBXBUVd0zyfVIkiTWHeZ/ARwNbAd8fK3PJjsC3OeADzfLfxj4GHDIaDMmWQYsA1i0aNEkNiVJ0vwwZphX1RnAGUk+WFUfno6NVdWda94n+TzwlXXMuxxYDrB06dKaju1LkjQXtbk17cNJ9gZe2jRdUlVjhvC6JFlYVXc0k/sBK9Y1vyRJGt+4YZ7kr4CdgFOapsOT7FJV7xtnuVOBXYEtk9zGYPCZXZPsyOA0+0rg7ZMvXZIkQbsR4PYCdqyqhwCSnAxcDawzzKvqwFGaT5hwhZIkaZ3a3mc+8kErT+yiEEmSNDltjsz/Crg6ycUMbk97KeMclUuSpJnTpgPcqUkuAV7IIMzfW1U/6bowSZLUTpsjc5oe6Od2XIskSZqELsdmlyRJM8AwlySp59YZ5knWG/kIU0mSNPusM8ybe8uvTeLg6JIkzVJtOsAtBG5I8m3gl2saq2rvzqrqgcVHn/eI6ZXH7dXJeqdz3TNpruyHJPVBmzA/tvMqJEnSpLW5z/zSJE8Gdqiqf03yeGD97kuTJEltjNubPcn/Bs4A/r5p2hY4u8uiJElSe21uTXsHsAtwL0BVfQ/YqsuiJElSe23C/IGq+u2aiSQbMHiEqSRJmgXadIC7NMn7gccleTnwJ8CXuy1rbrFntySpS22OzI8G7gKuB94OnA98oMuiJElSe216sz+U5GTgMgan179bVZ5mlyRplhg3zJPsBfxf4PsMHoG6JMnbq+qrXRcnSZLG1+aa+ceA3arqZoAkTwXOAwxzSZJmgTbXzFetCfLGLcCqjuqRJEkTNOaReZL9m7c3JDkf+CKDa+YHAJfPQG2SJKmFdZ1m/8MR7+8E/qB5fxeweWcVqTVveZMkwTrCvKreOpOFSJKkyWnTm30J8E5g8cj5x3sEapITgdcwuOb+rKZtC+D0Zl0rgddV1T2TK12SJEG7DnBnMwjezzDo2b7mNZ6TgD3XajsauKiqdgAuaqYlSdIUtLk17TdV9emJrriqvpFk8VrN+wC7Nu9PBi4B3jvRdUuSpIe1CfNPJTkGuAB4YE1jVV01ie1tXVV3NMvfkcSnr0mSNEVtwvzZwJuB3YGHmrZqpjuTZBmwDGDRokVdbmpard3DfLp6l8/Uete1bnvPS9Ls1CbM9wOeMvIxqFNwZ5KFzVH5QtYx+ExVLQeWAyxdutSx4CVJGkObDnDXAptN0/bOBQ5q3h8EnDNN65Ukad5qc2S+NXBTkst55DXz8W5NO5VBZ7ctk9wGHAMcB3wxyaHADxmMJidJkqagTZgfM5kVV9WBY3y0x2TWJ0mSRtfmeeaXzkQhkiRpctqMAHcfg97rABsBGwK/rKondFmYpl9XPeL7xl75kqZqtv09bXNkvunI6ST7Ajt1VpEkSZqQNr3ZH6Gqzqbje8wlSVJ7bU6z7z9icj1gKQ+fdpckSUPWpjf7yOear2bw0JV9OqlGkiRNWJtr5j7XXJKkWWzMME/yF+tYrqrqwx3UI0mSJmhdR+a/HKVtY+BQ4EmAYS5J0iwwZphX1cfWvE+yKXA48FbgNOBjYy0nSZJm1jqvmSfZAjgSeCNwMvD8qrpnJgqTJEntrOua+UeB/Rk8hvTZVXX/jFUlSZJaW9egMUcB2wAfAG5Pcm/zui/JvTNTniRJGs+6rplPeHQ4SZI089oMGqOOzLaB+jU3+CAZaf7x6FuSpJ4zzCVJ6jnDXJKknjPMJUnqOcNckqSeM8wlSeo5b03TlE3kFruubsebzbdjzZbaRvvez3Rts+V7Ic01HplLktRzQzkyT7ISuA94EFhdVUuHUYckSXPBME+z71ZVPx3i9iVJmhM8zS5JUs8NK8wLuCDJlUmWDakGSZLmhGGdZt+lqm5PshVwYZKbquobI2doQn4ZwKJFi4ZRo4ZsrJ7PM917fqZ64Pe1V/ds6aE+V76f0mQM5ci8qm5vvq4CzgJ2GmWe5VW1tKqWLliwYKZLlCSpN2Y8zJNsnGTTNe+BVwArZroOSZLmimGcZt8aOCvJmu3/c1V9bQh1SJI0J8x4mFfVLcBzZ3q7kiTNVd6aJklSzzk2+xw0m3twz5btTZVjmo9vNv+bzubapMnwyFySpJ4zzCVJ6jnDXJKknjPMJUnqOcNckqSeM8wlSeo5b02TGrPhdqU+3oI2HWbD917qM4/MJUnqOcNckqSeM8wlSeo5w1ySpJ4zzCVJ6jl7s0vTaK48oGY29y6fSG0TuTtgrHln8/dCWsMjc0mSes4wlySp5wxzSZJ6zjCXJKnnDHNJknrO3uySWpmv48ZLfeCRuSRJPWeYS5LUc0MJ8yR7JvlukpuTHD2MGiRJmitmPMyTrA98FngV8EzgwCTPnOk6JEmaK4ZxZL4TcHNV3VJVvwVOA/YZQh2SJM0JwwjzbYEfjZi+rWmTJEmTkKqa2Q0mBwCvrKq3NdNvBnaqqneuNd8yYFkz+bvAdzsoZ0vgpx2sdxjm0r6sbS7vG7h/am+2fC9nQx2zoQboro4nV9WCtjMP4z7z24DtR0xvB9y+9kxVtRxY3mUhSa6oqqVdbmOmzKV9Wdtc3jdw/9TebPlezoY6ZkMNs6mOYZxmvxzYIcmSJBsBrwfOHUIdkiTNCTN+ZF5Vq5McBnwdWB84sapumOk6JEmaK4YynGtVnQ+cP4xtr6XT0/gzbC7ty9rm8r6B+6f2Zsv3cjbUMRtqgFlSx4x3gJMkSdPL4VwlSeq5eRnmSU5MsirJimHXMlGj1Z5kiyQXJvle83XzYdY4FUm2T3JxkhuT3JDk8KZ9Tuxjkscm+XaSa5v9O7ZpX5Lksmb/Tm86h/ZSkvWTXJ3kK830nNm3riVZmeT6JNckuaJp6/RnfyJ/UzLw6WYo7uuSPH8a65jQ7/501TJd+5/koGb+7yU5aGrfjYmbl2EOnATsOewiJukkHl370cBFVbUDcFEz3VergaOq6hnAzsA7muF+58o+PgDsXlXPBXYE9kyyM3A88Ilm/+4BDh1ijVN1OHDjiOm5tG8zYbeq2nHE7U7j/uwnOSnJrpPc3km0/5vyKmCH5rUM+NwotRyc5EOTqGOiv/vTVctJTHH/k2wBHAO8iMEop8es/Z+uJIuTXDJOLZM2L8O8qr4B3D3sOiZjjNr3AU5u3p8M7DujRU2jqrqjqq5q3t/HIBS2ZY7sYw3c30xu2LwK2B04o2nv7f4l2Q7YC/iHZjrMkX0bok5/9if4N2Uf4B+bn+NvAZslWThNdUz0d39aapmm/X8lcGFV3V1V9wAXMsMHjPMyzOegravqDhj8QgBbDbmeaZFkMfA84DLm0D42p6GvAVYx+KX/PvDzqlrdzNLnIY4/CfwZ8FAz/STmzr7NhAIuSHJlBqNgwnB+9sfa5owMx93yd7/LWia6zaEPUz6UW9Ok8STZBDgTOKKq7h0c4M0NVfUgsGOSzYCzgGeMNtvMVjV1SV4DrKqqK0ec8h3tH653+zaDdqmq25NsBVyY5KaxZkzySgaXMAAWAS9Jcj/wQFW9qKP6Rv33TPIkBqejAbYANkqy5mj2zVV1fesNtP/d77yWtttcRztJzgKWABsBi5r/yAN8qqq+MIVaHsEwnxvuTLKwqu5oTvmsGnZBU5FkQwa/zKdU1Zea5jm1jwBV9fPmGtrODE7XbdAcwY46xHEP7ALsneTVwGOBJzA4Up8L+zYjqur25uuqJgR2Yoyf/ar6OoPBt0hyEnBSVV0yTaWM9fs26nDcVfUzBn1ASHIwsLiqPjTRjU7wd7/LWia0zaZ917XaLwGoqv2aWhYz+DcaOd+08TT73HAusKb35EHAOUOsZUqaa6wnADdW1cdHfDQn9jHJguaInCSPA17G4NrgxcBrm9l6uX9V9b6q2q6qFjMYpvnfquqNzIF9mwlJNk6y6Zr3wCuAFQznZ3+sbZ4LvKXp1b0z8Is1p6OnahK/+53VMoltfh14RZLNm45vr2jaZk5VzbsXcCpwB/DfDP5Hdeiwa5pK7QyuS14EfK/5usWw65zC/r2Ewemp64Brmter58o+As8Brm72bwXwF037U4BvAzcD/wI8Zti1TnE/dwW+Mhf3rcPv2VOAa5vXDcCfN+3j/uwz6JG96yS32/pvCoPTyZ9l0M/jemDpKOs7GPjQJOqY0O/+dNUyXfsPHNL8jN8MvHWU7SwGLunq58cR4CRJ6jlPs0uS1HOGuSRJPWeYS5LUc4a5JEk9Z5hLktRzhrk0iyR5MIMnZq1I8uU196RPw3oPTvK307GuCW53cZI3rOPzhXn46Wr7J7loxGcvab4XGyR5TZonzEl6NMNcml1+XYMnZj2LwcMf3jHsgqZoMTBmmANHAp8HqMGIX79J8oYkGwB/B/xJDUaOO4/B6HKP77heqZcMc2n2+k9GPKwhyXuSXN48R/nYEe1nNw/muGHEwzlI8tYk/5XkUgZDrT5Kkk2SfCGDZ2hfl+SPmvYDm7YVSY4fMf/9I96/thlGdM0jOD+d5JtJbkmyZsS344Dfb46w3z1KCX8EfG3E9DuBjwDHApdX1Tdh8LQ5BsNjvmb8b5s0/zg2uzQLJVkf2IPB8JYkeQWDZyjvxGAUqnOTvLQGj288pKruboaHvTzJmQwe6nAs8ALgFwyGVL16lE19kMGQlM9utrN5km0YPMDjBQyeP35Bkn2r6uxxyl7IYBSvpzMY9vIMBs+B/tOqelQIJ1kC3FNVD6xpq6pbkpwOHAY8da1FrgB+H/jiOHVI845H5tLs8rjmqUo/Y/DEpwub9lc0r6uBqxgE5g7NZ+9Kci3wLQYPgdgBeBGDoSPvqqrfAqePsb2XMRieEoAaPIv5hSOWXQ2cAry0Re1nV9VDVfUdYOsW8y8E7hrZkGS9pqb7gSevNf8qYJsW65XmHcNcml1+XVU7MgiyjXj4mnmAv2qup+9YVU+rqhOaR42+DHhxVT2XQdg/tlmmzVjNGWW+dT1vduS8j13rswdGvG/zzNpfj7KOdzAYs/5Q4LN55PMvH9ssI2kthrk0C1XVLykdiK4AAAFPSURBVIB3AX/aPBby68AhzbOeSbJtBs+8fiKDU9W/SvJ0Bo9TBbgM2DXJk5rlDxhjUxcwOKVNs97Nm2X/IMmWzen+A4FLm1nuTPKM5gh6vxa7ch+w6Rif/ReDDnJrtv07DDrE/VlVfQ34MfC2EfP/DwZBL2kthrk0S1XV1QyeoPX6qroA+GfgP5Ncz+B69KYMOo9tkOQ64MMMTrVTg8cyfohBJ7p/ZXBqfjQfATZvOrpdC+zWLPs+BtfZrwWuqqo1j4A8GvgK8G8MnjQ1nuuA1UmuXbsDXFX9Evh+kqc1TR8H/rqq1px6PwL48yRbNNO7MejVLmktPjVN0tAk2Q94QVV9YJz5tgb+uar2mJnKpH6xN7ukoamqs5I8qcWsi4Cjuq5H6iuPzCVJ6jmvmUuS1HOGuSRJPWeYS5LUc4a5JEk9Z5hLktRzhrkkST33/wFPD9SGGDr/QgAAAABJRU5ErkJggg==\n",
      "text/plain": [
       "<Figure size 576x360 with 2 Axes>"
      ]
     },
     "metadata": {
      "needs_background": "light"
     },
     "output_type": "display_data"
    }
   ],
   "source": [
    "cutoffs = [49,99,199,999]\n",
    "top_axes = [80,120]\n",
    "plot_S7_Poisson('S6.d3', '-3 bp', cutoffs, top_axes) "
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "The insertion libraries have a simpler distribution, no split axes needed."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 15,
   "metadata": {},
   "outputs": [],
   "source": [
    "def plot_S7_Poisson_ins(library, cutoff,title):\n",
    "    hist_data = df_hist_Poisson[library].iloc[:cutoff + 1]\n",
    "    main_max = hist_data.max()\n",
    "    axes_split = hist_data[2]\n",
    "    after_trunc = df_hist_Poisson[library].iloc[cutoff:].sum()\n",
    "\n",
    "    hist_data.iloc[cutoff] = after_trunc # that has the data with the last entry in series being cumulative\n",
    "\n",
    "    \n",
    "    f, ax2 = plt.subplots(1, figsize=(5,5))\n",
    "\n",
    "    f.suptitle(title)\n",
    "    \n",
    "    bins = list(range(1, cutoff+1)) + [cutoff + 3]\n",
    "\n",
    "    ax2.bar(bins,hist_data, color='purple')\n",
    " \n",
    "    ax2.set_ylabel('Count (X)')\n",
    "    ax2.set_xlabel('Number of mutations in library observed X-times')\n",
    "\n",
    "\n",
    "    # set the tick label for overflow bin\n",
    "    xlabels_pos = [1,2,3,4,5,6,7,8,9,10,12]\n",
    "    xlabels = [str(pos) for pos in xlabels_pos]\n",
    "    xlabels[-1] = str(cutoff) + '+'\n",
    "    ax2.xaxis.set_ticks(xlabels_pos)\n",
    "    ax2.set_xticklabels(xlabels)\n",
    "\n",
    "    f.savefig('FigS7_' + library +'.png', format='png', dpi=600)\n",
    "    plt.show()\n",
    "\n",
    "\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 16,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "image/png": "iVBORw0KGgoAAAANSUhEUgAAAVMAAAFhCAYAAADN+UnSAAAABHNCSVQICAgIfAhkiAAAAAlwSFlzAAALEgAACxIB0t1+/AAAADh0RVh0U29mdHdhcmUAbWF0cGxvdGxpYiB2ZXJzaW9uMy4xLjAsIGh0dHA6Ly9tYXRwbG90bGliLm9yZy+17YcXAAAenklEQVR4nO3deZRlZX3u8e9j04ojgzQsZLiNpuOYiNpBItGr4AUHlhgjEWO09RJJlDhlxJuhQWOiSYxGE3UhEnGISFRCOyIyaLy5Io0ig4iNitJCoBXBgYgCv/vHfgsORU1d9Z7uqub7WavW2fvd+7z1O0M9tYez35OqQpK0MHfb2gVI0rbAMJWkDgxTSerAMJWkDgxTSerAMJWkDgxT3aUleWGSz2/tOrT0Gaba5iR5X5Krk/wwydeT/M7WrknbPsNUS1KSY5McO83ivwFWVtX9gGcAf5XkMVusON0lGaba5lTVJVV108Rs+3nQDHdJkrcmuSHJ15IcNLLgnCR/k+SLbflpSXYeZ/1amgxTbZOSvC3JjcDXgKuBT8yw+mOBbwK7AGuBj0wKzBcA/xt4AHAz8JaxFK0lzTDVNqmqXgrcF3g88BHgphlWvxZ4c1X9vKo+CFwGPH1k+Xur6uKq+gnwF8BvJlk2ptK1RBmmWjKSfCzJ9UmuB44BjpmYT/KxyetX1S1V9XlgT+AlM3T93brjiD/fZtgKnXDlpGXLGbZipdtst7ULkOaqqg6dmJ44+VRVx87hrtsx8zHTPZJkJFD3BtaNLN9rZHpv4OfA9+bwe3UX4paptilJdk1yRJL7JFmW5BDgucBZM9xtV+DlSZYnORx4KHc8xvrbSR6W5F7Aa4APVdUtY3sQWpLcMtW2phh26d/BsLHwbeCVVXXaDPc5F1jFsLV5DfDsqvr+yPL3Au8GHgJ8lpkPGeguKg4OLU0vyTnA+6rqhK1dixY3d/MlqQPDVJI6cDdfkjpwy1SSOjBMJakDw1SSOjBMJakDw1SSOjBMJakDw1SSOjBMJakDw1SSOjBMJakDw1SSOjBMJakDw1SSOjBMJakDw1SSOjBMJakDw1SSOjBMJakDw1SSOjBMJakDw1SSOjBMJamD7bZ2AeOwyy671MqVK7d2GZK2Meeff/73qmrFVMu2yTBduXIl69ev39plSNrGJPn2dMvczZekDgxTSerAMJWkDgxTSerAMJWkDgxTSerAMJWkDgxTSerAMJWkDgxTSerAMJWkDrbJa/Pn47gct6D7r621nSqRtBS5ZSpJHRimktSBYSpJHRimktSBYSpJHRimktSBYSpJHRimktSBYSpJHRimktSBYSpJHRimktSBYSpJHRimktSBYSpJHRimktSBYSpJHRimktSBYSpJHYw1TJNckeSiJBckWd/adk5yRpIN7Xan1p4kb0lyeZILkzx6pJ81bf0NSdaMs2ZJmo8tsWX6pKrat6pWt/ljgDOrahVwZpsHeCqwqv0cBbwdhvAF1gKPBfYD1k4EsCQtFltjN/8w4KQ2fRLwzJH299TgC8COSXYHDgHOqKrrquoHwBnAU7Z00ZI0k3GHaQGfTnJ+kqNa225VdTVAu921te8BXDly342tbbp2SVo0thtz/wdU1VVJdgXOSPK1GdbNFG01Q/sd7zyE9VEAe++993xqlaR5G+uWaVVd1W6vBU5lOOZ5Tdt9p91e21bfCOw1cvc9gatmaJ/8u46vqtVVtXrFihW9H4okzWhsYZrk3knuOzENHAxcDKwDJs7IrwFOa9PrgBe0s/r7Aze0wwCnAwcn2amdeDq4tUnSojHO3fzdgFOTTPyef62qTyU5DzglyZHAd4DD2/qfAJ4GXA7cCLwIoKquS/Ja4Ly23muq6rox1i1Jm21sYVpV3wQeOUX794GDpmgv4Ohp+joROLF3jZLUi1dASVIHhqkkdWCYSlIHhqkkdWCYSlIHhqkkdWCYSlIHhqkkdWCYSlIHhqkkdWCYSlIHhqkkdWCYSlIHhqkkdWCYSlIHhqkkdWCYSlIHhqkkdWCYSlIHhqkkdWCYSlIHhqkkdWCYSlIHhqkkdWCYSlIHhqkkdWCYSlIHhqkkdWCYSlIHhqkkdWCYSlIHhqkkdWCYSlIHhqkkdWCYSlIHhqkkdWCYSlIHhqkkdTD2ME2yLMmXk3ysze+T5NwkG5J8MMndW/s92vzlbfnKkT5e3dovS3LIuGuWpM21JbZMXwFcOjL/BuBNVbUK+AFwZGs/EvhBVf0C8Ka2HkkeBhwBPBx4CvC2JMu2QN2SNGdjDdMkewJPB05o8wEOBD7UVjkJeGabPqzN05Yf1NY/DDi5qm6qqm8BlwP7jbNuSdpc494yfTPwJ8Ctbf7+wPVVdXOb3wjs0ab3AK4EaMtvaOvf1j7FfSRpURhbmCY5FLi2qs4fbZ5i1Zpl2Uz3Gf19RyVZn2T9pk2bNrteSVqIcW6ZHgA8I8kVwMkMu/dvBnZMsl1bZ0/gqja9EdgLoC3fAbhutH2K+9ymqo6vqtVVtXrFihX9H40kzWBsYVpVr66qPatqJcMJpLOq6nnA2cCz22prgNPa9Lo2T1t+VlVVaz+ine3fB1gFfHFcdUvSfGw3+yrd/SlwcpK/Ar4MvKu1vwt4b5LLGbZIjwCoqkuSnAJ8FbgZOLqqbtnyZUvS9LZImFbVOcA5bfqbTHE2vqp+Chw+zf1fB7xufBVK0sJ4BZQkdWCYSlIHhqkkdWCYSlIHhqkkdWCYSlIHhqkkdWCYSlIHhqkkdWCYSlIHhqkkdWCYSlIHhqkkdWCYSlIHhqkkdWCYSlIHhqkkdWCYSlIHhqkkdWCYSlIHhqkkdWCYSlIHhqkkdWCYSlIHhqkkdbDdXFZKsitwAPAA4L+Bi4H1VXXrGGtb0o7LcQu6/9pa26kSSVvCjGGa5EnAMcDOwJeBa4HtgWcCD0ryIeCNVfXDcRcqSYvZbFumTwNeXFXfmbwgyXbAocD/Aj48htokacmYMUyr6o9nWHYz8O/dK5KkJWjGE1BJ3prkvlO0PyTJZ8ZXliQtLbOdzf8v4IIkvwWQ5F5J/hZYB/zzuIuTpKVixjCtqtcBTwael+RzwIXAzcC+VXXqFqhPkpaEuXzOtNrtdm39S6vqxvGVJElLz2zHTP8c+Azwnqp6HPB44LAkn03ysC1RoCQtBbN9NGoF8Kiq+hFAVX0XeHaSpzJ8HOqhY65PkpaE2Y6ZvmIiSCe1fxLYd2xVSdISM+tufpKdp1pWVTclOTDJoeMpTZKWjtl28y8CPprkp8CXgE0Ml5OuYtgy/Qzw12OtUJKWgNl280+rqgOA3wMuAZYBPwTeB+xXVa+qqk1T3TfJ9km+mOQrSS5JhpE/kuyT5NwkG5J8MMndW/s92vzlbfnKkb5e3dovS3JIjwcuST3NadSoqtoAbNjMvm8CDqyqHydZDnw+ySeBPwDeVFUnJ3kHcCTw9nb7g6r6hSRHAG8AntM+NXAE8HCGUas+k+QXq+qWzaxHksZmbOOZ1uDHbXZ5+yngQOBDrf0khhGoAA5r87TlByVJaz+5qm6qqm8BlwP7jatuSZqPsQ4OnWRZkgsYhu47A/gGcH0bJAVgI7BHm94DuBJuG0TlBuD+o+1T3EeSFoU5hWmSA+bSNllV3VJV+wJ7MmxNTvW51IkrrDLNsunaJ9dzVJL1SdZv2jTlYVxJGpu5bpm+dY5tU6qq64FzgP2BHdtYqDCE7FVteiOwF9w2VuoOwHWj7VPcZ/R3HF9Vq6tq9YoVK+ZamiR1MdtI+78KPA5YkeQPRhbdj+HM/kz3XQH8vKquT3JPhgFT3gCcDTwbOBlYA5zW7rKuzf+/tvysqqok64B/TfIPDCegVgFf3KxHKUljNtvZ/LsD92nrjY5r+kOGwJvJ7sBJSZYxbAGfUlUfS/JV4OQkf8XwVSjvauu/C3hvkssZtkiPAKiqS5KcAnyVYcSqoz2TL2mxmW2k/c8Cn03y7qr69uZ0XFUXAo+aov2bTHE2vqp+Chw+TV+vA163Ob9fkrakOX3OFLhHkuOBlaP3qaoDx1GUJM3FYvoW4LmG6b8B7wBOANzFlqRJ5hqmN1fV28daiSQtYXP9aNRHk7w0ye5Jdp74GWtlkrSEzHXLdE27Hf3q5wIe2LccSVqa5jrQyT7jLkSSlrI5hWmSF0zVXlXv6VuOJC1Nc93N/5WR6e2BgxgGizZMJYm57+a/bHQ+yQ7Ae8dSkSQtQfMdgu9GhmvkJUnM/ZjpR7l92LtlDEPpnTKuoiRpqZnrMdO/H5m+Gfh2VW0cQz2StCTNaTe/DXjyNYaRo3YCfjbOoiRpqZnrSPu/yTCG6OHAbwLnJpltCD5JusuY627+nwG/UlXXwm0DP3+G278YT5Lu0uZ6Nv9uE0HafH8z7itJ27y5bpl+KsnpwAfa/HOAT46nJElaeub6of0/TvIs4NcYvi30+Ko6dayVSdISMtsX6v0CsFtV/d+q+gjwkdb+hCQPqqpvbIkiJWmxm+2455uBH03RfmNbJkli9jBd2b4Y7w6qaj3D90FJkpg9TLefYdk9exYiSUvZbGF6XpIXT25MciRw/nhKkqSlZ7az+a8ETk3yPG4Pz9XA3YFfH2dhkrSUzBimVXUN8LgkTwIe0Zo/XlVnjb0ySVpC5vo507OBs8dciyQtWV4SKkkdGKaS1IFhKkkdGKaS1MFcR43SVnZcjpv3fdfW2o6VSJqKW6aS1IFhKkkdGKaS1IFhKkkdGKaS1IFhKkkdGKaS1MHYwjTJXknOTnJpkkuSvKK175zkjCQb2u1OrT1J3pLk8iQXJnn0SF9r2vobkqwZV82SNF/j3DK9GfjDqnoosD9wdJKHAccAZ1bVKuDMNg/wVGBV+zkKeDsM4QusBR4L7AesnQhgSVosxhamVXV1VX2pTf8IuBTYAzgMOKmtdhLwzDZ9GPCeGnwB2DHJ7sAhwBlVdV1V/QA4A3jKuOqWpPnYIsdMk6wEHgWcy/DV0VfDELjArm21PYArR+62sbVN1y5Ji8bYwzTJfYAPA6+sqh/OtOoUbTVD++Tfc1SS9UnWb9q0aX7FStI8jTVMkyxnCNL3V9VHWvM1bfeddntta98I7DVy9z2Bq2Zov4OqOr6qVlfV6hUrVvR9IJI0i3GezQ/wLuDSqvqHkUXrgIkz8muA00baX9DO6u8P3NAOA5wOHJxkp3bi6eDWJkmLxjiH4DsAeD5wUZILWtv/AV4PnNK+Lvo7wOFt2SeApwGXAzcCLwKoquuSvBY4r633mqq6box1S9JmG1uYVtXnmfp4J8BBU6xfwNHT9HUicGK/6iSpL6+AkqQODFNJ6sAwlaQODFNJ6sAwlaQODFNJ6sAwlaQOxvmhfS1Sx+W4Bd1/ba3tVIm07XDLVJI6MEwlqQPDVJI6MEwlqQPDVJI6MEwlqQPDVJI6MEwlqQPDVJI6MEwlqQPDVJI6MEwlqQPDVJI6MEwlqQPDVJI6MEwlqQPDVJI6MEwlqQPDVJI6MEwlqQPDVJI68NtJtWB+26nklqkkdWGYSlIHhqkkdWCYSlIHhqkkdWCYSlIHhqkkdWCYSlIHYwvTJCcmuTbJxSNtOyc5I8mGdrtTa0+StyS5PMmFSR49cp81bf0NSdaMq15JWohxbpm+G3jKpLZjgDOrahVwZpsHeCqwqv0cBbwdhvAF1gKPBfYD1k4EsCQtJmML06r6HHDdpObDgJPa9EnAM0fa31ODLwA7JtkdOAQ4o6quq6ofAGdw54CWpK1uS1+bv1tVXQ1QVVcn2bW17wFcObLextY2Xbu2YV7rr6VosZyAyhRtNUP7nTtIjkqyPsn6TZs2dS1OkmazpcP0mrb7Tru9trVvBPYaWW9P4KoZ2u+kqo6vqtVVtXrFihXdC5ekmWzpMF0HTJyRXwOcNtL+gnZWf3/ghnY44HTg4CQ7tRNPB7c2SVpUxnbMNMkHgCcCuyTZyHBW/vXAKUmOBL4DHN5W/wTwNOBy4EbgRQBVdV2S1wLntfVeU1WTT2pJ0lY3tjCtqudOs+igKdYt4Ohp+jkROLFjaZLU3WI5ASVJS5phKkkdGKaS1IFfqKdtnhcBaEtwy1SSOjBMJakDw1SSOjBMJakDw1SSOvBsvrSZ/HSApuKWqSR1YJhKUgeGqSR1YJhKUgeGqSR1YJhKUgeGqSR14OdMpa1sIZ9b9TOri4dbppLUgWEqSR0YppLUgcdMpW2I4wZsPW6ZSlIHhqkkdWCYSlIHhqkkdWCYSlIHhqkkdWCYSlIHhqkkdWCYSlIHhqkkdWCYSlIHhqkkdWCYSlIHhqkkdWCYSlIHhqkkdbBkwjTJU5JcluTyJMds7XokadSSCNMky4B/Bp4KPAx4bpKHbd2qJOl2SyJMgf2Ay6vqm1X1M+Bk4LCtXJMk3WaphOkewJUj8xtbmyQtCqmqrV3DrJIcDhxSVb/T5p8P7FdVLxtZ5yjgqDb7YOCyzmXsAnzvLtLfYq7trtbfYq5tHBZ7ff+jqlZMtWCpfDvpRmCvkfk9gatGV6iq44Hjx1VAkvVVtfqu0N9iru2u1t9irm0cFnt9M1kqu/nnAauS7JPk7sARwLqtXJMk3WZJbJlW1c1Jfh84HVgGnFhVl2zlsiTpNksiTAGq6hPAJ7ZiCb0PISzm/hZzbXe1/hZzbeOw2Oub1pI4ASVJi91SOWYqSYuaYTqLJCcmuTbJxR362ivJ2UkuTXJJklcssL/tk3wxyVdaf8cttMbW77IkX07ysQ59XZHkoiQXJFnfob8dk3woydfa8/ir8+znwa2miZ8fJnnlAmt7VXsdLk7ygSTbL7C/V7S+LplPbVO9d5PsnOSMJBva7U4LqXEp19ObYTq7dwNP6dTXzcAfVtVDgf2Boxd4WexNwIFV9UhgX+ApSfbvUOcrgEs79DPhSVW1b6ePvPwj8KmqegjwSOZZZ1Vd1mraF3gMcCNw6nyLSrIH8HJgdVU9guFE6REL6O8RwIsZrv57JHBoklWb2c27ufN79xjgzKpaBZzZ5reUedWT5N1Jnjj26hbIMJ1FVX0OuK5TX1dX1Zfa9I8YgmDeV3LV4Mdtdnn7WdBB8CR7Ak8HTlhIP+OQ5H7AE4B3AVTVz6rq+g5dHwR8o6q+vcB+tgPumWQ74F5M+iz0Znoo8IWqurGqbgY+C/z65nQwzXv3MOCkNn0S8MwF1LhZFls9vRmmW0mSlcCjgHMX2M+yJBcA1wJnVNWC+gPeDPwJcOsC+5lQwKeTnN+uUluIBwKbgH9phyFOSHLvhZfIEcAHFtJBVX0X+HvgO8DVwA1V9ekFdHkx8IQk909yL+Bp3PHClfnaraquhuGfO7Brhz63pXrmzTDdCpLcB/gw8Mqq+uFC+qqqW9qu6p7Afm33cL51HQpcW1XnL6SmSQ6oqkczjPh1dJInLKCv7YBHA2+vqkcBP2GBu6ntIpBnAP+2wH52YtjK2gd4AHDvJL893/6q6lLgDcAZwKeArzAcJrpLSHLIxPFshtfnhDa/0I2FsTFMt7AkyxmC9P1V9ZFe/bbd3XNY2PHdA4BnJLmCYWSuA5O8b4F1XdVur2U4JrnfArrbCGwc2fr+EEO4LsRTgS9V1TUL7OfJwLeqalNV/Rz4CPC4hXRYVe+qqkdX1RMYdo83LLBGgGuS7A7Qbq/t0Gf3eqrq9JFj2uuA32nzj92Ktc7IMN2CkoTheN+lVfUPHfpbkWTHNn1Phj/or823v6p6dVXtWVUrGXZ9z6qqeW9dJbl3kvtOTAMHM+y+zre+/wKuTPLg1nQQ8NX59tc8lwXu4jffAfZPcq/2Oh/EAk/iJdm13e4NPIs+da4D1rTpNcBpHfpciMVWz/xVlT8z/DC8ga8Gfs6wZXTkAvr6NYZjiBcCF7Sfpy2gv18Gvtz6uxj4y46P+4nAxxbYxwMZdk+/AlwC/FmHuvYF1rfH/O/ATgvo617A94EdOj1nxzH8M7sYeC9wjwX29x8M/yy+Ahw0j/vf6b0L3J/hrPmGdrtzr/fMuOph+BTAE7dUnfP98QooSerA3XxJ6sAwlaQODFNJ6sAwlaQODFNJ6sAwnUWSSvLGkfk/SnJsp77fneTZPfqa5fcc3kZYOrtzvyuT/NbmrpdkdZK39Kxl0u97TZInb8b6T5wYISvJM5Ic06a3yOszV0l+PPtaW9Z0z1GSdRm++HJi/p1J/niK9bboe2OcDNPZ3QQ8K8kuW7uQUUmWbcbqRwIvraondS5jJTBrmE5er6rWV9XLO9dym6r6y6r6zDzvu66qXj/X9dugJnNZL0kW1d/bXGufp5cDr8kwZOLjgMcyjPsw2Uq24HtjnBbVi7tI3czwVQqvmrxg8n/liS2HtqXz2SSnJPl6ktcneV6GsUcvSvKgkW6enOQ/2nqHtvsvS/J3Sc5LcmGS3x3p9+wk/wpcNEU9z239X5zkDa3tLxkuFnhHkr+btP6c6pzucQKvBx7frpl+VdvK+I8kX2o/j5tmvdEtwZ2T/Ht7nF9I8sut/dgM41+ek+SbSV7e2u+d5OMZxnC9OMlzZnpdMoynelyr56IkD5n6Zb7tvi9M8k+zvD4vTPJvST7KMIjLfZKcOfI7DmvrrcywR/A24EvAXyR508jvenGSO10JN9XrOLLsje33nJlkRWt7eZKvtufw5JHn6cT2HvrySE2Ta/9gkqdNeu5+Y4b3YJL8U/t9H2eagUmq6gqGv5u/Bd4G/H4Nl9lONtN749gkJyX5dHsdn5Xkb9tz86kMl2aT5DHtfXx+ktNz++Wpd3pexmprXzWw2H+AHwP3A64AdgD+CDh25MqMZ4+u226fCFwP7A7cA/gucFxb9grgzSP3/xTDP7VVDFeFbA8cBfx5W+ceDFf87NP6/QmwzxR1PoDhksYVDAOCnAU8sy07h2Gczcn32Zw6p3ucHxtpvxewfZteBayfZr3b5oG3Amvb9IHABW36WOA/W127MFyptBz4DeCdI33d6eql0Xrb6/ayNv1S4IRpnoeJel4I/NMsr88L2/TObb3tgPu16V2Ay4EwbHXdCuzflt0b+AawvM3/J/BLm/E6FvC8Nv2XI3VeRbvaCtix3f418NsTbcDX2++fXPuvAye16bsDVwL3ZPr34LMYBl9Z1mq9npH3xqTHsrw9lvfP8Pc103vjWODzrZ9HMow5+9S27FSG4fqWt+dxRWt/DsMXbk75vIzzxy3TOahhZKf3MOy6zNV5NYxfehPDH9DEcGwXMfyRTTilqm6tqg3AN4GHMFzD/oIMI+acy3DJ3cTAwF+sqm9N8ft+BTinhoE2bgbezzD2Z68652I58M4kFzGMwjSXga9/jeHSS6rqLOD+SXZoyz5eVTdV1fcYBsDYrdX15CRvSPL4qrphDr9jYkCZ89n8xzTV6wPDcIcTY3MG+OskFwKfYRijdre27NtV9YX2+H7CEI6Hti3k5VU1eQ9jptfxVuCDbfp9DM8dDJfWvj/DKFUTI0sdDBzT3kPnMPwT2HuK2j/JMKDNPRgGfflcVf03078HnwB8oIbRyq5qj2c6v9yem4dk/oc4PlnDFu1FDAH+qdY+8f58MPAI4IxW658zjKAGUz8vY2OYzt2bGY49jo6feTPtOUwShv/sE24amb51ZP5W7vitsJOv5y2GN+DLqo2aU1X71O1jY/5kmvoy1wcyyVzqnOlxjnoVcA3DVsTqGdYbNVXdE8/JaG23ANtV1dcZRsa/CPibDIcxZjPRzy1s/jfyTvX6wB1fh+cxbEk+poZRjq5hCK/J68Ew6PYLgRcB/zLF79uc13GilqcD/8zwvJyf4VhogN8YeQ/tXcOwfneoqap+yhC2hzBs1U3sDs/0Hpz1GvQWnm8Dns9w3f1LWvtjc/tXxTxjDo/xplbnrcDPq21mcvv7M8AlI3X+UlUdPMPzMjaG6Ry1/+SnMATqhCsYXigYxrJcPo+uD09ytwzHJx8IXAacDrxk5JjQL2b2QZDPBf5nkl0ynJx6LsPo7D1cwdSP80fAfUfW2wG4ur3xn8+wJTHVeqM+xxBGZPhqiu/VDGO8JnkAcGNVvY9hMOaFDsE3m6len8l2YBgH9udJngT8j+k6q2H4wL0YTrpMNQrUTK/j3YCJY9e/BXy+hdZeVXU2w6DeOwL3YXgPvaz98yPJo2Z4jCczhPvj2/1g+vfg54Aj2jHV3YHpTmr+LrChqs4B/gD4kyQrqurckeBbx8zvjbm4DFiR9l1gSZYnefgMz8vYjDWpt0FvBH5/ZP6dwGlJvsgw4s10W40zuYzhj2U34Peq6qdJTmDYhflS+2PYxCxf51BVVyd5NXA2w3/rT1RVr+HMpnucFwI3J/kKw/HFtwEfTnJ4q2O69b480vexDCPnX8hwTGwNM/sl4O+S3Mow+tBLFvTIZjfV6zN5nfcDH83whYEXMPswiKcA+1bVDyYvmOV1/Anw8CTnAzcwbEkuA97XDo0EeFNVXZ/ktQx7Uxe299AVwKHT1PNphsNY66rqZ61tuvfgqQzHti9iOA57p3/YGYYO/FOG7zmjqq5K8o8MJ6NeNGn1md4bs6qqn2U42fiW9hxs1x7315niedmcvjeXo0ZJW1g7W/2mqjpza9eiftzNl7aQDJ+5/Drw3wbptsctU0nqwC1TSerAMJWkDgxTSerAMJWkDgxTSerAMJWkDv4/UC6YvwJHW68AAAAASUVORK5CYII=\n",
      "text/plain": [
       "<Figure size 360x360 with 1 Axes>"
      ]
     },
     "metadata": {
      "needs_background": "light"
     },
     "output_type": "display_data"
    }
   ],
   "source": [
    "plot_S7_Poisson_ins('S6.i3', 10, '+3 bp') "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 171,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "ACGT composition of insertions, for each position separately\n",
      "{3: {0: {'A': 4586, 'C': 3076, 'G': 3827, 'T': 3840},\n",
      "     1: {'A': 4359, 'C': 3328, 'G': 3434, 'T': 4208},\n",
      "     2: {'A': 3531, 'C': 3643, 'G': 3094, 'T': 5061}},\n",
      " 6: {0: {'A': 5057, 'C': 4941, 'G': 9202, 'T': 6622},\n",
      "     1: {'A': 5046, 'C': 5273, 'G': 8572, 'T': 6931},\n",
      "     2: {'A': 4982, 'C': 5071, 'G': 8194, 'T': 7575},\n",
      "     3: {'A': 5427, 'C': 4955, 'G': 7761, 'T': 7679},\n",
      "     4: {'A': 5596, 'C': 4631, 'G': 8005, 'T': 7590},\n",
      "     5: {'A': 3627, 'C': 5589, 'G': 6323, 'T': 10283}},\n",
      " 9: {0: {'A': 2122, 'C': 1903, 'G': 3591, 'T': 2443},\n",
      "     1: {'A': 2070, 'C': 2193, 'G': 3208, 'T': 2588},\n",
      "     2: {'A': 1992, 'C': 2139, 'G': 3128, 'T': 2800},\n",
      "     3: {'A': 2153, 'C': 2145, 'G': 3007, 'T': 2754},\n",
      "     4: {'A': 2042, 'C': 2208, 'G': 3092, 'T': 2717},\n",
      "     5: {'A': 2089, 'C': 2032, 'G': 2948, 'T': 2990},\n",
      "     6: {'A': 2156, 'C': 1844, 'G': 3019, 'T': 3040},\n",
      "     7: {'A': 2228, 'C': 1776, 'G': 3087, 'T': 2968},\n",
      "     8: {'A': 1496, 'C': 2060, 'G': 2560, 'T': 3943}}}\n",
      "\n"
     ]
    }
   ],
   "source": [
    "def insertion_composition(all_references, cutoff=2, l=(3,6,9)):\n",
    "\n",
    "    comp = {length: {k: {'A':0, 'C':0, 'T':0, 'G':0} for k in range(length)} for length in l}\n",
    "    for background in all_references.keys():\n",
    "        for fraction in all_references[background].keys():\n",
    "            for mutation in all_references[background][fraction].keys():\n",
    "                for dna_error, c in all_references[background][fraction][mutation]['dna'].items():\n",
    "                    if dna_error is None:\n",
    "                        continue\n",
    "                    if c >= cutoff and len(dna_error) == 1 and classify_dna(dna_error) in ('i3', 'i6', 'i9'):\n",
    "                        ins = dna_error[0][2]\n",
    "                        ins_len = len(ins)\n",
    "                        if ins_len in l:\n",
    "                            for pos in range(ins_len):\n",
    "                                comp[ins_len][pos][ins[pos]] += 1\n",
    "    return comp\n",
    "\n",
    "\n",
    "# # 4. Get position by position ACGT composition of insertions - SI Figure\n",
    "ins_composition = insertion_composition(all_ref)\n",
    "print('ACGT composition of insertions, for each position separately')\n",
    "pprint.pprint(ins_composition)\n",
    "print()\n",
    "#"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.7.3"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 4
}
