{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 19,
   "metadata": {},
   "outputs": [],
   "source": [
    "#!/usr/bin/env python3\n",
    "\n",
    "import sys\n",
    "import pickle\n",
    "import re\n",
    "import os\n",
    "import csv\n",
    "import pprint\n",
    "\n",
    "import pandas as pd\n",
    "import numpy as np\n",
    "import matplotlib.pyplot as plt\n",
    "from collections import defaultdict\n",
    "\n",
    "# Demand Python 3.\n",
    "if sys.version_info[0] < 3:\n",
    "    print(\"Python 3 is required, but you are using Python %i.%i.%i\") % (\n",
    "        sys.version_info[0], sys.version_info[1], sys.version_info[2])\n",
    "    sys.exit(1)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "We need to change directory to wherever the input files are located."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {},
   "outputs": [],
   "source": [
    "os.chdir(\"/mnt/c//Users/Maya/Dropbox/PTE_results\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Later, load in the results for analysis\n",
    "with open('S6_full_all.p', 'rb') as f:\n",
    "    all_ref = pickle.load(f)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "dict_keys(['d9', 'd6', 'i6', 'd3', 'i3', 'i9'])\n"
     ]
    }
   ],
   "source": [
    "print(all_ref['S6'].keys())"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "The `all_ref` dictionary contains all information collected from sequenced, aggregated by protein background (here wtPTE, dubbed S6 for all fractions) and then by fraction / type of library. There is no filterin on this point, so `all_ref` includes all sequencing error, frameshifts, variants with extra point mutations etc.\n",
    "\n",
    "Each key in the dictionary takes the format:\n",
    "\n",
    "```\n",
    "protein_mutation : {'total': integer_count, \n",
    "                    'dna_hgvs': {'dna_in_HGVS_format': count},\n",
    "                    'dna': {'dna_in_tuple_format': count} # this is for each DNA mutation that generates that protein mutation\n",
    "                   }\n",
    "```\n",
    "So, if we want to know all mutations in a certain position in a library, we need `all_ref['S6']['d3](protein_tuple)['dna'].items()` or some such."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 34,
   "metadata": {},
   "outputs": [],
   "source": [
    "def classify_dna(dna_error):\n",
    "    if dna_error is None:  # empty or broken reads\n",
    "        return 'b'\n",
    "    elif dna_error == ():\n",
    "        return 'wt'\n",
    "    elif len(dna_error) > 1:\n",
    "        # expect substitutions\n",
    "        if dna_error[-1][1] == 'f':  # frameshifts are always the last mutation\n",
    "            return 'f'\n",
    "        else:\n",
    "            for k in range(len(dna_error)):\n",
    "                if dna_error[k][1] == 'i':\n",
    "                    return 'si'\n",
    "                elif dna_error[k][1] == 'd':\n",
    "                    return 'sd'\n",
    "            return 's'\n",
    "    elif len(dna_error) == 1:\n",
    "        if dna_error[0][1] == 'f':\n",
    "            return 'f'\n",
    "        elif dna_error[0][1] == 'd':\n",
    "            return 'd' + dna_error[0][2]  # length of deletion\n",
    "        elif dna_error[0][1] == 'i':\n",
    "            return 'i' + str(len(dna_error[0][2]))\n",
    "        elif dna_error[0][1] == 's':\n",
    "            return 's'\n",
    "    else:\n",
    "        return 'b'\n",
    "\n",
    "\n",
    "def get_dna_composition(all_references, cutoff=10):\n",
    "    \"\"\"\n",
    "    Classify all detected mutations by what kind they are: substitution, deletion, insertion (including length), frameshift, or other.\n",
    "    Collect the results in a dataframe, for both proteins and DNA mutations.\n",
    "    Only mutations\n",
    "    \"\"\"\n",
    "\n",
    "    dna_count = {}\n",
    "    dna_reads = {}\n",
    "    for background in all_references.keys():\n",
    "        for fraction in all_references[background].keys():\n",
    "            # print('Analysing background {0} and fraction {1}. Unusal mutations: '.format(background, fraction))\n",
    "            distinct_mutations = 0\n",
    "            total_count = 0\n",
    "            dna_count[background + '.' + fraction] = {'s': 0, 'd3': 0, 'd6': 0, 'd9': 0, 'i3': 0, 'i6': 0, 'i9': 0, 'f': 0,\n",
    "                                                'b': 0, 'other': 0, 'sd': 0, 'si': 0}\n",
    "            dna_reads[background + '.' + fraction] = {'s': 0, 'd3': 0, 'd6': 0, 'd9': 0, 'i3': 0, 'i6': 0, 'i9': 0, 'f': 0,\n",
    "                                                'b': 0, 'other': 0, 'sd': 0, 'si': 0}\n",
    "            for mutation in all_references[background][fraction].keys():\n",
    "                mut_total = all_references[background][fraction][mutation]['total']\n",
    "                if mut_total >= cutoff:\n",
    "                    # find all DNA entries with high enough counts\n",
    "                    for dna_error, c in all_references[background][fraction][mutation]['dna'].items():\n",
    "                        # dna_error = max(all_references[background][fraction][mutation]['dna'],\n",
    "                        #  key=lambda key: all_references[background][fraction][mutation]['dna'][key])\n",
    "                        if c >= cutoff:\n",
    "                            try:\n",
    "                                dna_type = classify_dna(dna_error)\n",
    "                                if dna_type == 'wt':\n",
    "                                    continue\n",
    "                                elif dna_type == 'b':\n",
    "                                    print(dna_type, dna_error, mut_total)\n",
    "                                dna_count[background + '.' + fraction][dna_type] += 1\n",
    "                                dna_reads[background + '.' + fraction][dna_type] += c\n",
    "                                distinct_mutations += 1\n",
    "                                total_count += c\n",
    "                            except KeyError:\n",
    "                                #print(dna_type, dna_error, mut_total)\n",
    "                                dna_count[background + '.' + fraction]['other'] += 1\n",
    "                                dna_reads[background + '.' + fraction]['other'] += c\n",
    "            # print('In background {0} and fraction {1} found {2} distinct mutations with total read count {3}'.format(\n",
    "            #     background, fraction, distinct_mutations, total_count\n",
    "            # ))\n",
    "\n",
    "    return dna_count, dna_reads\n",
    "\n",
    "\n",
    "def classify_protein(mutation):\n",
    "    \"\"\"\n",
    "    Detect whether a protein mutation is an insertion/deletion/substituion and if more than 1, (non)consecutive\n",
    "    :param mutation:\n",
    "    :return:\n",
    "    \"\"\"\n",
    "    if mutation is None:  # came from empty or broken reads\n",
    "        return 'b'\n",
    "    else:\n",
    "        m = []\n",
    "        for pos in range(len(mutation)):\n",
    "            t = mutation[pos][1]\n",
    "            if t != 'i':\n",
    "                m.append(t)\n",
    "            else:\n",
    "                m.append(t + str(len(mutation[pos][2])))\n",
    "        # need to distinguish between consecutive mutations and likely sequencing errors\n",
    "        if 'f' in m:\n",
    "            return 'f'\n",
    "        elif len(m) == 1:\n",
    "            return ''.join(m)\n",
    "        elif len(m) >= 1:\n",
    "            c = is_mutation_consecutive(mutation) + '-' + ''.join(m)\n",
    "            return c\n",
    "        \n",
    "def is_mutation_consecutive(mutation):\n",
    "    \"\"\"\n",
    "    If only consecutve amino acids are affected, return 'c', else return 'nc'\n",
    "    :param mutation:\n",
    "    :return:\n",
    "    \"\"\"\n",
    "    for pos in range(1, len(mutation)):\n",
    "        if mutation[pos][0] != (mutation[pos - 1][0] + 1) :\n",
    "            return 'nc'\n",
    "    return 'c'\n",
    "\n",
    "def dna_protein_correspondence(dna_type):\n",
    "    \n",
    "    dna_to_prot_type = {'d3':['d', 'c-sd'], 'd6':['c-dd', 'c-sdd'], 'd9':['c-ddd', 'c-sddd'],\n",
    "                'i3':['i1','c-si1'], 'i6': ['i2','c-si2'], 'i9':['i3','c-si3']}\n",
    "    \n",
    "    try:\n",
    "        return dna_to_prot_type[dna_type]\n",
    "    except KeyError:\n",
    "        []\n",
    "\n",
    "        \n",
    "def get_protein_composition(all_references, cutoff=10):\n",
    "    protein_count = {}\n",
    "    protein_reads = {}\n",
    "    for background in all_references.keys():\n",
    "        for fraction in all_references[background].keys():\n",
    "            # print('Analysing protein composition in background {0} and fraction {1}. Unusal mutations: '.format(background, fraction))\n",
    "            distinct_mutations = 0\n",
    "            total_count = 0\n",
    "            protein_count[background + '.' + fraction] = {'s': 0, 'd': 0, 'c-dd': 0, 'c-ddd': 0, 'i1': 0, 'i2': 0,\n",
    "                            'i3': 0, 'c-ss': 0, 'c-sd': 0, 'c-sdd': 0, 'c-sddd': 0, 'c-si1': 0, 'c-si2': 0, 'c-si3': 0,\n",
    "                            'f': 0, 'other': 0, 'b': 0}\n",
    "            protein_reads[background + '.' + fraction] = {'s': 0, 'd': 0, 'c-dd': 0, 'c-ddd': 0, 'i1': 0, 'i2': 0,\n",
    "                            'i3': 0, 'c-ss': 0, 'c-sd': 0, 'c-sdd': 0, 'c-sddd': 0, 'c-si1': 0, 'c-si2': 0, 'c-si3': 0,\n",
    "                            'f': 0, 'other': 0, 'b': 0}\n",
    "            for mutation in all_references[background][fraction].keys():\n",
    "                mut_total = all_references[background][fraction][mutation]['total']\n",
    "                if mut_total >= cutoff:\n",
    "                    # find all DNA entries with high enough counts\n",
    "                    try:\n",
    "                        prot_type = classify_protein(mutation)\n",
    "                        protein_count[background + '.' + fraction][prot_type] += 1\n",
    "                        protein_reads[background + '.' + fraction][prot_type] += mut_total\n",
    "                        distinct_mutations += 1\n",
    "                        total_count += mut_total\n",
    "                    except KeyError:\n",
    "                        # print(prot_type, mutation, mut_total)\n",
    "                        protein_count[background + '.' + fraction]['other'] += 1\n",
    "                        protein_reads[background + '.' + fraction]['other'] += mut_total\n",
    "            # print('In background {0} and fraction {1} found {2} distinct mutations with total read count {3}'.format(\n",
    "            #     background, fraction, distinct_mutations, total_count\n",
    "            # ))\n",
    "\n",
    "    return protein_count, protein_reads"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Overall composition of libraries\n",
    "Tables"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Now start with statistics\n",
    "1. Generate CSV files that give overall composition of libraries on protein and DNA level"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "b None 16097\n",
      "b None 17718\n",
      "b None 97637\n",
      "b None 14958\n",
      "b None 72875\n",
      "b None 90659\n"
     ]
    }
   ],
   "source": [
    "dna_count, dna_reads = get_dna_composition(all_ref, 1)\n",
    "protein_count, protein_reads = get_protein_composition(all_ref, 1)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>S6.d3</th>\n",
       "      <th>S6.d6</th>\n",
       "      <th>S6.d9</th>\n",
       "      <th>S6.i3</th>\n",
       "      <th>S6.i6</th>\n",
       "      <th>S6.i9</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>d3</th>\n",
       "      <td>643</td>\n",
       "      <td>224</td>\n",
       "      <td>278</td>\n",
       "      <td>208</td>\n",
       "      <td>211</td>\n",
       "      <td>266</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>d6</th>\n",
       "      <td>28</td>\n",
       "      <td>694</td>\n",
       "      <td>53</td>\n",
       "      <td>126</td>\n",
       "      <td>177</td>\n",
       "      <td>171</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>d9</th>\n",
       "      <td>5</td>\n",
       "      <td>241</td>\n",
       "      <td>617</td>\n",
       "      <td>80</td>\n",
       "      <td>66</td>\n",
       "      <td>76</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>i3</th>\n",
       "      <td>2</td>\n",
       "      <td>10</td>\n",
       "      <td>0</td>\n",
       "      <td>20880</td>\n",
       "      <td>310</td>\n",
       "      <td>154</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>i6</th>\n",
       "      <td>2</td>\n",
       "      <td>5</td>\n",
       "      <td>5</td>\n",
       "      <td>4</td>\n",
       "      <td>107175</td>\n",
       "      <td>932</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>i9</th>\n",
       "      <td>0</td>\n",
       "      <td>8</td>\n",
       "      <td>1</td>\n",
       "      <td>18</td>\n",
       "      <td>18</td>\n",
       "      <td>103724</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>s</th>\n",
       "      <td>6140</td>\n",
       "      <td>7303</td>\n",
       "      <td>6909</td>\n",
       "      <td>18296</td>\n",
       "      <td>18923</td>\n",
       "      <td>14528</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>sd</th>\n",
       "      <td>2333</td>\n",
       "      <td>1901</td>\n",
       "      <td>1785</td>\n",
       "      <td>1038</td>\n",
       "      <td>1878</td>\n",
       "      <td>2164</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>si</th>\n",
       "      <td>4</td>\n",
       "      <td>474</td>\n",
       "      <td>7</td>\n",
       "      <td>4712</td>\n",
       "      <td>6600</td>\n",
       "      <td>5586</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>other</th>\n",
       "      <td>35</td>\n",
       "      <td>167</td>\n",
       "      <td>252</td>\n",
       "      <td>332</td>\n",
       "      <td>378</td>\n",
       "      <td>422</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "       S6.d3  S6.d6  S6.d9  S6.i3   S6.i6   S6.i9\n",
       "d3       643    224    278    208     211     266\n",
       "d6        28    694     53    126     177     171\n",
       "d9         5    241    617     80      66      76\n",
       "i3         2     10      0  20880     310     154\n",
       "i6         2      5      5      4  107175     932\n",
       "i9         0      8      1     18      18  103724\n",
       "s       6140   7303   6909  18296   18923   14528\n",
       "sd      2333   1901   1785   1038    1878    2164\n",
       "si         4    474      7   4712    6600    5586\n",
       "other     35    167    252    332     378     422"
      ]
     },
     "execution_count": 8,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "row_index_dna = ['d3', 'd6', 'd9', 'i3', 'i6', 'i9', 's', 'sd', 'si', 'other']\n",
    "fractions = ['S6.d3', 'S6.d6', 'S6.d9', 'S6.i3', 'S6.i6', 'S6.i9']\n",
    "\n",
    "df_dna_count = pd.DataFrame(dna_count, index=row_index_dna, columns=fractions)\n",
    "df_dna_reads = pd.DataFrame(dna_reads, index=row_index_dna, columns=fractions)\n",
    "\n",
    "df_dna_count"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>S6.d3</th>\n",
       "      <th>S6.d6</th>\n",
       "      <th>S6.d9</th>\n",
       "      <th>S6.i3</th>\n",
       "      <th>S6.i6</th>\n",
       "      <th>S6.i9</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>d</th>\n",
       "      <td>302</td>\n",
       "      <td>146</td>\n",
       "      <td>188</td>\n",
       "      <td>143</td>\n",
       "      <td>151</td>\n",
       "      <td>183</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>c-sd</th>\n",
       "      <td>267</td>\n",
       "      <td>73</td>\n",
       "      <td>77</td>\n",
       "      <td>148</td>\n",
       "      <td>157</td>\n",
       "      <td>173</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>c-dd</th>\n",
       "      <td>19</td>\n",
       "      <td>322</td>\n",
       "      <td>42</td>\n",
       "      <td>82</td>\n",
       "      <td>135</td>\n",
       "      <td>128</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>c-sdd</th>\n",
       "      <td>11</td>\n",
       "      <td>292</td>\n",
       "      <td>19</td>\n",
       "      <td>109</td>\n",
       "      <td>131</td>\n",
       "      <td>136</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>c-ddd</th>\n",
       "      <td>5</td>\n",
       "      <td>145</td>\n",
       "      <td>307</td>\n",
       "      <td>73</td>\n",
       "      <td>53</td>\n",
       "      <td>65</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>c-sddd</th>\n",
       "      <td>1</td>\n",
       "      <td>78</td>\n",
       "      <td>220</td>\n",
       "      <td>56</td>\n",
       "      <td>93</td>\n",
       "      <td>72</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>i1</th>\n",
       "      <td>2</td>\n",
       "      <td>21</td>\n",
       "      <td>0</td>\n",
       "      <td>4890</td>\n",
       "      <td>297</td>\n",
       "      <td>129</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>c-si1</th>\n",
       "      <td>0</td>\n",
       "      <td>72</td>\n",
       "      <td>0</td>\n",
       "      <td>3604</td>\n",
       "      <td>302</td>\n",
       "      <td>115</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>i2</th>\n",
       "      <td>3</td>\n",
       "      <td>4</td>\n",
       "      <td>4</td>\n",
       "      <td>4</td>\n",
       "      <td>37279</td>\n",
       "      <td>681</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>c-si2</th>\n",
       "      <td>0</td>\n",
       "      <td>2</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>21686</td>\n",
       "      <td>450</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>i3</th>\n",
       "      <td>0</td>\n",
       "      <td>5</td>\n",
       "      <td>1</td>\n",
       "      <td>14</td>\n",
       "      <td>15</td>\n",
       "      <td>64187</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>c-si3</th>\n",
       "      <td>0</td>\n",
       "      <td>4</td>\n",
       "      <td>0</td>\n",
       "      <td>4</td>\n",
       "      <td>3</td>\n",
       "      <td>31000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>s</th>\n",
       "      <td>2057</td>\n",
       "      <td>2097</td>\n",
       "      <td>2094</td>\n",
       "      <td>2890</td>\n",
       "      <td>2473</td>\n",
       "      <td>2392</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>c-ss</th>\n",
       "      <td>251</td>\n",
       "      <td>292</td>\n",
       "      <td>284</td>\n",
       "      <td>1170</td>\n",
       "      <td>1231</td>\n",
       "      <td>983</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>f</th>\n",
       "      <td>969</td>\n",
       "      <td>2647</td>\n",
       "      <td>2056</td>\n",
       "      <td>5374</td>\n",
       "      <td>6587</td>\n",
       "      <td>7687</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>other</th>\n",
       "      <td>3565</td>\n",
       "      <td>4381</td>\n",
       "      <td>3775</td>\n",
       "      <td>12009</td>\n",
       "      <td>15401</td>\n",
       "      <td>12647</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "        S6.d3  S6.d6  S6.d9  S6.i3  S6.i6  S6.i9\n",
       "d         302    146    188    143    151    183\n",
       "c-sd      267     73     77    148    157    173\n",
       "c-dd       19    322     42     82    135    128\n",
       "c-sdd      11    292     19    109    131    136\n",
       "c-ddd       5    145    307     73     53     65\n",
       "c-sddd      1     78    220     56     93     72\n",
       "i1          2     21      0   4890    297    129\n",
       "c-si1       0     72      0   3604    302    115\n",
       "i2          3      4      4      4  37279    681\n",
       "c-si2       0      2      1      0  21686    450\n",
       "i3          0      5      1     14     15  64187\n",
       "c-si3       0      4      0      4      3  31000\n",
       "s        2057   2097   2094   2890   2473   2392\n",
       "c-ss      251    292    284   1170   1231    983\n",
       "f         969   2647   2056   5374   6587   7687\n",
       "other    3565   4381   3775  12009  15401  12647"
      ]
     },
     "execution_count": 9,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "row_index_protein = [ 'd', 'c-sd', 'c-dd', 'c-sdd', 'c-ddd','c-sddd', 'i1','c-si1', 'i2','c-si2', 'i3','c-si3','s', 'c-ss','f', 'other']\n",
    "df_protein_reads = pd.DataFrame(protein_reads, index=row_index_protein, columns=fractions)\n",
    "df_protein_count = pd.DataFrame(protein_count, index=row_index_protein, columns=fractions)\n",
    "\n",
    "df_protein_count"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Figure 3A (after normalization for baseline from codon ambiguity).\n",
    "\n",
    "Generate data for a histogram of tranposon insertion sites: best used for -3 bp library, it shows how many times a mutation is detected at each DNA position. Spikes corresponds to sites close to tranposon preferred insertion sequence (GC rich). This data still needs to be normalized for whether a position can be detected due to codon ambiguity.\n",
    "\n",
    "InDel placement is inheretnly ambiguous. For example, if the original sequence is GCT TCA and we observe GC- --A, it could just as well come from G-- -CA. Any algorithm will arbitrarily favour one over another. Here, if the true deletions were one of each, the use of Needleman-Wunsch algorith to place deletions will favour placing InDels earlier in the sequence. Hence, G-- -CA will be reported twice and GC- --A zero times. When we are considering the transposon preference, the best we can do is to split the observed count across all positions where it could have come from, equally.\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "metadata": {},
   "outputs": [],
   "source": [
    "def find_transposon_histogram(all_references, background, baseline='baseline', transposon='d3'):\n",
    "    \"\"\"\n",
    "    Find all mutation of a certain type and count where they are in DNA\n",
    "    :return: dict\n",
    "    \"\"\"\n",
    "    hist = defaultdict(int)\n",
    "\n",
    "    for prot_mutation in all_references[background][baseline]:\n",
    "        if prot_mutation is None:\n",
    "            continue\n",
    "        elif len(prot_mutation) <= 2:\n",
    "            for dna_mutation, count in all_references[background][baseline][prot_mutation]['dna'].items():\n",
    "                if classify_dna(dna_mutation) == transposon:\n",
    "                    hist[int(dna_mutation[0][0])] += count\n",
    "\n",
    "    return hist\n",
    "\n",
    "histogram = find_transposon_histogram(all_ref, 'S6', baseline='d3', transposon='d3')\n",
    "\n",
    "df_hist = pd.DataFrame(histogram, index=['raw_insertions'], columns=list(range(1000)))\n",
    "df_hist = df_hist.fillna(0).transpose()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Figure 3B: Number of observed deletions per position"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "metadata": {},
   "outputs": [],
   "source": [
    "def insertion_frequencies(all_references):\n",
    "    \"\"\"\n",
    "    Generate data for number of different insertions per position.\n",
    "    :param all_references:\n",
    "    :return:\n",
    "    \"\"\"\n",
    "    ins_freq = {}  # for each library ('d3', 'i6', etc.) give a dictionary {1: 5 times, 2: 3 times...}\n",
    "    for background in all_references.keys():\n",
    "        for fraction in all_references[background].keys():\n",
    "            ins_freq['.'.join([background,fraction])] = defaultdict(int)\n",
    "            for prot_mutation in all_references[background][fraction].keys():  # this will go through all protein mutations\n",
    "                for dna_mutation, c in all_references[background][fraction][prot_mutation]['dna'].items():\n",
    "                    \n",
    "                    # sometimes a mutations will have a sequencing error / silent substitution elsewhere, which biases the count\n",
    "                    dna_mut_type = classify_dna(dna_mutation)\n",
    "                    prot_mut_type = classify_protein(prot_mutation)\n",
    "                    if (dna_mut_type == fraction) and (prot_mut_type in dna_protein_correspondence(dna_mut_type)):\n",
    "                        ins_freq['.'.join([background,fraction])][int(dna_mutation[0][0])] += 1\n",
    "    return ins_freq\n",
    "\n",
    "ins_freq = insertion_frequencies(all_ref)\n",
    "\n",
    "df_ins_freq = pd.DataFrame(ins_freq, columns=fractions).fillna(0).astype('int32')\n",
    "df_ins_freq['S6.deletions'] = df_ins_freq['S6.d3'] + df_ins_freq['S6.d6'] + df_ins_freq['S6.d9']\n",
    "df_ins_freq.to_csv('20190626_Figure_3BC.csv')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 52,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "array([[<matplotlib.axes._subplots.AxesSubplot object at 0x7f2d60598630>]],\n",
       "      dtype=object)"
      ]
     },
     "execution_count": 52,
     "metadata": {},
     "output_type": "execute_result"
    },
    {
     "data": {
      "image/png": "iVBORw0KGgoAAAANSUhEUgAAAXcAAAEICAYAAACktLTqAAAABHNCSVQICAgIfAhkiAAAAAlwSFlzAAALEgAACxIB0t1+/AAAADh0RVh0U29mdHdhcmUAbWF0cGxvdGxpYiB2ZXJzaW9uMy4xLjAsIGh0dHA6Ly9tYXRwbG90bGliLm9yZy+17YcXAAAUBklEQVR4nO3df5Bd5X3f8fcnCGwVJRJBVAFJtchA0tgocUDFap22K5MmAmLEpNDioTZycZVJnYldyx0rno7TdDoJnoSQOmnDyMFFTl0LBttFBtyEYHbsJIUY+QeCyI4FQ2wBhfBLICCO5X77xz0im2WXvbt7V/fuk/dr5o7OOc9zznm+91x97rln749UFZKktnzXsAcgSRo8w12SGmS4S1KDDHdJapDhLkkNMtwlqUGGu/7WSDKe5B199q0kZ8xxP+9P8jtzWVcaFMNdIyfJjyX54ySHkjyV5I+S/IOu7ZQk/zPJM0meTvKxIY91LMnBicuq6perqq8nEWmhLBn2AKSJknwPcAvws8CNwAnAPwa+1XX5JPAF4DXAC8BZQximNPI8c9eo+QGAqvp4VX2nql6sqt+vqnuT/ASwFvj3VXWoqr5dVV+abkNJ/lmSr3avAH4LyKT2f51kf/cK4PeSvGaa7bwqya8l+UaSx5Jcm2RpkhOBzwCnJTnc3U5L8h+T/I8J61+U5P7u1cZ4kh+a0PZQkvcmubcb5w1JXt21rUxyS7feU0k+n8T/s+qLDxSNmj8DvpNkV5Lzk5w0oW0j8DVgV5Ink3whyT+daiNJVgKfAP4DsBJ4AHjjhPaLgfcDPw2cAnwe+Pg0Y/ogvSed1wNnAKuBD1TV88D5wCNVtay7PTJpHD/Qbffd3X5uAz6d5IQJ3f4FsBk4HfhhYGu3fDtwsFtvVTdevy9EfTHcNVKq6lngx+iF2IeBv0iyJ8kqYA3wE8CdwPcBVwM3d0E+2QXAn1bVTVX1beA3gP87of1ngF+pqv1VdQT4ZeD1k8/ekwT4N8C/q6qnquq5ru9lfZb0L4Fbq+r2bhy/BiwF/tGEPh+qqkeq6ing0/SeRAC+DZwKvKZ7lfL58sug1CfDXSOnC9ytVbWG3jX10+iF84vAQ1V1XRd2u4FvMuGMfILTuraj26yJ8/Su2f+X7pLHM8BT9C7brJ60nVOAvwPsndD3f3fL+3Ea8OcTxvH/unFM3M/EJ50XgGXd9K8CB4DfT/Jgkh197lMy3DXaquqrwPX0Qv5e+r8s8Si96/PAS2fgaye0fxP4mapaMeG2tKr+eNJ2nqD3pPK6Cf2WV9XRAJ5pPI/QeyKZPI6HZyqgqp6rqu1V9f3Am4H3JDlvpvUkMNw1YpL8/STbk6zp5tcCbwHuAj4FnJTkiiTHJbmE3hnwH02xqVuB1yX56SRLgJ+ndynnqGuBX0jyum4/y5NcOnkj3Zn2h4Frkvzdru/qJD/ZdXkMODnJ8mlKuhG4MMl5SY6ndx39W8DkJ5Gp7oufSnJG94TwLPCd7ibNyHDXqHkOeANwd5Ln6YX6fcD27pr0RcB7gUPADmBLVT0B0L2L5VqAbtmlwFXAk8CZTHgSqKpP0ftD6e4kz3b7OH+aMb2P3uWRu7q+fwD8YLedr9L7g+mD3WWb0yauWFVfA/4V8Jv0XgW8GXhzVf1VH/fFmd2+DgP/B/hvVTXex3oS8e8zktQez9wlqUGGuyQ1yHCXpAYZ7pLUoJH44rCVK1fWunXr5rTu888/z4knnjjYAQ2JtYyeVuoAaxlV86ll7969T1TVlB+oG4lwX7duHffcc8+c1h0fH2dsbGywAxoSaxk9rdQB1jKq5lNLkj+frs3LMpLUIMNdkhpkuEtSgwx3SWqQ4S5JDTLcJalBhrskNchwl6QGGe6S1KCR+ISqpJdbt+PWgW5v+/ojbO1jmw9ddeFA96vh8MxdkhpkuEtSgwx3SWqQ4S5JDTLcJalBhrskNchwl6QGGe6S1CDDXZIaZLhLUoMMd0lqkOEuSQ0y3CWpQYa7JDXIcJekBhnuktQgw12SGmS4S1KDDHdJapDhLkkNMtwlqUF9h3uS45J8Kckt3fzpSe5O8vUkNyQ5oVv+qm7+QNe+bmGGLkmazmzO3N8F7J8w/0Hgmqo6E3gauLJbfiXwdFWdAVzT9ZMkHUN9hXuSNcCFwO908wHeBNzUddkFXNxNb+nm6drP6/pLko6RVNXMnZKbgF8Bvht4L7AVuKs7OyfJWuAzVXVWkvuAzVV1sGt7AHhDVT0xaZvbgG0Aq1atOmf37t1zKuDw4cMsW7ZsTuuOGmsZPcOsY9/Dhwa6vVVL4bEXZ+63fvXyge53IbTy+IL51bJp06a9VbVhqrYlM62c5KeAx6tqb5Kxo4un6Fp9tP31gqqdwE6ADRs21NjY2OQufRkfH2eu644aaxk9w6xj645bB7q97euPcPW+Gf/L89DlYwPd70Jo5fEFC1fLzEca3ghclOQC4NXA9wC/AaxIsqSqjgBrgEe6/geBtcDBJEuA5cBTAx+5JGlaM15zr6pfqKo1VbUOuAz4bFVdDtwJXNJ1uwK4uZve083TtX+2+rn2I0kamPm8z/19wHuSHABOBq7rll8HnNwtfw+wY35DlCTNVj+XZV5SVePAeDf9IHDuFH3+Erh0AGOTJM2Rn1CVpAYZ7pLUIMNdkhpkuEtSgwx3SWqQ4S5JDTLcJalBhrskNchwl6QGGe6S1CDDXZIaZLhLUoMMd0lqkOEuSQ0y3CWpQYa7JDXIcJekBhnuktQgw12SGmS4S1KDDHdJapDhLkkNMtwlqUGGuyQ1yHCXpAYZ7pLUIMNdkhpkuEtSgwx3SWqQ4S5JDTLcJalBhrskNchwl6QGGe6S1CDDXZIaZLhLUoNmDPckr07yJ0m+kuT+JL/ULT89yd1Jvp7khiQndMtf1c0f6NrXLWwJkqTJ+jlz/xbwpqr6EeD1wOYkG4EPAtdU1ZnA08CVXf8rgaer6gzgmq6fJOkYmjHcq+dwN3t8dyvgTcBN3fJdwMXd9JZunq79vCQZ2IglSTPq65p7kuOSfBl4HLgdeAB4pqqOdF0OAqu76dXANwG69kPAyYMctCTplaWq+u+crAA+BXwA+O/dpReSrAVuq6r1Se4HfrKqDnZtDwDnVtWTk7a1DdgGsGrVqnN27949pwIOHz7MsmXL5rTuqLGW0TPMOvY9fGig21u1FB57ceZ+61cvH+h+F0Irjy+YXy2bNm3aW1UbpmpbMpsNVdUzScaBjcCKJEu6s/M1wCNdt4PAWuBgkiXAcuCpKba1E9gJsGHDhhobG5vNUF4yPj7OXNcdNdYyeoZZx9Ydtw50e9vXH+HqfTP/l3/o8rGB7nchtPL4goWrpZ93y5zSnbGTZCnw48B+4E7gkq7bFcDN3fSebp6u/bM1m5cHkqR56+fM/VRgV5Lj6D0Z3FhVtyT5U2B3kv8MfAm4rut/HfC7SQ7QO2O/bAHGLUl6BTOGe1XdC/zoFMsfBM6dYvlfApcOZHSSpDnxE6qS1CDDXZIaZLhLUoMMd0lqkOEuSQ0y3CWpQYa7JDXIcJekBhnuktQgw12SGmS4S1KDDHdJapDhLkkNMtwlqUGz+iUmSWrRugH/6tVsXL/5xAXZrmfuktQgw12SGmS4S1KDDHdJapDhLkkNMtwlqUGGuyQ1yPe5a1b6fT/w9vVH2DrA9w4/dNWFA9uW9LeBZ+6S1CDDXZIaZLhLUoMMd0lqkOEuSQ0y3CWpQYa7JDXIcJekBhnuktQgw12SGmS4S1KDDHdJapDhLkkNMtwlqUEzhnuStUnuTLI/yf1J3tUt/94ktyf5evfvSd3yJPlQkgNJ7k1y9kIXIUn6m/o5cz8CbK+qHwI2Au9M8lpgB3BHVZ0J3NHNA5wPnNndtgG/PfBRS5Je0YzhXlWPVtUXu+nngP3AamALsKvrtgu4uJveAny0eu4CViQ5deAjlyRNK1XVf+dkHfA54CzgG1W1YkLb01V1UpJbgKuq6g+75XcA76uqeyZtaxu9M3tWrVp1zu7du+dUwOHDh1m2bNmc1h01i6GWfQ8f6qvfqqXw2IuD2+/61csHt7FZGOYx6fe+7le/x2RY9/VsDPq4DPq+no3Tlx8351o2bdq0t6o2TNXW98/sJVkGfAJ4d1U9m2TarlMse9kzSFXtBHYCbNiwocbGxvodyt8wPj7OXNcdNYuhln5/Om/7+iNcvW9wv+L40OVjA9vWbAzzmAzyZwqh/2MyrPt6NgZ9XAZ9X8/G9ZtPXJDHWF/vlklyPL1g/1hVfbJb/NjRyy3dv493yw8CayesvgZ4ZDDDlST1o593ywS4DthfVb8+oWkPcEU3fQVw84Tlb+veNbMROFRVjw5wzJKkGfTzuvmNwFuBfUm+3C17P3AVcGOSK4FvAJd2bbcBFwAHgBeAtw90xJKkGc0Y7t0fRqe7wH7eFP0LeOc8xyVJmgc/oSpJDTLcJalBhrskNchwl6QGGe6S1CDDXZIaZLhLUoMMd0lqkOEuSQ0y3CWpQYa7JDXIcJekBhnuktQgw12SGmS4S1KDDHdJapDhLkkNMtwlqUGGuyQ1yHCXpAYZ7pLUIMNdkhpkuEtSgwx3SWqQ4S5JDTLcJalBhrskNchwl6QGGe6S1CDDXZIaZLhLUoMMd0lqkOEuSQ0y3CWpQYa7JDXIcJekBs0Y7kk+kuTxJPdNWPa9SW5P8vXu35O65UnyoSQHktyb5OyFHLwkaWr9nLlfD2yetGwHcEdVnQnc0c0DnA+c2d22Ab89mGFKkmZjxnCvqs8BT01avAXY1U3vAi6esPyj1XMXsCLJqYMarCSpP6mqmTsl64Bbquqsbv6Zqloxof3pqjopyS3AVVX1h93yO4D3VdU9U2xzG72ze1atWnXO7t2751TA4cOHWbZs2ZzWHTWLoZZ9Dx/qq9+qpfDYi4Pb7/rVywe3sVkY5jHp977uV7/HZFj39WwM+rgM+r6ejdOXHzfnWjZt2rS3qjZM1bZkXqN6uUyxbMpnj6raCewE2LBhQ42Njc1ph+Pj48x13VGzGGrZuuPWvvptX3+Eq/cN7uH10OVjA9vWbAzzmPR7X/er32MyrPt6NgZ9XAZ9X8/G9ZtPXJDH2FzfLfPY0cst3b+Pd8sPAmsn9FsDPDL34UmS5mKu4b4HuKKbvgK4ecLyt3XvmtkIHKqqR+c5RknSLM34Gi3Jx4ExYGWSg8AvAlcBNya5EvgGcGnX/TbgAuAA8ALw9gUYsyRpBjOGe1W9ZZqm86boW8A75zsoSdL8+AlVSWqQ4S5JDTLcJalBhrskNchwl6QGGe6S1CDDXZIaZLhLUoMMd0lqkOEuSQ0y3CWpQYa7JDXIcJekBhnuktQgw12SGmS4S1KDDHdJatDgfp5+SPY9fGhov1z+0FUXDmW/kjQTz9wlqUGGuyQ1yHCXpAYZ7pLUIMNdkhpkuEtSgwx3SWqQ4S5JDTLcJalBhrskNchwl6QGGe6S1CDDXZIaZLhLUoMMd0lqkOEuSQ0y3CWpQYa7JDXIcJekBi1IuCfZnORrSQ4k2bEQ+5AkTW/g4Z7kOOC/AucDrwXekuS1g96PJGl6C3Hmfi5woKoerKq/AnYDWxZgP5KkaaSqBrvB5BJgc1W9o5t/K/CGqvq5Sf22Adu62R8EvjbHXa4EnpjjuqPGWkZPK3WAtYyq+dTymqo6ZaqGJXMfz7QyxbKXPYNU1U5g57x3ltxTVRvmu51RYC2jp5U6wFpG1ULVshCXZQ4CayfMrwEeWYD9SJKmsRDh/gXgzCSnJzkBuAzYswD7kSRNY+CXZarqSJKfA34POA74SFXdP+j9TDDvSzsjxFpGTyt1gLWMqgWpZeB/UJUkDZ+fUJWkBhnuktSgRRPuM32lQZJXJbmha787ybpjP8r+9FHL1iR/keTL3e0dwxjnTJJ8JMnjSe6bpj1JPtTVeW+Ss4/1GPvVRy1jSQ5NOCYfONZj7EeStUnuTLI/yf1J3jVFn0VxXPqsZbEcl1cn+ZMkX+lq+aUp+gw2w6pq5G/0/jD7APD9wAnAV4DXTurzb4Fru+nLgBuGPe551LIV+K1hj7WPWv4JcDZw3zTtFwCfoffZh43A3cMe8zxqGQNuGfY4+6jjVODsbvq7gT+b4vG1KI5Ln7UsluMSYFk3fTxwN7BxUp+BZthiOXPv5ysNtgC7uumbgPOSTPWBqmFr5usZqupzwFOv0GUL8NHquQtYkeTUYzO62emjlkWhqh6tqi92088B+4HVk7otiuPSZy2LQndfH+5mj+9uk9/NMtAMWyzhvhr45oT5g7z8IL/Up6qOAIeAk4/J6Gann1oA/nn3kvmmJGunaF8M+q11sfiH3cvqzyR53bAHM5PuZf2P0jtLnGjRHZdXqAUWyXFJclySLwOPA7dX1bTHZRAZtljCvZ+vNOjraw9GQD/j/DSwrqp+GPgD/vrZfLFZLMekH1+k9z0ePwL8JvC/hjyeV5RkGfAJ4N1V9ezk5ilWGdnjMkMti+a4VNV3qur19D61f26SsyZ1GehxWSzh3s9XGrzUJ8kSYDmj+TJ7xlqq6smq+lY3+2HgnGM0tkFr5qsoqurZoy+rq+o24PgkK4c8rCklOZ5eGH6sqj45RZdFc1xmqmUxHZejquoZYBzYPKlpoBm2WMK9n6802ANc0U1fAny2ur9MjJgZa5l0/fMietcaF6M9wNu6d2dsBA5V1aPDHtRcJPm+o9c/k5xL7//Ok8Md1ct1Y7wO2F9Vvz5Nt0VxXPqpZREdl1OSrOimlwI/Dnx1UreBZthCfCvkwNU0X2mQ5D8B91TVHnoPgt9NcoDes91lwxvx9Pqs5eeTXAQcoVfL1qEN+BUk+Ti9dyusTHIQ+EV6fyiiqq4FbqP3zowDwAvA24cz0pn1UcslwM8mOQK8CFw2oicPbwTeCuzrru8CvB/4e7Dojks/tSyW43IqsCu9HzP6LuDGqrplITPMrx+QpAYtlssykqRZMNwlqUGGuyQ1yHCXpAYZ7pLUIMNdkhpkuEtSg/4/NuEevogoXGgAAAAASUVORK5CYII=\n",
      "text/plain": [
       "<Figure size 432x288 with 1 Axes>"
      ]
     },
     "metadata": {
      "needs_background": "light"
     },
     "output_type": "display_data"
    }
   ],
   "source": [
    "df_ins_freq.hist('S6.deletions')"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Figure ? - not currently in the paper\n",
    "Reverse the data: how many mutations occer once, twice, three times etc."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 199,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>S6.d3</th>\n",
       "      <th>S6.d6</th>\n",
       "      <th>S6.d9</th>\n",
       "      <th>S6.i3</th>\n",
       "      <th>S6.i6</th>\n",
       "      <th>S6.i9</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>30</td>\n",
       "      <td>30</td>\n",
       "      <td>30</td>\n",
       "      <td>5508</td>\n",
       "      <td>80377</td>\n",
       "      <td>92496</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>17</td>\n",
       "      <td>35</td>\n",
       "      <td>28</td>\n",
       "      <td>3351</td>\n",
       "      <td>17777</td>\n",
       "      <td>8381</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>23</td>\n",
       "      <td>23</td>\n",
       "      <td>16</td>\n",
       "      <td>2257</td>\n",
       "      <td>5013</td>\n",
       "      <td>1165</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>13</td>\n",
       "      <td>19</td>\n",
       "      <td>22</td>\n",
       "      <td>1749</td>\n",
       "      <td>1600</td>\n",
       "      <td>253</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>10</td>\n",
       "      <td>12</td>\n",
       "      <td>10</td>\n",
       "      <td>1272</td>\n",
       "      <td>569</td>\n",
       "      <td>89</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "   S6.d3  S6.d6  S6.d9  S6.i3  S6.i6  S6.i9\n",
       "1     30     30     30   5508  80377  92496\n",
       "2     17     35     28   3351  17777   8381\n",
       "3     23     23     16   2257   5013   1165\n",
       "4     13     19     22   1749   1600    253\n",
       "5     10     12     10   1272    569     89"
      ]
     },
     "execution_count": 199,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "def dna_mutation_frequencies(all_references):\n",
    "    \"\"\"\n",
    "    Generate data for a histogram of DNA mutation frequencies - how many occur once, twice, ...\n",
    "    :param all_references:\n",
    "    :return:\n",
    "    \"\"\"\n",
    "    freq = {}  # for each library ('d3', 'i6', etc.) give a dictionary {1: 5 times, 2: 3 times...}\n",
    "    for background in all_references.keys():\n",
    "        for fraction in all_references[background].keys():\n",
    "            freq['.'.join([background,fraction])] = defaultdict(int)\n",
    "            for prot_mutation in all_references[background][fraction].keys():\n",
    "                for dna_mutation, c in all_references[background][fraction][prot_mutation]['dna'].items():\n",
    "                    dna_mut_type = classify_dna(dna_mutation)\n",
    "                    prot_mut_type = classify_protein(prot_mutation)\n",
    "                    if (dna_mut_type == fraction) and (prot_mut_type in dna_protein_correspondence(dna_mut_type)):\n",
    "                        freq['.'.join([background,fraction])][c] += 1\n",
    "    return freq\n",
    "\n",
    "hist_Poisson = dna_mutation_frequencies(all_ref)\n",
    "df_hist_Poisson = pd.DataFrame(hist_Poisson, columns=fractions).fillna(0).astype('int32')\n",
    "df_hist_Poisson.head()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "index = number of times a mutation is observed\n",
    "column = number of such mutations\n",
    "\n",
    "This still needs manual clipping."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 230,
   "metadata": {},
   "outputs": [],
   "source": [
    "def plot_S7_Poisson(library, cutoff,title):\n",
    "    hist_data = df_hist_Poisson[library].iloc[:cutoff + 1]\n",
    "    main_max = hist_data.max()\n",
    "    after_trunc = df_hist_Poisson[library].iloc[cutoff:].sum()\n",
    "\n",
    "    hist_data.iloc[cutoff] = after_trunc # that has the data with the last entry in series being cumulative\n",
    "\n",
    "    \n",
    "    f, (ax, ax2) = plt.subplots(2, 1, sharex=True, gridspec_kw={'height_ratios': [1, 4]}, figsize=(5,5))\n",
    "\n",
    "    f.suptitle(title)\n",
    "    \n",
    "    bins = list(range(1, cutoff+1)) + [cutoff + 5]\n",
    "\n",
    "    ax.bar(bins, hist_data) # first arguement is x-position, second is height\n",
    "    ax2.bar(bins,hist_data)\n",
    "\n",
    "    # zoom-in / limit the view to different portions of the data\n",
    "    ax.set_ylim(after_trunc - 2, after_trunc + 2)  # outliers only\n",
    "    ax2.set_ylim(0, main_max + 1)  # most of the data\n",
    "    \n",
    "    ax2.set_ylabel('Count (X)')\n",
    "    ax2.set_xlabel('Number of mutations in library observed X-times')\n",
    "\n",
    "    # hide the spines between ax and ax2\n",
    "    ax.spines['bottom'].set_visible(False)\n",
    "    ax2.spines['top'].set_visible(False)\n",
    "\n",
    "    # set the tick label for overflow bin\n",
    "    xlabels_pos = [1,5,10,20,30,40,50,55]\n",
    "    xlabels = [str(pos) for pos in xlabels_pos]\n",
    "    xlabels[-1] = str(cutoff) + '+'\n",
    "    ax2.xaxis.set_ticks(xlabels_pos)\n",
    "    ax2.set_xticklabels(xlabels)\n",
    "    # hide the axes labels on top\n",
    "    ax.axes.get_xaxis().set_visible(False) \n",
    "\n",
    "\n",
    "    # This looks pretty good, and was fairly painless, but you can get that\n",
    "    # cut-out diagonal lines look with just a bit more work. The important\n",
    "    # thing to know here is that in axes coordinates, which are always\n",
    "    # between 0-1, spine endpoints are at these locations (0,0), (0,1),\n",
    "    # (1,0), and (1,1).  Thus, we just need to put the diagonals in the\n",
    "    # appropriate corners of each of our axes, and so long as we use the\n",
    "    # right transform and disable clipping.\n",
    "\n",
    "    d = .012  # how big to make the diagonal lines in axes coordinates\n",
    "    # arguments to pass to plot, just so we don't keep repeating them\n",
    "    kwargs = dict(transform=ax.transAxes, color='k', clip_on=False)\n",
    "    ax.plot((-d, +d), (-4*d, +4*d), **kwargs)        # top-left diagonal\n",
    "    ax.plot((1 - d, 1 + d), (-4*d, +4*d), **kwargs)  # top-right diagonal\n",
    "\n",
    "    kwargs.update(transform=ax2.transAxes)  # switch to the bottom axes\n",
    "    ax2.plot((-d, +d), (1 - d, 1 + d), **kwargs)  # bottom-left diagonal\n",
    "    ax2.plot((1 - d, 1 + d), (1 - d, 1 + d), **kwargs)  # bottom-right diagonal\n",
    "\n",
    "    f.savefig('FigS7_' + library +'.png', format='png', dpi=600)\n",
    "    plt.show()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 233,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "image/png": "iVBORw0KGgoAAAANSUhEUgAAAUoAAAFhCAYAAAAWQuPwAAAABHNCSVQICAgIfAhkiAAAAAlwSFlzAAALEgAACxIB0t1+/AAAADh0RVh0U29mdHdhcmUAbWF0cGxvdGxpYiB2ZXJzaW9uMy4xLjAsIGh0dHA6Ly9tYXRwbG90bGliLm9yZy+17YcXAAAfx0lEQVR4nO3deZRcZZ3G8e9DiKDskQYhEIMaNgWCNosCDiAgCLIIjBMxwogEFwQ3nKgIYZxxQEUQ1xMWA5JBUQmETQghEBcmkkBIAmFRJigmQ8ImIBoI+c0f920omup6q7vrVlV3P59z+tStu/6q6tbTd31LEYGZmfVsjVYXYGbW7hyUZmYZDkozswwHpZlZhoPSzCzDQWlmluGgtAFPUkh6S6vrsMHLQWltR9Jaks6VtFTSk5J+IGl4q+uyoctBae1oItAJvA3YGng7cFpLK7IhzUFp7ej9wPkR8URErADOBz6ameZ9kh6S9Jikb0paA0DScZJ+K+m7kv4q6T5J7yn7Bdjg4qC0dqT0V/l8C0kb1JjmCIqt0LcDh/HKYN0NeAjYGDgDuFLSiIZWbIOag9La0Q3AKZI6JL0BODn1f12Nac5OW6B/As4DxlUMWw6cFxEvRMTPgPuBg8so3AYnB6W1lKRjJD2b/m5Ivf8TuAuYD/wOuAp4gSLwevLniu6Hgc0rnv8lXtn6S/fhZjU5KK2lImJqRKyb/g5K/f4eESdFxMiIeBPwODAvIl6sMastK7pHAUsrno+UpBrDzWpyUFrbkTRS0uYq7A58leLYYi2nStpI0pbAKcDPKoZtApwsabiko4HtgOtLKd4GpTVbXYBZFW8GLqUIuD8DEyPipsw0VwPzgA2AKcBFFcPmAGOAx4BHgaMi4vEG12yDmNxwrw1mko4DPhYRe7a6Fhu4vOttZpbhoDQzy/Cut5lZhrcozcwyHJRmZhkOSjOzDAelmVmGg9LMLMNBaWaW4aA0M8twUJqZZTgozcwyHJRmZhkOSjOzDAelmVmGg9LMLMNBaWaW4aA0M8twUJqZZTgozcwyHJRmZhn9DkpJF0taLmlRRb+dJN0uaaGkayStXzFsxzTsnjR87f7WYGZWpkZsUU4BDuzW70KK32LeAZgGnAogaU3gMuDjEfFWYG/ghQbUYGZWmn4HZUTMBp7o1nsbYHbqngEcmboPABZExN1p2scj4sX+1mBmVqayjlEuAg5N3UcDW6burYGQdKOkOyV9saTlm5k1zJolzfejwPmSTgemA89XLG9PYBfgOWCmpHkRMbP7DCRNACYArLPOOu/YdtttSyrVzIaqefPmPRYRHbnxSgnKiLiPYjcbSVsDB6dBjwC3RcRjadj1wNuBVwVlREwGJgN0dnbG3LlzyyjVzIYwSQ/XM14pu96SNkmPawCnAT9Kg24EdpT0unRi55+Ae8uowcysURpxedDlwO3ANpIekXQ8ME7SA8B9wFLgxwAR8STwbeAOYD5wZ0Rc198azMzK1O9d74gY18Og7/Qw/mUUlwiZmQ0IvjPHzCzDQWlmluGgNDPLcFCamWU4KM3MMhyUZmYZDkozswwHpZlZhoPSzCzDQWlmluGgNDPLcFCamWU4KM3MMhyUZmYZDkozswwHpZlZRiNaOL9Y0nJJiyr67STpdkkLJV0jaf1u04yS9KykL/R3+WZmZWvEFuUU4MBu/S4EJkbEDsA04NRuw88FbmjAss3MStfvoIyI2cAT3XpvA8xO3TOAI7sGSDoceAi4p7/LNjNrhrKOUS4CDk3dRwNbAkhaB/g34MySlmtm1nCl/K438FHgfEmnA9OB51P/M4FzI+JZSTVnIGkCMAFg1KhRJZVpZgPZ6Ikv/4jrkrMOLm05pQRlRNwHHAAgaWug6xXsBhwl6RvAhsBqSf+IiO9VmcdkYDJAZ2dnlFGnmVk9SglKSZtExHJJawCnAT8CiIi9KsaZBDxbLSTNzNpJIy4Puhy4HdhG0iOSjgfGSXoAuA9YCvy4v8sxM2uVfm9RRsS4HgZ9JzPdpP4u28ysGXxnjplZhoPSzCzDQWlmluGgNDPLcFCamWU4KM3MMhyUZmYZDkozswwHpZlZhoPSzCzDQWlmluGgNDPLcFCamWU4KM3MMhyUZmYZDkozs4xGtHB+saTlkhZV9NtJ0u2SFkq6RtL6qf/+kual/vMk7dubZV111VXsv//+rFy5sr9lm5m9RNLhkmZIWqva8EZsUU4BDuzW70JgYkTsAEwDTk39HwPen/ofC/ykngWsXr2aE044gSOOOIInnniCxx9/vAFlm5mxhqQLKHJqBPD6qiP1dykRMRt4olvvbYDZqXsGcGQa966IWJr63wOs3VOCd5G0y7333stFF13ExIkTuf3229l88837W7aZDXF33HEHwPbA8cBZwDsr8ukVFNH/X4KVNBq4NiLelp7/Djg7Iq6W9DngzIhYr9s0RwEfj4j9epjnicBXgc2BVcD+EXFbv4s1syFN0jBgIjCJIlsOzGVLWSdzPgp8StI8YD3g+cqBkt4KnA2cWG1iSW8EjgFGAlcAmzokzay/UrbMAv4D+CWweT3ZUsrvekfEfcABqbCtgYMrCt2C4njARyLij92nlbQzcCsg4CPAZdGIzV4zG9L6ky2lbFFK2iQ9rgGcBvwoPd8QuA74UkT8tofJ7wEuB3aKiJ84JM2sQfqcLf0+RinpcmBvYGPgUeAMYF3gU2mUKymCMSSdBnwJeLBiFgdExPJ+FWFmVqKGnMwxMxvMfGeOmVmGg9LMLMNBaWaW4aA0M8twUJqZZTgozcwyHJRmZhkOSjOzDAelmVmGg9LMLMNBaWaW4aA0M8twUJqZZTgozcwyHJRmZhkDLiglbdrqGsxs8KmVLQMmKCWtI2kysFjSyFbXY2aDQz3ZUsqPizWapF2AqcBbKH69cUVrKzKzwaDebGnrLUpJwyR9BfgdsDawb0R8KSKez0xqZtaj3mZL2/5mTvr93Z8Ae73hDW/g3nvvZaONNmp1WWbWRkZPvO6l7iVnHVxjzJc9/PDDjB8/nl//+tcA/wdsHxFP1pqmLbcoJR0DLADGAh8ZOXKkQ9LM+m3q1KnsuOOOzJ8/n0svvRTgL7mQhDYMSkm7AZcBC0m/v9viksxsEJgzZw4f/vCH2WGHHbj77rsZP3583dOWdjJH0trAbGCttJxfRMQZkrYCfgqMAO4ExlceF4iIOZIOBW6IiFVl1WdmQ8tuu+3G9OnTOeigg1hzzd5FX5lblCspDpDuRLELfaCk3SnOLJ0bEWOAJ4Hju08YEdc4JM2s0d7//vf3OiShxKCMwrPp6fD0F8C+wC9S/0uAw8uqwcysEUo9RplOwc8HlgMzgD8CT1VsLT4C+OJxM2trpV5wHhEvAmMlbQhMA7arNlq1aSVNACYAjBo1qtfL7stlA2Zm1TTlrHdEPAXcCuwObCipK6C3AJb2MM3kiOiMiM6Ojo5mlGlmVlVpQSmpI21JIum1wH7AYmAWcFQa7Vjg6rJqMDNrhDJ3vTcDLpE0jCKQr4iIayXdC/xU0n8AdwEXlViDmVm/lRaUEbEA2LlK/4eAXctarplZo7XdnTlmZu3GQWlmluGgNDPLcFCamWU4KM3MMhyUZmYZDkozs4wB8eNijVB57zf4/m8zq5+3KM3MMhyUZmYZDkozswwHpZlZhoPSzCzDQWlmluGgNDPLcFCamWU4KM3MMhyUZmYZDkozswwHpZlZhoPSzCzDQWlmluGgNDPLKC0oJW0paZakxZLukXRK6j9J0l8kzU9/7yurBjOzRiiz4d5VwOcj4k5J6wHzJM1Iw86NiG+VuGwzs4YpLSgjYhmwLHU/I2kxMLKs5ZmZlaUpxygljQZ2BuakXidJWiDpYkkb9TDNBElzJc1dsWJFM8o0M6uq9KCUtC7wS+AzEfE08EPgzcBYii3Oc6pNFxGTI6IzIjo7OjrKLtPMrEelBqWk4RQhOTUirgSIiEcj4sWIWA1cAOxaZg1mZv1V5llvARcBiyPi2xX9N6sY7QhgUVk1mJk1QplnvfcAxgMLJc1P/b4MjJM0FghgCXBiiTWYmfVbmWe9fwOoyqDry1qmmVkZfGeOmVmGg9LMLMNBaWaW4aA0M8twUJqZZTgozcwyHJRmZhkOSjOzDAelmVmGg9LMLMNBaWaW4aA0M8twUJqZZTgozcwyymyPckAaPfG6l7qXnHVwCysxs3bhLUozswwHpZlZhoPSzCzDQWlmluGgNDPLcFCamWU4KM3MMkoLSklbSpolabGkeySdkvqPkDRD0oPpcaOyajAza4QytyhXAZ+PiO2A3YFPSdoemAjMjIgxwMz03MysbZUWlBGxLCLuTN3PAIuBkcBhwCVptEuAw8uqwcysEZpyjFLSaGBnYA6waUQsgyJMgU2aUYOZWV+VHpSS1gV+CXwmIp7uxXQTJM2VNHfFihXlFWhmllFqUEoaThGSUyPiytT7UUmbpeGbAcurTRsRkyOiMyI6Ozo6yizTzKymMs96C7gIWBwR364YNB04NnUfC1xdVg1mZo1QZjNrewDjgYWS5qd+XwbOAq6QdDzwJ+DoEmswM+u30oIyIn4DqIfB7ylruWZmjeY7c8zMMhyUZmYZde16S9qE4pjj5sDfgUXA3IhYXWJtbcs/F2E2tNQMSkn7UNxiOAK4i+JSnrUp7qZ5s6RfAOf05vpIM7OBJrdF+T7ghIj4U/cBktYEDgH2p7hW0sxsUKoZlBFxao1hq4CrGl6RmVmbqXkyR9J3Ja1Xpf+2km4urywzs/aRO+v9f8B8SR8CkPQ6Sd+guLvm+2UXZ2bWDmoGZUT8J7AfcIyk2cACinYmx0bEtCbUZ2bWcvVcRxnpcc00/uKIeK68kszM2kvuGOVpwM3ApRHxLmAv4DBJt6XWys3MBr3c5UEdwM6phXIi4i/AUZIOorgkaLuS6zMza7ncMcpTukKyW/8bgLGlVWVm1kayu96SRlQbFhErJe0r6ZBySjMzaw+5Xe+FwDWS/gHcCayguIVxDMUW5c3A10ut0MysxXJ35lwNXC1pDEWjGJsBTwOXARMi4u/ll2hm1lp1tR4UEQ8CD5Zci5lZW3J7lGZmGQ5KM7OMuoJS0h719DMzG4zq3aL8bp39zMwGnVwL5+8E3gV0SPpcxaD1gWFlFtYMlT/pAM3/WYdWL9/M6pM76/0aYN00XmW7lE8DR5VVlJlZO8ldR3kbcJukKRHxcG9mLOliip+KWB4Rb0v9JgEnUFy4DvDliLi+11WbmTVRXddRAmtJmgyMrpwmIvatMc0U4HvApd36nxsR3+pFjWZmLVVvUP4c+BFwIfBiPRNExGxJo/tWlplZ+6g3KFdFxA8btMyTJH0EmAt8PiKebNB8zcxKUW9QXiPpk8A0YGVXz4h4opfL+yHwNYpW078GnAN8tNqIkiYAEwBGjRrVy8U0V71nryvH8xlus4Gj3qA8Nj1W/nxtAG/qzcIi4tGubkkXANfWGHcyMBmgs7MzehrPzKxs9TaKsVUjFiZps4hYlp4eASxqxHzNzMpUV1CmY4qvEhHdz2hXTnM5sDewsaRHgDOAvSWNpdgaXQKc2Mt6zcyart5d710qutcG3kPRkG+PQRkR46r0vqj+0szM2kO9u96frnwuaQPgJ6VUZGbWZvrazNpzFD8HYWY26NV7jPIaiuOKUDSGsR1wRVlFWW1uTMOsueo9Rll5y+Eq4OGIeKSEeszM2k5du96pcYz7KFoQ2gh4vsyizMzaSb0tnP8z8HvgaOCfgTmS3MyamQ0J9e56fwXYJSKWA0jqoPhN71+UVZiZWbuo96z3Gl0hmTzei2nNzAa0ercofyXpRuDy9PyDwA3llGTduTENs9aq94LzUyV9ANgTEDA5IqaVWpmZWZvI/bjYW4BNI+K3EXElcGXq/25Jb46IPzajSDOzVsodZzwPeKZK/+fSMDOzQS8XlKMjYkH3nhExl+L3c8zMBr1cUK5dY9hrG1mImVm7yp3MuUPSCRFxQWVPSccD88ory9qJ7y23oS4XlJ8Bpkk6hpeDsRN4DUUL5WZmg17NoEy/cfMuSfsAb0u9r4uIW0qvzMysTdR7HeUsYFbJtZiZtSXfhmhmluGgNDPLcFCamWXU2yiG9ZIvqTEbPLxFaWaWUVpQSrpY0nJJiyr6jZA0Q9KD6XGjspZvZtYoZW5RTgEO7NZvIjAzIsYAM9NzM7O2VlpQRsRs4IluvQ8DLkndlwCHl7V8M7NGafYxyk0jYhlAetykycs3M+u1tj3rLWkCMAFg1KhRLatjsJ+99s9MmOU1e4vyUUmbAaTH5T2NGBGTI6IzIjo7OjqaVqCZWXfNDsrpwLGp+1jg6iYv38ys18q8POhy4HZgG0mPpDYszwL2l/QgsH96bmbW1ko7RhkR43oY9J6ylmlmVgbfmWNmltG2Z72HqsF8lr2vr60d35N2rMnK4y1KM7MMB6WZWYaD0swsw0FpZpbhoDQzy3BQmpll+PKgQarsS3HcmIYNJd6iNDPLcFCamWU4KM3MMhyUZmYZDkozswyf9R5CymzIYTA1EjGYXos1hrcozcwyHJRmZhkOSjOzDAelmVmGg9LMLMNBaWaW4cuDBonul7QMFI28FKcvDXWU3QjIUGs8ZLC+Xm9RmplltGSLUtIS4BngRWBVRHS2og4zs3q0ctd7n4h4rIXLNzOri3e9zcwyWhWUAdwkaZ6kCS2qwcysLq3a9d4jIpZK2gSYIem+iJhdOUIK0AkAo0aNakWN1k9lnolv1E9dDBT1vN52bMyjHWvqi5ZsUUbE0vS4HJgG7FplnMkR0RkRnR0dHc0u0czsJU0PSknrSFqvqxs4AFjU7DrMzOrVil3vTYFpkrqW/98R8asW1GFmVpemB2VEPATs1Ozlmpn1lS8PMjPL8L3eZg0yUM9MW563KM3MMhyUZmYZDkozswwHpZlZhoPSzCzDQWlmluHLg6ztNOrnBAbKpTjV6mzmTyoMlPeplbxFaWaW4aA0M8twUJqZZTgozcwyHJRmZhk+6202iLX6jPpg4S1KM7MMB6WZWYaD0swsw0FpZpbhoDQzy/BZbzN7lXrOlg+le8S9RWlmluGgNDPLaElQSjpQ0v2S/iBpYitqMDOrV9ODUtIw4PvAQcD2wDhJ2ze7DjOzerVii3JX4A8R8VBEPA/8FDisBXWYmdWlFUE5EvhzxfNHUj8zs7akiGjuAqWjgfdGxMfS8/HArhHx6W7jTQAmpKfbAPf3YXEbA4/1o9yy5tXMeZdpoNYNA7d2193Yeb8xIjpyI7XiOspHgC0rnm8BLO0+UkRMBib3Z0GS5kZEZ3/mUca8mjnvMg3UumHg1u66mzvvLq3Y9b4DGCNpK0mvAf4FmN6COszM6tL0LcqIWCXpJOBGYBhwcUTc0+w6zMzq1ZJbGCPieuD6JiyqX7vuJc6rmfMu00CtGwZu7a67ufMGWnAyx8xsoPEtjGZmGYMyKCVdLGm5pEUNmt8SSQslzZc0t9G1SRohaYakB9PjRv2vurEkbSlplqTFku6RdErq39a1S1pb0u8l3Z3qPjP130rSnFT3z9KJxbYjaZikuyRdm54PlLpf9Z3p67rSyHn11aAMSmAKcGCD57lPRIxtwGUIU3h1bROBmRExBpiZnrebVcDnI2I7YHfgU+nW03avfSWwb0TsBIwFDpS0O3A2cG6q+0ng+BbWWMspwOKK5wOlbnj1d6Y/60qv5yVpiqS9+17+ywZlUEbEbOCJVtdRTQ+1HQZckrovAQ5valF1iIhlEXFn6n6G4ss7kjavPQrPpqfD018A+wK/SP3brm4ASVsABwMXpudiANRdQyPXlaaud4MyKEsQwE2S5qU7hhpt04hYBkUgAZuUsIyGkTQa2BmYwwCoPe2+zgeWAzOAPwJPRcSqNEq73kZ7HvBFYHV6/noGRt1Q/TvT13WlkfPqE7dwXp89ImKppE2AGZLuS1uGQ46kdYFfAp+JiKeLjZz2FhEvAmMlbQhMA7arNlpzq6pN0iHA8oiYV7H7WO3Nbqu6K7zqO9OMeUl6L8XhCYBRwJ6SngVWRsRufS3AW5R1iIil6XE5xRdt1wYv4lFJmwGkx+UNnn9DSBpOEZJTI+LK1HtA1A4QEU8Bt1IcY91QUteGQtXbaFtsD+BQSUsoWtjal2ILs93rBnr8zvRpXenNvCLixnQscyzFHX8fS8/7HJLgoMyStI6k9bq6gQOAhpxNrzAdODZ1Hwtc3eD591s6PnYRsDgivl0xqK1rl9SRtiSR9FpgP4rjq7OAo9JobVd3RHwpIraIiNEUt/neEhHH0OZ1Q83vTK/XlUbOq18iYtD9AZcDy4AXKI7jHN+Peb0JuDv93QN8pdG1URx7mgk8mB5HtPo9rFL3nhS7eQuA+envfe1eO7AjcFeqexFwesXn+nvgD8DPgbVaXWuN17A3cO1Aqbun70xf1pX+zIviCpO9G/GafGeOmVmGd73NzDIclGZmGQ5KM7MMB6WZWYaD0swsY0gFpaSQdE7F8y9ImtSgeU+RdFR+zH4v5+jUgs+sBs93tKQP9XY8SZ2Szm9kLd2W9++S9uvF+HtXtLRzqKSJqbspn0+90t0ibaWn90jSdBU/Atj1/AJJp1YZr6nrRjMNqaCkaEnmA5I2bnUhlSQN68XoxwOfjIh9GlzGaCAblN3Hi4i5EXFyg2t5SUScHhE393Ha6RFxVr3jV9zxkhtPktrqu1Nv7X10MvDvkjaU9C5gN4q7hLobTRPXjWZqqw+7CVZRNBv/2e4Duv837fqPn7ZQbpN0haQHJJ0l6RgVbRwulPTmitnsJ+nXabxD0vTDJH1T0h2SFkg6sWK+syT9N7CwSj3j0vwXSTo79Tud4sLvH0n6Zrfx66qzp9cJnAXspaLNv8+mrYNfS7oz/b2rh/Eqt+BGSLoqvc7/kbRj6j9JRTuct0p6SNLJqf86kq5T0VbkIkkfrPW5qGiX8MxUz0JJ21b/mF+a9jhJ38t8PsdJ+rmkaygaXlhX0syKZRyWxhutYkv+B8CdwFclnVuxrBMkVd6x1OPnWDHsnLScmZI6Ur+TJd2b3sOfVrxPF6d16K6KmrrX/jNJ7+v23h1ZYx2UpO+l5V1HDw1LRMQSiu/NN4AfACdFxAtVRq21bkySdImkm9Ln+AFJ30jvza9U3B6LpHek9XiepBv18m2Kr3pfmqrVV/E3+Y6BZ4H1gSXABsAXgEkVV/EfVTluetwbeArYDFgL+AtwZhp2CnBexfS/ovjnM4birpu1KX6b/LQ0zlrAXGCrNN+/AVtVqXNz4E9AB0XDJbcAh6dhtwKdVabpTZ09vc5rK/q/Dlg7dY8B5vYw3kvPge8CZ6TufYH5qXsS8LtU18bA4xTNnR0JXFAxrw16uLviqNS9BPh06v4kcGEP70NXPccB38t8Psel7hFpvDWB9VP3xhR3wIhia2k1sHsatg5FK0TD0/PfATv04nMM4JjUfXpFnUtJd9sAG6bHrwMf7uoHPJCW3732I4BLUvdrgD8Dr6XndfADFK0pDUu1PkXFutHttQxPr2Vqje9XrXVjEvCbNJ+dgOeAg9KwaRTNpA1P72NH6v9Bih8frPq+NPNvqG1REhFPA5dS7E7U644o2mNcSfHluCn1X0jxBepyRUSsjogHgYeAbSnuTf2Iima+5lDcejUmjf/7iPjfKsvbBbg1IlZE0aTWVODdDayzHsOBCyQtpLhVbvs6ptkT+AlARNwCvF7SBmnYdRGxMiIeo2jAYNNU136Szpa0V0T8tY5ldDXGMY/ev6Zqnw/AjIjoaiNUwNclLQBupmjGbNM07OGI+J/0+v5GEXyHpC3b4RHRfc+g1ue4GvhZ6r6M4r2D4lbLqZI+TLEHBMU6NDGtQ7dSBPyoKrXfAOwraS3gIGB2RPydntfBdwOXR8SLUTQ8cUuN927H9N5sq74fdrghii3RhRTh/KvUv2v93AZ4G0ULQfOB0yga/oDq70vTDLmgTM6jONa3TkW/VaT3Q5Io/iN3WVnRvbri+Wpe2VRd9/tBg2Ll+nSkFk0iYquI6Aqwv/VQX1/bLqunzlqvs9JngUcp/vt31hivUq1mwCprexFYMyIeAN5B8UX5LxWHFnK65vMivW8msNrnA6/8HI6h2AJ8RxQt0DxKEUzdx4OiQd3jgH8Fflxleb35HLtqORj4PsX7Mk/FsUcBR1asQ6MioqvV85dqioh/UATpeym2xrp2UWutg9l7mFMw/gAYT3Fv9SdS/93SbvZ8SYfW8RpXpjpXAy9E2jzk5fVTwD0Vde4QEQfUeF+aZkgGZfoPfAWvbEZ/CcWHAEXrycP7MOujJa2h4njgm4D7KX6//BMVx2C2VtEKSi1zgH+StLGKEz3jgNv6UE81S6j+Op8B1qsYbwNgWVqpx1NsAVQbr9JsiqBBRRuKj6Ut+KokbQ48FxGXAd8C3t7L19Jb1T6f7jagaAfyBUn7AG/saWYRMQfYkuIExuVVRqn1Oa7By60AfQj4TQqkLSNiFkWDvRsC61KsQ59O/9iQtHON1/hTiuDeK00HPa+Ds4F/SccwNwN6OkF4IvBgRNwKfA74oqSOiJhTEWrTqb1u1ON+oEPSO1OdwyW9tcb70jRDueHec4CTKp5fAFwt6fcUrZH0tLVXy/0UX4RNgY9HxD8kXUixW3FnWtFXkGm2PiKWSfoSRZNaAq6PiEY1I9XT61wArJJ0N8XxvB8Av5R0dKqjp/Huqpj3JODHabf1OV5uBqsnOwDflLSaojWlT/TrleVV+3y6jzMVuEbFj1jNB3INzl4BjI2IJ7sPyHyOfwPeKmke8FeKLcBhwGXpcIUofhvnKUlfo9gLWpDWoSXAIT3UcxPFoaXpEfF86tfTOjiN4ljyQorjnq/6Z6yisdx/o2jDkyga0P0OxYmdf+02eq11Iysinldx4u789B6smV73A1R5X3oz7/5y60Fm/ZDO6p4bETNbXYuVZ0juepv1l4prCh8A/u6QHPy8RWlmluEtSjOzDAelmVmGg9LMLMNBaWaW4aA0M8twUJqZZfw/VWc8O+EpO0AAAAAASUVORK5CYII=\n",
      "text/plain": [
       "<Figure size 360x360 with 2 Axes>"
      ]
     },
     "metadata": {
      "needs_background": "light"
     },
     "output_type": "display_data"
    }
   ],
   "source": [
    "plot_S7_Poisson('S6.d9', 50, '-9 bp') "
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "The insertion libraries have a simpler distribution, no split axes needed."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 220,
   "metadata": {},
   "outputs": [],
   "source": [
    "def plot_S7_Poisson_ins(library, cutoff,title):\n",
    "    hist_data = df_hist_Poisson[library].iloc[:cutoff + 1]\n",
    "    main_max = hist_data.max()\n",
    "    axes_split = hist_data[2]\n",
    "    after_trunc = df_hist_Poisson[library].iloc[cutoff:].sum()\n",
    "\n",
    "    hist_data.iloc[cutoff] = after_trunc # that has the data with the last entry in series being cumulative\n",
    "\n",
    "    \n",
    "    f, ax2 = plt.subplots(1, figsize=(5,5))\n",
    "\n",
    "    f.suptitle(title)\n",
    "    \n",
    "    bins = list(range(1, cutoff+1)) + [cutoff + 3]\n",
    "\n",
    "    ax2.bar(bins,hist_data, color='purple')\n",
    " \n",
    "    ax2.set_ylabel('Count (X)')\n",
    "    ax2.set_xlabel('Number of mutations in library observed X-times')\n",
    "\n",
    "\n",
    "    # set the tick label for overflow bin\n",
    "    xlabels_pos = [1,2,3,4,5,6,7,8,9,10,12]\n",
    "    xlabels = [str(pos) for pos in xlabels_pos]\n",
    "    xlabels[-1] = str(cutoff) + '+'\n",
    "    ax2.xaxis.set_ticks(xlabels_pos)\n",
    "    ax2.set_xticklabels(xlabels)\n",
    "\n",
    "    f.savefig('FigS7_' + library +'.png', format='png', dpi=600)\n",
    "    plt.show()\n",
    "\n",
    "\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 223,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "image/png": "iVBORw0KGgoAAAANSUhEUgAAAVMAAAFhCAYAAADN+UnSAAAABHNCSVQICAgIfAhkiAAAAAlwSFlzAAALEgAACxIB0t1+/AAAADh0RVh0U29mdHdhcmUAbWF0cGxvdGxpYiB2ZXJzaW9uMy4xLjAsIGh0dHA6Ly9tYXRwbG90bGliLm9yZy+17YcXAAAenklEQVR4nO3deZRlZX3u8e9j04ojgzQsZLiNpuOYiNpBItGr4AUHlhgjEWO09RJJlDhlxJuhQWOiSYxGE3UhEnGISFRCOyIyaLy5Io0ig4iNitJCoBXBgYgCv/vHfgsORU1d9Z7uqub7WavW2fvd+7z1O0M9tYez35OqQpK0MHfb2gVI0rbAMJWkDgxTSerAMJWkDgxTSerAMJWkDgxT3aUleWGSz2/tOrT0Gaba5iR5X5Krk/wwydeT/M7WrknbPsNUS1KSY5McO83ivwFWVtX9gGcAf5XkMVusON0lGaba5lTVJVV108Rs+3nQDHdJkrcmuSHJ15IcNLLgnCR/k+SLbflpSXYeZ/1amgxTbZOSvC3JjcDXgKuBT8yw+mOBbwK7AGuBj0wKzBcA/xt4AHAz8JaxFK0lzTDVNqmqXgrcF3g88BHgphlWvxZ4c1X9vKo+CFwGPH1k+Xur6uKq+gnwF8BvJlk2ptK1RBmmWjKSfCzJ9UmuB44BjpmYT/KxyetX1S1V9XlgT+AlM3T93brjiD/fZtgKnXDlpGXLGbZipdtst7ULkOaqqg6dmJ44+VRVx87hrtsx8zHTPZJkJFD3BtaNLN9rZHpv4OfA9+bwe3UX4paptilJdk1yRJL7JFmW5BDgucBZM9xtV+DlSZYnORx4KHc8xvrbSR6W5F7Aa4APVdUtY3sQWpLcMtW2phh26d/BsLHwbeCVVXXaDPc5F1jFsLV5DfDsqvr+yPL3Au8GHgJ8lpkPGeguKg4OLU0vyTnA+6rqhK1dixY3d/MlqQPDVJI6cDdfkjpwy1SSOjBMJakDw1SSOjBMJakDw1SSOjBMJakDw1SSOjBMJakDw1SSOjBMJakDw1SSOjBMJakDw1SSOjBMJakDw1SSOjBMJakDw1SSOjBMJakDw1SSOjBMJakDw1SSOjBMJamD7bZ2AeOwyy671MqVK7d2GZK2Meeff/73qmrFVMu2yTBduXIl69ev39plSNrGJPn2dMvczZekDgxTSerAMJWkDgxTSerAMJWkDgxTSerAMJWkDgxTSerAMJWkDgxTSerAMJWkDrbJa/Pn47gct6D7r621nSqRtBS5ZSpJHRimktSBYSpJHRimktSBYSpJHRimktSBYSpJHRimktSBYSpJHRimktSBYSpJHRimktSBYSpJHRimktSBYSpJHRimktSBYSpJHRimktSBYSpJHYw1TJNckeSiJBckWd/adk5yRpIN7Xan1p4kb0lyeZILkzx6pJ81bf0NSdaMs2ZJmo8tsWX6pKrat6pWt/ljgDOrahVwZpsHeCqwqv0cBbwdhvAF1gKPBfYD1k4EsCQtFltjN/8w4KQ2fRLwzJH299TgC8COSXYHDgHOqKrrquoHwBnAU7Z00ZI0k3GHaQGfTnJ+kqNa225VdTVAu921te8BXDly342tbbp2SVo0thtz/wdU1VVJdgXOSPK1GdbNFG01Q/sd7zyE9VEAe++993xqlaR5G+uWaVVd1W6vBU5lOOZ5Tdt9p91e21bfCOw1cvc9gatmaJ/8u46vqtVVtXrFihW9H4okzWhsYZrk3knuOzENHAxcDKwDJs7IrwFOa9PrgBe0s/r7Aze0wwCnAwcn2amdeDq4tUnSojHO3fzdgFOTTPyef62qTyU5DzglyZHAd4DD2/qfAJ4GXA7cCLwIoKquS/Ja4Ly23muq6rox1i1Jm21sYVpV3wQeOUX794GDpmgv4Ohp+joROLF3jZLUi1dASVIHhqkkdWCYSlIHhqkkdWCYSlIHhqkkdWCYSlIHhqkkdWCYSlIHhqkkdWCYSlIHhqkkdWCYSlIHhqkkdWCYSlIHhqkkdWCYSlIHhqkkdWCYSlIHhqkkdWCYSlIHhqkkdWCYSlIHhqkkdWCYSlIHhqkkdWCYSlIHhqkkdWCYSlIHhqkkdWCYSlIHhqkkdWCYSlIHhqkkdWCYSlIHhqkkdWCYSlIHhqkkdTD2ME2yLMmXk3ysze+T5NwkG5J8MMndW/s92vzlbfnKkT5e3dovS3LIuGuWpM21JbZMXwFcOjL/BuBNVbUK+AFwZGs/EvhBVf0C8Ka2HkkeBhwBPBx4CvC2JMu2QN2SNGdjDdMkewJPB05o8wEOBD7UVjkJeGabPqzN05Yf1NY/DDi5qm6qqm8BlwP7jbNuSdpc494yfTPwJ8Ctbf7+wPVVdXOb3wjs0ab3AK4EaMtvaOvf1j7FfSRpURhbmCY5FLi2qs4fbZ5i1Zpl2Uz3Gf19RyVZn2T9pk2bNrteSVqIcW6ZHgA8I8kVwMkMu/dvBnZMsl1bZ0/gqja9EdgLoC3fAbhutH2K+9ymqo6vqtVVtXrFihX9H40kzWBsYVpVr66qPatqJcMJpLOq6nnA2cCz22prgNPa9Lo2T1t+VlVVaz+ine3fB1gFfHFcdUvSfGw3+yrd/SlwcpK/Ar4MvKu1vwt4b5LLGbZIjwCoqkuSnAJ8FbgZOLqqbtnyZUvS9LZImFbVOcA5bfqbTHE2vqp+Chw+zf1fB7xufBVK0sJ4BZQkdWCYSlIHhqkkdWCYSlIHhqkkdWCYSlIHhqkkdWCYSlIHhqkkdWCYSlIHhqkkdWCYSlIHhqkkdWCYSlIHhqkkdWCYSlIHhqkkdWCYSlIHhqkkdWCYSlIHhqkkdWCYSlIHhqkkdWCYSlIHhqkkdbDdXFZKsitwAPAA4L+Bi4H1VXXrGGtb0o7LcQu6/9pa26kSSVvCjGGa5EnAMcDOwJeBa4HtgWcCD0ryIeCNVfXDcRcqSYvZbFumTwNeXFXfmbwgyXbAocD/Aj48htokacmYMUyr6o9nWHYz8O/dK5KkJWjGE1BJ3prkvlO0PyTJZ8ZXliQtLbOdzf8v4IIkvwWQ5F5J/hZYB/zzuIuTpKVixjCtqtcBTwael+RzwIXAzcC+VXXqFqhPkpaEuXzOtNrtdm39S6vqxvGVJElLz2zHTP8c+Azwnqp6HPB44LAkn03ysC1RoCQtBbN9NGoF8Kiq+hFAVX0XeHaSpzJ8HOqhY65PkpaE2Y6ZvmIiSCe1fxLYd2xVSdISM+tufpKdp1pWVTclOTDJoeMpTZKWjtl28y8CPprkp8CXgE0Ml5OuYtgy/Qzw12OtUJKWgNl280+rqgOA3wMuAZYBPwTeB+xXVa+qqk1T3TfJ9km+mOQrSS5JhpE/kuyT5NwkG5J8MMndW/s92vzlbfnKkb5e3dovS3JIjwcuST3NadSoqtoAbNjMvm8CDqyqHydZDnw+ySeBPwDeVFUnJ3kHcCTw9nb7g6r6hSRHAG8AntM+NXAE8HCGUas+k+QXq+qWzaxHksZmbOOZ1uDHbXZ5+yngQOBDrf0khhGoAA5r87TlByVJaz+5qm6qqm8BlwP7jatuSZqPsQ4OnWRZkgsYhu47A/gGcH0bJAVgI7BHm94DuBJuG0TlBuD+o+1T3EeSFoU5hWmSA+bSNllV3VJV+wJ7MmxNTvW51IkrrDLNsunaJ9dzVJL1SdZv2jTlYVxJGpu5bpm+dY5tU6qq64FzgP2BHdtYqDCE7FVteiOwF9w2VuoOwHWj7VPcZ/R3HF9Vq6tq9YoVK+ZamiR1MdtI+78KPA5YkeQPRhbdj+HM/kz3XQH8vKquT3JPhgFT3gCcDTwbOBlYA5zW7rKuzf+/tvysqqok64B/TfIPDCegVgFf3KxHKUljNtvZ/LsD92nrjY5r+kOGwJvJ7sBJSZYxbAGfUlUfS/JV4OQkf8XwVSjvauu/C3hvkssZtkiPAKiqS5KcAnyVYcSqoz2TL2mxmW2k/c8Cn03y7qr69uZ0XFUXAo+aov2bTHE2vqp+Chw+TV+vA163Ob9fkrakOX3OFLhHkuOBlaP3qaoDx1GUJM3FYvoW4LmG6b8B7wBOANzFlqRJ5hqmN1fV28daiSQtYXP9aNRHk7w0ye5Jdp74GWtlkrSEzHXLdE27Hf3q5wIe2LccSVqa5jrQyT7jLkSSlrI5hWmSF0zVXlXv6VuOJC1Nc93N/5WR6e2BgxgGizZMJYm57+a/bHQ+yQ7Ae8dSkSQtQfMdgu9GhmvkJUnM/ZjpR7l92LtlDEPpnTKuoiRpqZnrMdO/H5m+Gfh2VW0cQz2StCTNaTe/DXjyNYaRo3YCfjbOoiRpqZnrSPu/yTCG6OHAbwLnJpltCD5JusuY627+nwG/UlXXwm0DP3+G278YT5Lu0uZ6Nv9uE0HafH8z7itJ27y5bpl+KsnpwAfa/HOAT46nJElaeub6of0/TvIs4NcYvi30+Ko6dayVSdISMtsX6v0CsFtV/d+q+gjwkdb+hCQPqqpvbIkiJWmxm+2455uBH03RfmNbJkli9jBd2b4Y7w6qaj3D90FJkpg9TLefYdk9exYiSUvZbGF6XpIXT25MciRw/nhKkqSlZ7az+a8ETk3yPG4Pz9XA3YFfH2dhkrSUzBimVXUN8LgkTwIe0Zo/XlVnjb0ySVpC5vo507OBs8dciyQtWV4SKkkdGKaS1IFhKkkdGKaS1MFcR43SVnZcjpv3fdfW2o6VSJqKW6aS1IFhKkkdGKaS1IFhKkkdGKaS1IFhKkkdGKaS1MHYwjTJXknOTnJpkkuSvKK175zkjCQb2u1OrT1J3pLk8iQXJnn0SF9r2vobkqwZV82SNF/j3DK9GfjDqnoosD9wdJKHAccAZ1bVKuDMNg/wVGBV+zkKeDsM4QusBR4L7AesnQhgSVosxhamVXV1VX2pTf8IuBTYAzgMOKmtdhLwzDZ9GPCeGnwB2DHJ7sAhwBlVdV1V/QA4A3jKuOqWpPnYIsdMk6wEHgWcy/DV0VfDELjArm21PYArR+62sbVN1y5Ji8bYwzTJfYAPA6+sqh/OtOoUbTVD++Tfc1SS9UnWb9q0aX7FStI8jTVMkyxnCNL3V9VHWvM1bfeddntta98I7DVy9z2Bq2Zov4OqOr6qVlfV6hUrVvR9IJI0i3GezQ/wLuDSqvqHkUXrgIkz8muA00baX9DO6u8P3NAOA5wOHJxkp3bi6eDWJkmLxjiH4DsAeD5wUZILWtv/AV4PnNK+Lvo7wOFt2SeApwGXAzcCLwKoquuSvBY4r633mqq6box1S9JmG1uYVtXnmfp4J8BBU6xfwNHT9HUicGK/6iSpL6+AkqQODFNJ6sAwlaQODFNJ6sAwlaQODFNJ6sAwlaQOxvmhfS1Sx+W4Bd1/ba3tVIm07XDLVJI6MEwlqQPDVJI6MEwlqQPDVJI6MEwlqQPDVJI6MEwlqQPDVJI6MEwlqQPDVJI6MEwlqQPDVJI6MEwlqQPDVJI6MEwlqQPDVJI6MEwlqQPDVJI6MEwlqQPDVJI68NtJtWB+26nklqkkdWGYSlIHhqkkdWCYSlIHhqkkdWCYSlIHhqkkdWCYSlIHYwvTJCcmuTbJxSNtOyc5I8mGdrtTa0+StyS5PMmFSR49cp81bf0NSdaMq15JWohxbpm+G3jKpLZjgDOrahVwZpsHeCqwqv0cBbwdhvAF1gKPBfYD1k4EsCQtJmML06r6HHDdpObDgJPa9EnAM0fa31ODLwA7JtkdOAQ4o6quq6ofAGdw54CWpK1uS1+bv1tVXQ1QVVcn2bW17wFcObLextY2Xbu2YV7rr6VosZyAyhRtNUP7nTtIjkqyPsn6TZs2dS1OkmazpcP0mrb7Tru9trVvBPYaWW9P4KoZ2u+kqo6vqtVVtXrFihXdC5ekmWzpMF0HTJyRXwOcNtL+gnZWf3/ghnY44HTg4CQ7tRNPB7c2SVpUxnbMNMkHgCcCuyTZyHBW/vXAKUmOBL4DHN5W/wTwNOBy4EbgRQBVdV2S1wLntfVeU1WTT2pJ0lY3tjCtqudOs+igKdYt4Ohp+jkROLFjaZLU3WI5ASVJS5phKkkdGKaS1IFfqKdtnhcBaEtwy1SSOjBMJakDw1SSOjBMJakDw1SSOvBsvrSZ/HSApuKWqSR1YJhKUgeGqSR1YJhKUgeGqSR1YJhKUgeGqSR14OdMpa1sIZ9b9TOri4dbppLUgWEqSR0YppLUgcdMpW2I4wZsPW6ZSlIHhqkkdWCYSlIHhqkkdWCYSlIHhqkkdWCYSlIHhqkkdWCYSlIHhqkkdWCYSlIHhqkkdWCYSlIHhqkkdWCYSlIHhqkkdbBkwjTJU5JcluTyJMds7XokadSSCNMky4B/Bp4KPAx4bpKHbd2qJOl2SyJMgf2Ay6vqm1X1M+Bk4LCtXJMk3WaphOkewJUj8xtbmyQtCqmqrV3DrJIcDhxSVb/T5p8P7FdVLxtZ5yjgqDb7YOCyzmXsAnzvLtLfYq7trtbfYq5tHBZ7ff+jqlZMtWCpfDvpRmCvkfk9gatGV6iq44Hjx1VAkvVVtfqu0N9iru2u1t9irm0cFnt9M1kqu/nnAauS7JPk7sARwLqtXJMk3WZJbJlW1c1Jfh84HVgGnFhVl2zlsiTpNksiTAGq6hPAJ7ZiCb0PISzm/hZzbXe1/hZzbeOw2Oub1pI4ASVJi91SOWYqSYuaYTqLJCcmuTbJxR362ivJ2UkuTXJJklcssL/tk3wxyVdaf8cttMbW77IkX07ysQ59XZHkoiQXJFnfob8dk3woydfa8/ir8+znwa2miZ8fJnnlAmt7VXsdLk7ygSTbL7C/V7S+LplPbVO9d5PsnOSMJBva7U4LqXEp19ObYTq7dwNP6dTXzcAfVtVDgf2Boxd4WexNwIFV9UhgX+ApSfbvUOcrgEs79DPhSVW1b6ePvPwj8KmqegjwSOZZZ1Vd1mraF3gMcCNw6nyLSrIH8HJgdVU9guFE6REL6O8RwIsZrv57JHBoklWb2c27ufN79xjgzKpaBZzZ5reUedWT5N1Jnjj26hbIMJ1FVX0OuK5TX1dX1Zfa9I8YgmDeV3LV4Mdtdnn7WdBB8CR7Ak8HTlhIP+OQ5H7AE4B3AVTVz6rq+g5dHwR8o6q+vcB+tgPumWQ74F5M+iz0Znoo8IWqurGqbgY+C/z65nQwzXv3MOCkNn0S8MwF1LhZFls9vRmmW0mSlcCjgHMX2M+yJBcA1wJnVNWC+gPeDPwJcOsC+5lQwKeTnN+uUluIBwKbgH9phyFOSHLvhZfIEcAHFtJBVX0X+HvgO8DVwA1V9ekFdHkx8IQk909yL+Bp3PHClfnaraquhuGfO7Brhz63pXrmzTDdCpLcB/gw8Mqq+uFC+qqqW9qu6p7Afm33cL51HQpcW1XnL6SmSQ6oqkczjPh1dJInLKCv7YBHA2+vqkcBP2GBu6ntIpBnAP+2wH52YtjK2gd4AHDvJL893/6q6lLgDcAZwKeArzAcJrpLSHLIxPFshtfnhDa/0I2FsTFMt7AkyxmC9P1V9ZFe/bbd3XNY2PHdA4BnJLmCYWSuA5O8b4F1XdVur2U4JrnfArrbCGwc2fr+EEO4LsRTgS9V1TUL7OfJwLeqalNV/Rz4CPC4hXRYVe+qqkdX1RMYdo83LLBGgGuS7A7Qbq/t0Gf3eqrq9JFj2uuA32nzj92Ktc7IMN2CkoTheN+lVfUPHfpbkWTHNn1Phj/or823v6p6dVXtWVUrGXZ9z6qqeW9dJbl3kvtOTAMHM+y+zre+/wKuTPLg1nQQ8NX59tc8lwXu4jffAfZPcq/2Oh/EAk/iJdm13e4NPIs+da4D1rTpNcBpHfpciMVWz/xVlT8z/DC8ga8Gfs6wZXTkAvr6NYZjiBcCF7Sfpy2gv18Gvtz6uxj4y46P+4nAxxbYxwMZdk+/AlwC/FmHuvYF1rfH/O/ATgvo617A94EdOj1nxzH8M7sYeC9wjwX29x8M/yy+Ahw0j/vf6b0L3J/hrPmGdrtzr/fMuOph+BTAE7dUnfP98QooSerA3XxJ6sAwlaQODFNJ6sAwlaQODFNJ6sAwnUWSSvLGkfk/SnJsp77fneTZPfqa5fcc3kZYOrtzvyuT/NbmrpdkdZK39Kxl0u97TZInb8b6T5wYISvJM5Ic06a3yOszV0l+PPtaW9Z0z1GSdRm++HJi/p1J/niK9bboe2OcDNPZ3QQ8K8kuW7uQUUmWbcbqRwIvraondS5jJTBrmE5er6rWV9XLO9dym6r6y6r6zDzvu66qXj/X9dugJnNZL0kW1d/bXGufp5cDr8kwZOLjgMcyjPsw2Uq24HtjnBbVi7tI3czwVQqvmrxg8n/liS2HtqXz2SSnJPl6ktcneV6GsUcvSvKgkW6enOQ/2nqHtvsvS/J3Sc5LcmGS3x3p9+wk/wpcNEU9z239X5zkDa3tLxkuFnhHkr+btP6c6pzucQKvBx7frpl+VdvK+I8kX2o/j5tmvdEtwZ2T/Ht7nF9I8sut/dgM41+ek+SbSV7e2u+d5OMZxnC9OMlzZnpdMoynelyr56IkD5n6Zb7tvi9M8k+zvD4vTPJvST7KMIjLfZKcOfI7DmvrrcywR/A24EvAXyR508jvenGSO10JN9XrOLLsje33nJlkRWt7eZKvtufw5JHn6cT2HvrySE2Ta/9gkqdNeu5+Y4b3YJL8U/t9H2eagUmq6gqGv5u/Bd4G/H4Nl9lONtN749gkJyX5dHsdn5Xkb9tz86kMl2aT5DHtfXx+ktNz++Wpd3pexmprXzWw2H+AHwP3A64AdgD+CDh25MqMZ4+u226fCFwP7A7cA/gucFxb9grgzSP3/xTDP7VVDFeFbA8cBfx5W+ceDFf87NP6/QmwzxR1PoDhksYVDAOCnAU8sy07h2Gczcn32Zw6p3ucHxtpvxewfZteBayfZr3b5oG3Amvb9IHABW36WOA/W127MFyptBz4DeCdI33d6eql0Xrb6/ayNv1S4IRpnoeJel4I/NMsr88L2/TObb3tgPu16V2Ay4EwbHXdCuzflt0b+AawvM3/J/BLm/E6FvC8Nv2XI3VeRbvaCtix3f418NsTbcDX2++fXPuvAye16bsDVwL3ZPr34LMYBl9Z1mq9npH3xqTHsrw9lvfP8Pc103vjWODzrZ9HMow5+9S27FSG4fqWt+dxRWt/DsMXbk75vIzzxy3TOahhZKf3MOy6zNV5NYxfehPDH9DEcGwXMfyRTTilqm6tqg3AN4GHMFzD/oIMI+acy3DJ3cTAwF+sqm9N8ft+BTinhoE2bgbezzD2Z68652I58M4kFzGMwjSXga9/jeHSS6rqLOD+SXZoyz5eVTdV1fcYBsDYrdX15CRvSPL4qrphDr9jYkCZ89n8xzTV6wPDcIcTY3MG+OskFwKfYRijdre27NtV9YX2+H7CEI6Hti3k5VU1eQ9jptfxVuCDbfp9DM8dDJfWvj/DKFUTI0sdDBzT3kPnMPwT2HuK2j/JMKDNPRgGfflcVf03078HnwB8oIbRyq5qj2c6v9yem4dk/oc4PlnDFu1FDAH+qdY+8f58MPAI4IxW658zjKAGUz8vY2OYzt2bGY49jo6feTPtOUwShv/sE24amb51ZP5W7vitsJOv5y2GN+DLqo2aU1X71O1jY/5kmvoy1wcyyVzqnOlxjnoVcA3DVsTqGdYbNVXdE8/JaG23ANtV1dcZRsa/CPibDIcxZjPRzy1s/jfyTvX6wB1fh+cxbEk+poZRjq5hCK/J68Ew6PYLgRcB/zLF79uc13GilqcD/8zwvJyf4VhogN8YeQ/tXcOwfneoqap+yhC2hzBs1U3sDs/0Hpz1GvQWnm8Dns9w3f1LWvtjc/tXxTxjDo/xplbnrcDPq21mcvv7M8AlI3X+UlUdPMPzMjaG6Ry1/+SnMATqhCsYXigYxrJcPo+uD09ytwzHJx8IXAacDrxk5JjQL2b2QZDPBf5nkl0ynJx6LsPo7D1cwdSP80fAfUfW2wG4ur3xn8+wJTHVeqM+xxBGZPhqiu/VDGO8JnkAcGNVvY9hMOaFDsE3m6len8l2YBgH9udJngT8j+k6q2H4wL0YTrpMNQrUTK/j3YCJY9e/BXy+hdZeVXU2w6DeOwL3YXgPvaz98yPJo2Z4jCczhPvj2/1g+vfg54Aj2jHV3YHpTmr+LrChqs4B/gD4kyQrqurckeBbx8zvjbm4DFiR9l1gSZYnefgMz8vYjDWpt0FvBH5/ZP6dwGlJvsgw4s10W40zuYzhj2U34Peq6qdJTmDYhflS+2PYxCxf51BVVyd5NXA2w3/rT1RVr+HMpnucFwI3J/kKw/HFtwEfTnJ4q2O69b480vexDCPnX8hwTGwNM/sl4O+S3Mow+tBLFvTIZjfV6zN5nfcDH83whYEXMPswiKcA+1bVDyYvmOV1/Anw8CTnAzcwbEkuA97XDo0EeFNVXZ/ktQx7Uxe299AVwKHT1PNphsNY66rqZ61tuvfgqQzHti9iOA57p3/YGYYO/FOG7zmjqq5K8o8MJ6NeNGn1md4bs6qqn2U42fiW9hxs1x7315niedmcvjeXo0ZJW1g7W/2mqjpza9eiftzNl7aQDJ+5/Drw3wbptsctU0nqwC1TSerAMJWkDgxTSerAMJWkDgxTSerAMJWkDv4/UC6YvwJHW68AAAAASUVORK5CYII=\n",
      "text/plain": [
       "<Figure size 360x360 with 1 Axes>"
      ]
     },
     "metadata": {
      "needs_background": "light"
     },
     "output_type": "display_data"
    }
   ],
   "source": [
    "plot_S7_Poisson_ins('S6.i3', 10, '+3 bp') "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 234,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "ACGT composition of insertions, for each position separately\n",
      "{3: {0: {'A': 4586, 'C': 3076, 'G': 3827, 'T': 3840},\n",
      "     1: {'A': 4359, 'C': 3328, 'G': 3434, 'T': 4208},\n",
      "     2: {'A': 3531, 'C': 3643, 'G': 3094, 'T': 5061}},\n",
      " 6: {0: {'A': 5057, 'C': 4941, 'G': 9202, 'T': 6622},\n",
      "     1: {'A': 5046, 'C': 5273, 'G': 8572, 'T': 6931},\n",
      "     2: {'A': 4982, 'C': 5071, 'G': 8194, 'T': 7575},\n",
      "     3: {'A': 5427, 'C': 4955, 'G': 7761, 'T': 7679},\n",
      "     4: {'A': 5596, 'C': 4631, 'G': 8005, 'T': 7590},\n",
      "     5: {'A': 3627, 'C': 5589, 'G': 6323, 'T': 10283}},\n",
      " 9: {0: {'A': 2122, 'C': 1903, 'G': 3591, 'T': 2443},\n",
      "     1: {'A': 2070, 'C': 2193, 'G': 3208, 'T': 2588},\n",
      "     2: {'A': 1992, 'C': 2139, 'G': 3128, 'T': 2800},\n",
      "     3: {'A': 2153, 'C': 2145, 'G': 3007, 'T': 2754},\n",
      "     4: {'A': 2042, 'C': 2208, 'G': 3092, 'T': 2717},\n",
      "     5: {'A': 2089, 'C': 2032, 'G': 2948, 'T': 2990},\n",
      "     6: {'A': 2156, 'C': 1844, 'G': 3019, 'T': 3040},\n",
      "     7: {'A': 2228, 'C': 1776, 'G': 3087, 'T': 2968},\n",
      "     8: {'A': 1496, 'C': 2060, 'G': 2560, 'T': 3943}}}\n",
      "\n"
     ]
    }
   ],
   "source": [
    "def insertion_composition(all_references, cutoff=2, l=(3,6,9)):\n",
    "\n",
    "    comp = {length: {k: {'A':0, 'C':0, 'T':0, 'G':0} for k in range(length)} for length in l}\n",
    "    for background in all_references.keys():\n",
    "        for fraction in all_references[background].keys():\n",
    "            for mutation in all_references[background][fraction].keys():\n",
    "                for dna_error, c in all_references[background][fraction][mutation]['dna'].items():\n",
    "                    if dna_error is None:\n",
    "                        continue\n",
    "                    if c >= cutoff and len(dna_error) == 1 and classify_dna(dna_error) in ('i3', 'i6', 'i9'):\n",
    "                        ins = dna_error[0][2]\n",
    "                        ins_len = len(ins)\n",
    "                        if ins_len in l:\n",
    "                            for pos in range(ins_len):\n",
    "                                comp[ins_len][pos][ins[pos]] += 1\n",
    "    return comp\n",
    "\n",
    "\n",
    "# # 4. Get position by position ACGT composition of insertions - SI Figure\n",
    "ins_composition = insertion_composition(all_ref)\n",
    "print('ACGT composition of insertions, for each position separately')\n",
    "pprint.pprint(ins_composition)\n",
    "print()\n",
    "#"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "def rev_comp(nt):\n",
    "    complement = {'A': 'T', 'T': 'A', 'G': 'C', 'C': 'G'}\n",
    "    return complement[nt]\n",
    "\n",
    "\n",
    "def transposon_consensus_seq(all_references, reference, fraction='d3', transposon='d3', start_offset=3):\n",
    "    \"\"\"\n",
    "    Determine the consensus sequence for transposon insertion, by analysing the position of d3 or i3 mutations\n",
    "    :param all_references: dictionary containing all mutation data\n",
    "    :param reference: BioSeq fasta reference\n",
    "    :param fraction: name of the activity fraction / library to be analysed\n",
    "    :param transposon: which type of mutations are we counting\n",
    "    :return: dict with composition by position\n",
    "    \"\"\"\n",
    "    consensus = {pos: {'A': 0, 'T': 0, 'C': 0, 'G': 0} for pos in range(5)}\n",
    "    baseline = {pos: {'A': 0, 'T': 0, 'C': 0, 'G': 0} for pos in range(5)}\n",
    "    ref = SeqIO.read(reference, 'fasta')\n",
    "\n",
    "    for i in range(2, len(ref)-7):\n",
    "        trans_seq = str(ref[i:i+5].seq)\n",
    "        for pos in range(5):\n",
    "            baseline[pos][trans_seq[pos]] += 1\n",
    "            baseline[pos][rev_comp(trans_seq[4 - pos])] += 1\n",
    "\n",
    "    background = str(ref.name)\n",
    "\n",
    "    deletions = all_references[background][fraction]\n",
    "    for prot_mutation in deletions.keys():\n",
    "        for dna_mutation, count in deletions[prot_mutation]['dna'].items():\n",
    "            if classify_dna(dna_mutation) == transposon:  # found the simple mutations: d3 or i3\n",
    "                n1 = int(dna_mutation[0][0]) + start_offset - 2\n",
    "                # this position refers to the nt in the first position of  the indel in 1-count:\n",
    "                # hence this is the 2nd nt of 5 nt site -> need to retrieve start-1:start+4\n",
    "                # Correct for offset: if start = 1: this is ref[start_offset]\n",
    "                # Therefore N1 = start_offset + start -2\n",
    "                if transposon == 'd3':\n",
    "                    trans_seq = str(ref[n1:n1+5].seq)\n",
    "                if len(trans_seq) != 5:\n",
    "                    continue\n",
    "                for pos in range(5):\n",
    "                    consensus[pos][trans_seq[pos]] += count\n",
    "                    consensus[pos][rev_comp(trans_seq[4 - pos])] += count\n",
    "\n",
    "    return baseline, consensus\n",
    "\n",
    "# 3. Find TransDel consensus sequence in -3 bp library - gives number of observations for the NNNNN target site\n",
    "# This is used for WebLogo in Fig. 3A\n",
    "d3_baseline, d3_cons = transposon_consensus_seq(all_ref, args.reference, fraction='d3', start_offset=args.start_offset)\n",
    "print('TransDel consensus sequence: baseline counts (reflect GC composition)')\n",
    "pprint.pprint(d3_baseline)\n",
    "print('TransDel consensus counts')\n",
    "pprint.pprint(d3_cons)\n",
    "print()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.7.3"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
